// Purescriptify.Root/index.js
"use strict";
var Control_Applicative = require("../Control.Applicative/index.js");
var Control_Bind = require("../Control.Bind/index.js");
var Control_Monad_State_Class = require("../Control.Monad.State.Class/index.js");
var Data_Maybe = require("../Data.Maybe/index.js");
var Data_Unit = require("../Data.Unit/index.js");
var Foreign_Toast = require("../Foreign.Toast/index.js");
var Purescriptify_Capability_Halo = require("../Purescriptify.Capability.Halo/index.js");
var Purescriptify_Capability_Routing = require("../Purescriptify.Capability.Routing/index.js");
var Purescriptify_Component_GlobalContext = require("../Purescriptify.Component.GlobalContext/index.js");
var Purescriptify_Data_Route = require("../Purescriptify.Data.Route/index.js");
var Purescriptify_Page_Home = require("../Purescriptify.Page.Home/index.js");
var React_Basic = require("../React.Basic/index.js");
var React_Basic_Hooks_Internal = require("../React.Basic.Hooks.Internal/index.js");
var React_Halo_Internal_Control = require("../React.Halo.Internal.Control/index.js");
var React_Halo_Internal_Eval = require("../React.Halo.Internal.Eval/index.js");
var Type_Equality = require("../Type.Equality/index.js");
var Initialize = (function () {
    function Initialize() {

    };
    Initialize.value = new Initialize();
    return Initialize;
})();
var UpdateRoute = (function () {
    function UpdateRoute(value0) {
        this.value0 = value0;
    };
    UpdateRoute.create = function (value0) {
        return new UpdateRoute(value0);
    };
    return UpdateRoute;
})();
var Navigate = (function () {
    function Navigate(value0) {
        this.value0 = value0;
    };
    Navigate.create = function (value0) {
        return new Navigate(value0);
    };
    return Navigate;
})();
var mkRoot = function (dictMonadAff) {
    return function (dictMonadGlobal) {
        return function (dictMonadRouting) {
            return function (dictMonadHalo) {
                return function (dictMonadUsers) {
                    return function (dictNow) {
                        return function (dictLogMessages) {
                            var mkRender = Control_Bind.bind((dictLogMessages.Monad0()).Bind1())(Purescriptify_Page_Home.mkHomePage(dictMonadAff.MonadEffect0())(dictMonadUsers)(dictNow)(dictLogMessages)(dictMonadHalo))(function (homePage) {
                                return Control_Bind.bind((dictLogMessages.Monad0()).Bind1())(Purescriptify_Component_GlobalContext.mkGlobalContext(dictMonadGlobal)(dictMonadHalo)(dictMonadAff.MonadEffect0()))(function (globalContext) {
                                    return Control_Applicative.pure((dictLogMessages.Monad0()).Applicative0())(function (v) {
                                        var contents = (function () {
                                            if (v.state.route instanceof Purescriptify_Data_Route.Home) {
                                                return homePage({});
                                            };
                                            if (v.state.route instanceof Purescriptify_Data_Route["Error"]) {
                                                return homePage({});
                                            };
                                            throw new Error("Failed pattern match at Purescriptify.Root (line 79, column 24 - line 81, column 35): " + [ v.state.route.constructor.name ]);
                                        })();
                                        return React_Basic.fragment([ globalContext(Data_Unit.unit), Foreign_Toast.toastContainer, contents ]);
                                    });
                                });
                            });
                            var initialState = function (v) {
                                return function (v1) {
                                    return {
                                        route: Purescriptify_Data_Route["Error"].value
                                    };
                                };
                            };
                            var handleAction = function (v) {
                                if (v instanceof Initialize) {
                                    return Control_Bind.discard(Control_Bind.discardUnit)(React_Halo_Internal_Control.bindHaloM)(Control_Bind.bindFlipped(React_Halo_Internal_Control.bindHaloM)(function ($24) {
                                        return handleAction(UpdateRoute.create($24));
                                    })(Purescriptify_Capability_Routing.read(Purescriptify_Capability_Routing.monadRoutingHaloM(dictMonadRouting))))(function () {
                                        return Purescriptify_Capability_Routing.subscribe(dictMonadRouting)(UpdateRoute.create);
                                    });
                                };
                                if (v instanceof UpdateRoute) {
                                    return Control_Monad_State_Class.modify_(React_Halo_Internal_Control.monadStateHaloM)(function (v1) {
                                        var $19 = {};
                                        for (var $20 in v1) {
                                            if ({}.hasOwnProperty.call(v1, $20)) {
                                                $19[$20] = v1[$20];
                                            };
                                        };
                                        $19.route = v.value0;
                                        return $19;
                                    });
                                };
                                if (v instanceof Navigate) {
                                    return Purescriptify_Capability_Routing.navigate(Purescriptify_Capability_Routing.monadRoutingHaloM(dictMonadRouting))(v.value0);
                                };
                                throw new Error("Failed pattern match at Purescriptify.Root (line 63, column 18 - line 71, column 29): " + [ v.constructor.name ]);
                            };
                            var $$eval = React_Halo_Internal_Eval.mkEval(function (v) {
                                return {
                                    onInitialize: function (v1) {
                                        return new Data_Maybe.Just(Initialize.value);
                                    },
                                    onUpdate: v.onUpdate,
                                    onAction: handleAction,
                                    onFinalize: v.onFinalize
                                };
                            });
                            var context = function (v) {
                                return Control_Applicative.pure(React_Basic_Hooks_Internal.applicativeRender(Type_Equality.refl))(Data_Unit.unit);
                            };
                            return Control_Bind.bind((dictLogMessages.Monad0()).Bind1())(mkRender)(function (render) {
                                return Purescriptify_Capability_Halo.component(dictMonadHalo)("Root")({
                                    context: context,
                                    initialState: initialState,
                                    "eval": $$eval,
                                    render: render
                                });
                            });
                        };
                    };
                };
            };
        };
    };
};
module.exports = {
    Initialize: Initialize,
    UpdateRoute: UpdateRoute,
    Navigate: Navigate,
    mkRoot: mkRoot
};
