// PureScript.CST.Range/index.js
"use strict";
var Control_Lazy = require("../Control.Lazy/index.js");
var Data_Array = require("../Data.Array/index.js");
var Data_Array_NonEmpty = require("../Data.Array.NonEmpty/index.js");
var Data_Array_NonEmpty_Internal = require("../Data.Array.NonEmpty.Internal/index.js");
var Data_Foldable = require("../Data.Foldable/index.js");
var Data_Maybe = require("../Data.Maybe/index.js");
var Data_Semigroup = require("../Data.Semigroup/index.js");
var Data_Tuple = require("../Data.Tuple/index.js");
var Data_Void = require("../Data.Void/index.js");
var PureScript_CST_Range_TokenList = require("../PureScript.CST.Range.TokenList/index.js");
var PureScript_CST_Types = require("../PureScript.CST.Types/index.js");
var tokensOfVoid = {
    tokensOf: Data_Void.absurd
};
var tokensOfRecoveredError = {
    tokensOf: function (v) {
        return PureScript_CST_Range_TokenList.fromArray(v.tokens);
    }
};
var tokensOfQualifiedName = {
    tokensOf: function (v) {
        return PureScript_CST_Range_TokenList.singleton(v.token);
    }
};
var tokensOfName = {
    tokensOf: function (v) {
        return PureScript_CST_Range_TokenList.singleton(v.token);
    }
};
var tokensOf = function (dict) {
    return dict.tokensOf;
};
var tokensOfArray = function (dictTokensOf) {
    return {
        tokensOf: Data_Foldable.foldMap(Data_Foldable.foldableArray)(PureScript_CST_Range_TokenList.monoidTokenList)(function (a) {
            return Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v) {
                return tokensOf(dictTokensOf)(a);
            });
        })
    };
};
var tokensOfFixityOp = {
    tokensOf: function (v) {
        if (v instanceof PureScript_CST_Types.FixityValue) {
            return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfQualifiedName)(v.value0))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(PureScript_CST_Range_TokenList.singleton(v.value1))(tokensOf(tokensOfName)(v.value2)));
        };
        if (v instanceof PureScript_CST_Types.FixityType) {
            return PureScript_CST_Range_TokenList.cons(v.value0)(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfQualifiedName)(v.value1))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(PureScript_CST_Range_TokenList.singleton(v.value2))(tokensOf(tokensOfName)(v.value3))));
        };
        throw new Error("Failed pattern match at PureScript.CST.Range (line 633, column 14 - line 637, column 59): " + [ v.constructor.name ]);
    }
};
var tokensOfLabeled = function (dictTokensOf) {
    return function (dictTokensOf1) {
        return {
            tokensOf: function (v) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(dictTokensOf)(v.label))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(PureScript_CST_Range_TokenList.singleton(v.separator))(tokensOf(dictTokensOf1)(v.value)));
            }
        };
    };
};
var tokensOfMaybe = function (dictTokensOf) {
    return {
        tokensOf: Data_Foldable.foldMap(Data_Foldable.foldableMaybe)(PureScript_CST_Range_TokenList.monoidTokenList)(tokensOf(dictTokensOf))
    };
};
var tokensOfNonEmptyArray = function (dictTokensOf) {
    return {
        tokensOf: Data_Foldable.foldMap(Data_Array_NonEmpty_Internal.foldableNonEmptyArray)(PureScript_CST_Range_TokenList.monoidTokenList)(function (a) {
            return Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v) {
                return tokensOf(dictTokensOf)(a);
            });
        })
    };
};
var tokensOfClassFundep = {
    tokensOf: function (v) {
        if (v instanceof PureScript_CST_Types.FundepDetermined) {
            return PureScript_CST_Range_TokenList.cons(v.value0)(tokensOf(tokensOfNonEmptyArray(tokensOfName))(v.value1));
        };
        if (v instanceof PureScript_CST_Types.FundepDetermines) {
            return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfNonEmptyArray(tokensOfName))(v.value0))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(PureScript_CST_Range_TokenList.singleton(v.value1))(tokensOf(tokensOfNonEmptyArray(tokensOfName))(v.value2)));
        };
        throw new Error("Failed pattern match at PureScript.CST.Range (line 554, column 14 - line 558, column 50): " + [ v.constructor.name ]);
    }
};
var tokensOfRecordLabeled = function (dictTokensOf) {
    return {
        tokensOf: function (v) {
            if (v instanceof PureScript_CST_Types.RecordPun) {
                return tokensOf(tokensOfName)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.RecordField) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfName)(v.value0))(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return PureScript_CST_Range_TokenList.cons(v.value1)(tokensOf(dictTokensOf)(v.value2));
                }));
            };
            throw new Error("Failed pattern match at PureScript.CST.Range (line 986, column 14 - line 990, column 52): " + [ v.constructor.name ]);
        }
    };
};
var tokensOfSeparated = function (dictTokensOf) {
    return {
        tokensOf: function (v) {
            return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(dictTokensOf)(v.head))(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                return Data_Foldable.foldMap(Data_Foldable.foldableArray)(PureScript_CST_Range_TokenList.monoidTokenList)(function (v2) {
                    return PureScript_CST_Range_TokenList.cons(v2.value0)(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v3) {
                        return tokensOf(dictTokensOf)(v2.value1);
                    }));
                })(v.tail);
            }));
        }
    };
};
var tokensOfTuple = function (dictTokensOf) {
    return function (dictTokensOf1) {
        return {
            tokensOf: function (v) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(dictTokensOf)(v.value0))(tokensOf(dictTokensOf1)(v.value1));
            }
        };
    };
};
var tokensOfWrapped = function (dictTokensOf) {
    return {
        tokensOf: function (v) {
            return PureScript_CST_Range_TokenList.wrap(v.open)(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                return tokensOf(dictTokensOf)(v.value);
            }))(v.close);
        }
    };
};
var tokensOfDataMembers = {
    tokensOf: function (v) {
        if (v instanceof PureScript_CST_Types.DataAll) {
            return PureScript_CST_Range_TokenList.singleton(v.value0);
        };
        if (v instanceof PureScript_CST_Types.DataEnumerated) {
            return tokensOf(tokensOfWrapped(tokensOfMaybe(tokensOfSeparated(tokensOfName))))(v.value0);
        };
        throw new Error("Failed pattern match at PureScript.CST.Range (line 313, column 14 - line 317, column 17): " + [ v.constructor.name ]);
    }
};
var tokensOfExport = function (dictTokensOf) {
    return {
        tokensOf: function (v) {
            if (v instanceof PureScript_CST_Types.ExportValue) {
                return tokensOf(tokensOfName)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.ExportOp) {
                return tokensOf(tokensOfName)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.ExportType) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfName)(v.value0))(Data_Foldable.foldMap(Data_Foldable.foldableMaybe)(PureScript_CST_Range_TokenList.monoidTokenList)(tokensOf(tokensOfDataMembers))(v.value1));
            };
            if (v instanceof PureScript_CST_Types.ExportTypeOp) {
                return PureScript_CST_Range_TokenList.cons(v.value0)(tokensOf(tokensOfName)(v.value1));
            };
            if (v instanceof PureScript_CST_Types.ExportClass) {
                return PureScript_CST_Range_TokenList.cons(v.value0)(tokensOf(tokensOfName)(v.value1));
            };
            if (v instanceof PureScript_CST_Types.ExportKind) {
                return PureScript_CST_Range_TokenList.cons(v.value0)(tokensOf(tokensOfName)(v.value1));
            };
            if (v instanceof PureScript_CST_Types.ExportModule) {
                return PureScript_CST_Range_TokenList.cons(v.value0)(tokensOf(tokensOfName)(v.value1));
            };
            if (v instanceof PureScript_CST_Types.ExportError) {
                return tokensOf(dictTokensOf)(v.value0);
            };
            throw new Error("Failed pattern match at PureScript.CST.Range (line 287, column 14 - line 303, column 17): " + [ v.constructor.name ]);
        }
    };
};
var tokensOfImport = function (dictTokensOf) {
    return {
        tokensOf: function (v) {
            if (v instanceof PureScript_CST_Types.ImportValue) {
                return tokensOf(tokensOfName)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.ImportOp) {
                return tokensOf(tokensOfName)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.ImportType) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfName)(v.value0))(Data_Foldable.foldMap(Data_Foldable.foldableMaybe)(PureScript_CST_Range_TokenList.monoidTokenList)(tokensOf(tokensOfDataMembers))(v.value1));
            };
            if (v instanceof PureScript_CST_Types.ImportTypeOp) {
                return PureScript_CST_Range_TokenList.cons(v.value0)(tokensOf(tokensOfName)(v.value1));
            };
            if (v instanceof PureScript_CST_Types.ImportClass) {
                return PureScript_CST_Range_TokenList.cons(v.value0)(tokensOf(tokensOfName)(v.value1));
            };
            if (v instanceof PureScript_CST_Types.ImportKind) {
                return PureScript_CST_Range_TokenList.cons(v.value0)(tokensOf(tokensOfName)(v.value1));
            };
            if (v instanceof PureScript_CST_Types.ImportError) {
                return tokensOf(dictTokensOf)(v.value0);
            };
            throw new Error("Failed pattern match at PureScript.CST.Range (line 372, column 14 - line 386, column 17): " + [ v.constructor.name ]);
        }
    };
};
var tokensOfImportDecl = function (dictTokensOf) {
    return {
        tokensOf: function (v) {
            return PureScript_CST_Range_TokenList.cons(v.keyword)(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfName)(v.module))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(Data_Foldable.foldMap(Data_Foldable.foldableMaybe)(PureScript_CST_Range_TokenList.monoidTokenList)(function (v2) {
                    return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(Data_Foldable.foldMap(Data_Foldable.foldableMaybe)(PureScript_CST_Range_TokenList.monoidTokenList)(PureScript_CST_Range_TokenList.singleton)(v2.value0))(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v3) {
                        return tokensOf(tokensOfWrapped(tokensOfSeparated(tokensOfImport(dictTokensOf))))(v2.value1);
                    }));
                })(v.names))(Data_Foldable.foldMap(Data_Foldable.foldableMaybe)(PureScript_CST_Range_TokenList.monoidTokenList)(function (v2) {
                    return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(PureScript_CST_Range_TokenList.singleton(v2.value0))(tokensOf(tokensOfName)(v2.value1));
                })(v.qualified)));
            }));
        }
    };
};
var tokensOfOneOrDelimited = function (dictTokensOf) {
    return {
        tokensOf: function (v) {
            if (v instanceof PureScript_CST_Types.One) {
                return tokensOf(dictTokensOf)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.Many) {
                return tokensOf(tokensOfWrapped(tokensOfSeparated(dictTokensOf)))(v.value0);
            };
            throw new Error("Failed pattern match at PureScript.CST.Range (line 130, column 14 - line 132, column 27): " + [ v.constructor.name ]);
        }
    };
};
var tokensOfTypeVarBinding = function (dictTokensOf) {
    return {
        tokensOf: function (v) {
            if (v instanceof PureScript_CST_Types.TypeVarKinded) {
                return tokensOf(tokensOfWrapped(tokensOfLabeled(tokensOfName)(tokensOfType(dictTokensOf))))(v.value0);
            };
            if (v instanceof PureScript_CST_Types.TypeVarName) {
                return tokensOf(tokensOfName)(v.value0);
            };
            throw new Error("Failed pattern match at PureScript.CST.Range (line 247, column 14 - line 251, column 17): " + [ v.constructor.name ]);
        }
    };
};
var tokensOfType = function (dictTokensOf) {
    return {
        tokensOf: function (v) {
            if (v instanceof PureScript_CST_Types.TypeVar) {
                return tokensOf(tokensOfName)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.TypeConstructor) {
                return tokensOf(tokensOfQualifiedName)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.TypeWildcard) {
                return PureScript_CST_Range_TokenList.singleton(v.value0);
            };
            if (v instanceof PureScript_CST_Types.TypeHole) {
                return tokensOf(tokensOfName)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.TypeString) {
                return PureScript_CST_Range_TokenList.singleton(v.value0);
            };
            if (v instanceof PureScript_CST_Types.TypeRow) {
                return tokensOf(tokensOfWrapped(tokensOfRow(dictTokensOf)))(v.value0);
            };
            if (v instanceof PureScript_CST_Types.TypeRecord) {
                return tokensOf(tokensOfWrapped(tokensOfRow(dictTokensOf)))(v.value0);
            };
            if (v instanceof PureScript_CST_Types.TypeForall) {
                return PureScript_CST_Range_TokenList.cons(v.value0)(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfNonEmptyArray(tokensOfTypeVarBinding(dictTokensOf)))(v.value1))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(PureScript_CST_Range_TokenList.singleton(v.value2))(tokensOf(tokensOfType(dictTokensOf))(v.value3)));
                }));
            };
            if (v instanceof PureScript_CST_Types.TypeKinded) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfType(dictTokensOf))(v.value0))(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(PureScript_CST_Range_TokenList.singleton(v.value1))(tokensOf(tokensOfType(dictTokensOf))(v.value2));
                }));
            };
            if (v instanceof PureScript_CST_Types.TypeApp) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfType(dictTokensOf))(v.value0))(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return tokensOf(tokensOfNonEmptyArray(tokensOfType(dictTokensOf)))(v.value1);
                }));
            };
            if (v instanceof PureScript_CST_Types.TypeOp) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfType(dictTokensOf))(v.value0))(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return Data_Foldable.foldMap(Data_Array_NonEmpty_Internal.foldableNonEmptyArray)(PureScript_CST_Range_TokenList.monoidTokenList)(function (v2) {
                        return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfQualifiedName)(v2.value0))(tokensOf(tokensOfType(dictTokensOf))(v2.value1));
                    })(v.value1);
                }));
            };
            if (v instanceof PureScript_CST_Types.TypeOpName) {
                return tokensOf(tokensOfQualifiedName)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.TypeArrow) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfType(dictTokensOf))(v.value0))(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(PureScript_CST_Range_TokenList.singleton(v.value1))(tokensOf(tokensOfType(dictTokensOf))(v.value2));
                }));
            };
            if (v instanceof PureScript_CST_Types.TypeArrowName) {
                return PureScript_CST_Range_TokenList.singleton(v.value0);
            };
            if (v instanceof PureScript_CST_Types.TypeConstrained) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfType(dictTokensOf))(v.value0))(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(PureScript_CST_Range_TokenList.singleton(v.value1))(tokensOf(tokensOfType(dictTokensOf))(v.value2));
                }));
            };
            if (v instanceof PureScript_CST_Types.TypeParens) {
                return tokensOf(tokensOfWrapped(tokensOfType(dictTokensOf)))(v.value0);
            };
            if (v instanceof PureScript_CST_Types.TypeUnaryRow) {
                return PureScript_CST_Range_TokenList.cons(v.value0)(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return tokensOf(tokensOfType(dictTokensOf))(v.value1);
                }));
            };
            if (v instanceof PureScript_CST_Types["TypeError"]) {
                return tokensOf(dictTokensOf)(v.value0);
            };
            throw new Error("Failed pattern match at PureScript.CST.Range (line 188, column 14 - line 232, column 17): " + [ v.constructor.name ]);
        }
    };
};
var tokensOfRow = function (dictTokensOf) {
    return {
        tokensOf: function (v) {
            return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(Data_Foldable.foldMap(Data_Foldable.foldableMaybe)(PureScript_CST_Range_TokenList.monoidTokenList)(tokensOf(tokensOfSeparated(tokensOfLabeled(tokensOfName)(tokensOfType(dictTokensOf)))))(v.labels))(Data_Foldable.foldMap(Data_Foldable.foldableMaybe)(PureScript_CST_Range_TokenList.monoidTokenList)(function (v1) {
                return PureScript_CST_Range_TokenList.cons(v1.value0)(tokensOf(tokensOfType(dictTokensOf))(v1.value1));
            })(v.tail));
        }
    };
};
var tokensOfBinder = function (dictTokensOf) {
    return {
        tokensOf: function (v) {
            if (v instanceof PureScript_CST_Types.BinderWildcard) {
                return PureScript_CST_Range_TokenList.singleton(v.value0);
            };
            if (v instanceof PureScript_CST_Types.BinderVar) {
                return tokensOf(tokensOfName)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.BinderNamed) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfName)(v.value0))(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return PureScript_CST_Range_TokenList.cons(v.value1)(tokensOf(tokensOfBinder(dictTokensOf))(v.value2));
                }));
            };
            if (v instanceof PureScript_CST_Types.BinderConstructor) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfQualifiedName)(v.value0))(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return tokensOf(tokensOfArray(tokensOfBinder(dictTokensOf)))(v.value1);
                }));
            };
            if (v instanceof PureScript_CST_Types.BinderBoolean) {
                return PureScript_CST_Range_TokenList.singleton(v.value0);
            };
            if (v instanceof PureScript_CST_Types.BinderChar) {
                return PureScript_CST_Range_TokenList.singleton(v.value0);
            };
            if (v instanceof PureScript_CST_Types.BinderString) {
                return PureScript_CST_Range_TokenList.singleton(v.value0);
            };
            if (v instanceof PureScript_CST_Types.BinderInt) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(Data_Foldable.foldMap(Data_Foldable.foldableMaybe)(PureScript_CST_Range_TokenList.monoidTokenList)(PureScript_CST_Range_TokenList.singleton)(v.value0))(PureScript_CST_Range_TokenList.singleton(v.value1));
            };
            if (v instanceof PureScript_CST_Types.BinderNumber) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(Data_Foldable.foldMap(Data_Foldable.foldableMaybe)(PureScript_CST_Range_TokenList.monoidTokenList)(PureScript_CST_Range_TokenList.singleton)(v.value0))(PureScript_CST_Range_TokenList.singleton(v.value1));
            };
            if (v instanceof PureScript_CST_Types.BinderArray) {
                return tokensOf(tokensOfWrapped(tokensOfMaybe(tokensOfSeparated(tokensOfBinder(dictTokensOf)))))(v.value0);
            };
            if (v instanceof PureScript_CST_Types.BinderRecord) {
                return tokensOf(tokensOfWrapped(tokensOfMaybe(tokensOfSeparated(tokensOfRecordLabeled(tokensOfBinder(dictTokensOf))))))(v.value0);
            };
            if (v instanceof PureScript_CST_Types.BinderParens) {
                return tokensOf(tokensOfWrapped(tokensOfBinder(dictTokensOf)))(v.value0);
            };
            if (v instanceof PureScript_CST_Types.BinderTyped) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfBinder(dictTokensOf))(v.value0))(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return PureScript_CST_Range_TokenList.cons(v.value1)(tokensOf(tokensOfType(dictTokensOf))(v.value2));
                }));
            };
            if (v instanceof PureScript_CST_Types.BinderOp) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfBinder(dictTokensOf))(v.value0))(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return tokensOf(tokensOfNonEmptyArray(tokensOfTuple(tokensOfQualifiedName)(tokensOfBinder(dictTokensOf))))(v.value1);
                }));
            };
            if (v instanceof PureScript_CST_Types.BinderError) {
                return tokensOf(dictTokensOf)(v.value0);
            };
            throw new Error("Failed pattern match at PureScript.CST.Range (line 953, column 14 - line 983, column 17): " + [ v.constructor.name ]);
        }
    };
};
var tokensOfDataCtor = function (dictTokensOf) {
    return {
        tokensOf: function (v) {
            return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfName)(v.name))(tokensOf(tokensOfArray(tokensOfType(dictTokensOf)))(v.fields));
        }
    };
};
var tokensOfForeign = function (dictTokensOf) {
    return {
        tokensOf: function (v) {
            if (v instanceof PureScript_CST_Types.ForeignValue) {
                return tokensOf(tokensOfLabeled(tokensOfName)(tokensOfType(dictTokensOf)))(v.value0);
            };
            if (v instanceof PureScript_CST_Types.ForeignData) {
                return PureScript_CST_Range_TokenList.cons(v.value0)(tokensOf(tokensOfLabeled(tokensOfName)(tokensOfType(dictTokensOf)))(v.value1));
            };
            if (v instanceof PureScript_CST_Types.ForeignKind) {
                return PureScript_CST_Range_TokenList.cons(v.value0)(tokensOf(tokensOfName)(v.value1));
            };
            throw new Error("Failed pattern match at PureScript.CST.Range (line 653, column 14 - line 659, column 26): " + [ v.constructor.name ]);
        }
    };
};
var tokensOfWhere = function (dictTokensOf) {
    return {
        tokensOf: function (v) {
            return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfExpr(dictTokensOf))(v.expr))(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                return Data_Foldable.foldMap(Data_Foldable.foldableMaybe)(PureScript_CST_Range_TokenList.monoidTokenList)(function (v2) {
                    return PureScript_CST_Range_TokenList.cons(v2.value0)(tokensOf(tokensOfNonEmptyArray(tokensOfLetBinding(dictTokensOf)))(v2.value1));
                })(v.bindings);
            }));
        }
    };
};
var tokensOfRecordUpdate = function (dictTokensOf) {
    return {
        tokensOf: function (v) {
            if (v instanceof PureScript_CST_Types.RecordUpdateLeaf) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfName)(v.value0))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(PureScript_CST_Range_TokenList.singleton(v.value1))(tokensOf(tokensOfExpr(dictTokensOf))(v.value2)));
            };
            if (v instanceof PureScript_CST_Types.RecordUpdateBranch) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfName)(v.value0))(tokensOf(tokensOfWrapped(tokensOfSeparated(tokensOfRecordUpdate(dictTokensOf))))(v.value1));
            };
            throw new Error("Failed pattern match at PureScript.CST.Range (line 837, column 14 - line 841, column 32): " + [ v.constructor.name ]);
        }
    };
};
var tokensOfPatternGuard = function (dictTokensOf) {
    return {
        tokensOf: function (v) {
            return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(Data_Foldable.foldMap(Data_Foldable.foldableMaybe)(PureScript_CST_Range_TokenList.monoidTokenList)(function (v1) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfBinder(dictTokensOf))(v1.value0))(PureScript_CST_Range_TokenList.singleton(v1.value1));
            })(v.binder))(tokensOf(tokensOfExpr(dictTokensOf))(v.expr));
        }
    };
};
var tokensOfLetBinding = function (dictTokensOf) {
    return {
        tokensOf: function (v) {
            if (v instanceof PureScript_CST_Types.LetBindingSignature) {
                return tokensOf(tokensOfLabeled(tokensOfName)(tokensOfType(dictTokensOf)))(v.value0);
            };
            if (v instanceof PureScript_CST_Types.LetBindingName) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfName)(v.value0.name))(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfArray(tokensOfBinder(dictTokensOf)))(v.value0.binders))(tokensOf(tokensOfGuarded(dictTokensOf))(v.value0.guarded));
                }));
            };
            if (v instanceof PureScript_CST_Types.LetBindingPattern) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfBinder(dictTokensOf))(v.value0))(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return PureScript_CST_Range_TokenList.cons(v.value1)(tokensOf(tokensOfWhere(dictTokensOf))(v.value2));
                }));
            };
            if (v instanceof PureScript_CST_Types.LetBindingError) {
                return tokensOf(dictTokensOf)(v.value0);
            };
            throw new Error("Failed pattern match at PureScript.CST.Range (line 885, column 14 - line 893, column 17): " + [ v.constructor.name ]);
        }
    };
};
var tokensOfGuardedExpr = function (dictTokensOf) {
    return {
        tokensOf: function (v) {
            return PureScript_CST_Range_TokenList.cons(v.bar)(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfSeparated(tokensOfPatternGuard(dictTokensOf)))(v.patterns))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(PureScript_CST_Range_TokenList.singleton(v.separator))(tokensOf(tokensOfWhere(dictTokensOf))(v.where)));
            }));
        }
    };
};
var tokensOfGuarded = function (dictTokensOf) {
    return {
        tokensOf: function (v) {
            if (v instanceof PureScript_CST_Types.Unconditional) {
                return PureScript_CST_Range_TokenList.cons(v.value0)(tokensOf(tokensOfWhere(dictTokensOf))(v.value1));
            };
            if (v instanceof PureScript_CST_Types.Guarded) {
                return tokensOf(tokensOfNonEmptyArray(tokensOfGuardedExpr(dictTokensOf)))(v.value0);
            };
            throw new Error("Failed pattern match at PureScript.CST.Range (line 597, column 14 - line 601, column 18): " + [ v.constructor.name ]);
        }
    };
};
var tokensOfExpr = function (dictTokensOf) {
    return {
        tokensOf: function (v) {
            if (v instanceof PureScript_CST_Types.ExprHole) {
                return tokensOf(tokensOfName)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.ExprSection) {
                return PureScript_CST_Range_TokenList.singleton(v.value0);
            };
            if (v instanceof PureScript_CST_Types.ExprIdent) {
                return tokensOf(tokensOfQualifiedName)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.ExprConstructor) {
                return tokensOf(tokensOfQualifiedName)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.ExprBoolean) {
                return PureScript_CST_Range_TokenList.singleton(v.value0);
            };
            if (v instanceof PureScript_CST_Types.ExprChar) {
                return PureScript_CST_Range_TokenList.singleton(v.value0);
            };
            if (v instanceof PureScript_CST_Types.ExprString) {
                return PureScript_CST_Range_TokenList.singleton(v.value0);
            };
            if (v instanceof PureScript_CST_Types.ExprInt) {
                return PureScript_CST_Range_TokenList.singleton(v.value0);
            };
            if (v instanceof PureScript_CST_Types.ExprNumber) {
                return PureScript_CST_Range_TokenList.singleton(v.value0);
            };
            if (v instanceof PureScript_CST_Types.ExprArray) {
                return tokensOf(tokensOfWrapped(tokensOfMaybe(tokensOfSeparated(tokensOfExpr(dictTokensOf)))))(v.value0);
            };
            if (v instanceof PureScript_CST_Types.ExprRecord) {
                return tokensOf(tokensOfWrapped(tokensOfMaybe(tokensOfSeparated(tokensOfRecordLabeled(tokensOfExpr(dictTokensOf))))))(v.value0);
            };
            if (v instanceof PureScript_CST_Types.ExprParens) {
                return tokensOf(tokensOfWrapped(tokensOfExpr(dictTokensOf)))(v.value0);
            };
            if (v instanceof PureScript_CST_Types.ExprTyped) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfExpr(dictTokensOf))(v.value0))(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return PureScript_CST_Range_TokenList.cons(v.value1)(tokensOf(tokensOfType(dictTokensOf))(v.value2));
                }));
            };
            if (v instanceof PureScript_CST_Types.ExprInfix) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfExpr(dictTokensOf))(v.value0))(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return tokensOf(tokensOfNonEmptyArray(tokensOfTuple(tokensOfWrapped(tokensOfExpr(dictTokensOf)))(tokensOfExpr(dictTokensOf))))(v.value1);
                }));
            };
            if (v instanceof PureScript_CST_Types.ExprOp) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfExpr(dictTokensOf))(v.value0))(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return tokensOf(tokensOfNonEmptyArray(tokensOfTuple(tokensOfQualifiedName)(tokensOfExpr(dictTokensOf))))(v.value1);
                }));
            };
            if (v instanceof PureScript_CST_Types.ExprOpName) {
                return tokensOf(tokensOfQualifiedName)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.ExprNegate) {
                return PureScript_CST_Range_TokenList.cons(v.value0)(tokensOf(tokensOfExpr(dictTokensOf))(v.value1));
            };
            if (v instanceof PureScript_CST_Types.ExprRecordAccessor) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfExpr(dictTokensOf))(v.value0.expr))(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return PureScript_CST_Range_TokenList.cons(v.value0.dot)(tokensOf(tokensOfSeparated(tokensOfName))(v.value0.path));
                }));
            };
            if (v instanceof PureScript_CST_Types.ExprRecordUpdate) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfExpr(dictTokensOf))(v.value0))(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return tokensOf(tokensOfWrapped(tokensOfSeparated(tokensOfRecordUpdate(dictTokensOf))))(v.value1);
                }));
            };
            if (v instanceof PureScript_CST_Types.ExprApp) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfExpr(dictTokensOf))(v.value0))(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return tokensOf(tokensOfNonEmptyArray(tokensOfExpr(dictTokensOf)))(v.value1);
                }));
            };
            if (v instanceof PureScript_CST_Types.ExprLambda) {
                return PureScript_CST_Range_TokenList.cons(v.value0.symbol)(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfNonEmptyArray(tokensOfBinder(dictTokensOf)))(v.value0.binders))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(PureScript_CST_Range_TokenList.singleton(v.value0.arrow))(tokensOf(tokensOfExpr(dictTokensOf))(v.value0.body)));
                }));
            };
            if (v instanceof PureScript_CST_Types.ExprIf) {
                return PureScript_CST_Range_TokenList.cons(v.value0.keyword)(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfExpr(dictTokensOf))(v.value0.cond))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(PureScript_CST_Range_TokenList.singleton(v.value0.then))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfExpr(dictTokensOf))(v["value0"]["true"]))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(PureScript_CST_Range_TokenList.singleton(v["value0"]["else"]))(tokensOf(tokensOfExpr(dictTokensOf))(v["value0"]["false"])))));
                }));
            };
            if (v instanceof PureScript_CST_Types.ExprCase) {
                return PureScript_CST_Range_TokenList.cons(v.value0.keyword)(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfSeparated(tokensOfExpr(dictTokensOf)))(v.value0.head))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(PureScript_CST_Range_TokenList.singleton(v.value0.of))(tokensOf(tokensOfNonEmptyArray(tokensOfTuple(tokensOfSeparated(tokensOfBinder(dictTokensOf)))(tokensOfGuarded(dictTokensOf))))(v.value0.branches)));
                }));
            };
            if (v instanceof PureScript_CST_Types.ExprLet) {
                return PureScript_CST_Range_TokenList.cons(v.value0.keyword)(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfNonEmptyArray(tokensOfLetBinding(dictTokensOf)))(v.value0.bindings))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(PureScript_CST_Range_TokenList.singleton(v["value0"]["in"]))(tokensOf(tokensOfExpr(dictTokensOf))(v.value0.body)));
                }));
            };
            if (v instanceof PureScript_CST_Types.ExprDo) {
                return PureScript_CST_Range_TokenList.cons(v.value0.keyword)(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return tokensOf(tokensOfNonEmptyArray(tokensOfDoStatement(dictTokensOf)))(v.value0.statements);
                }));
            };
            if (v instanceof PureScript_CST_Types.ExprAdo) {
                return PureScript_CST_Range_TokenList.cons(v.value0.keyword)(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfArray(tokensOfDoStatement(dictTokensOf)))(v.value0.statements))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(PureScript_CST_Range_TokenList.singleton(v["value0"]["in"]))(tokensOf(tokensOfExpr(dictTokensOf))(v.value0.result)));
                }));
            };
            if (v instanceof PureScript_CST_Types.ExprError) {
                return tokensOf(dictTokensOf)(v.value0);
            };
            throw new Error("Failed pattern match at PureScript.CST.Range (line 763, column 14 - line 834, column 17): " + [ v.constructor.name ]);
        }
    };
};
var tokensOfDoStatement = function (dictTokensOf) {
    return {
        tokensOf: function (v) {
            if (v instanceof PureScript_CST_Types.DoLet) {
                return PureScript_CST_Range_TokenList.cons(v.value0)(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return tokensOf(tokensOfNonEmptyArray(tokensOfLetBinding(dictTokensOf)))(v.value1);
                }));
            };
            if (v instanceof PureScript_CST_Types.DoDiscard) {
                return tokensOf(tokensOfExpr(dictTokensOf))(v.value0);
            };
            if (v instanceof PureScript_CST_Types.DoBind) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfBinder(dictTokensOf))(v.value0))(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return PureScript_CST_Range_TokenList.cons(v.value1)(tokensOf(tokensOfExpr(dictTokensOf))(v.value2));
                }));
            };
            if (v instanceof PureScript_CST_Types.DoError) {
                return tokensOf(dictTokensOf)(v.value0);
            };
            throw new Error("Failed pattern match at PureScript.CST.Range (line 859, column 14 - line 867, column 17): " + [ v.constructor.name ]);
        }
    };
};
var tokensOfInstanceBinding = function (dictTokensOf) {
    return {
        tokensOf: function (v) {
            if (v instanceof PureScript_CST_Types.InstanceBindingSignature) {
                return tokensOf(tokensOfLabeled(tokensOfName)(tokensOfType(dictTokensOf)))(v.value0);
            };
            if (v instanceof PureScript_CST_Types.InstanceBindingName) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfName)(v.value0.name))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfArray(tokensOfBinder(dictTokensOf)))(v.value0.binders))(tokensOf(tokensOfGuarded(dictTokensOf))(v.value0.guarded)));
            };
            throw new Error("Failed pattern match at PureScript.CST.Range (line 671, column 14 - line 677, column 28): " + [ v.constructor.name ]);
        }
    };
};
var tokensOfInstance = function (dictTokensOf) {
    return {
        tokensOf: function (v) {
            return PureScript_CST_Range_TokenList.cons(v.head.keyword)(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(Data_Foldable.foldMap(Data_Foldable.foldableMaybe)(PureScript_CST_Range_TokenList.monoidTokenList)(function (v2) {
                    return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfName)(v2.value0))(PureScript_CST_Range_TokenList.singleton(v2.value1));
                })(v.head.name))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(Data_Foldable.foldMap(Data_Foldable.foldableMaybe)(PureScript_CST_Range_TokenList.monoidTokenList)(function (v2) {
                    return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfOneOrDelimited(tokensOfType(dictTokensOf)))(v2.value0))(PureScript_CST_Range_TokenList.singleton(v2.value1));
                })(v.head.constraints))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfQualifiedName)(v.head.className))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfArray(tokensOfType(dictTokensOf)))(v.head.types))(Data_Foldable.foldMap(Data_Foldable.foldableMaybe)(PureScript_CST_Range_TokenList.monoidTokenList)(function (v2) {
                    return PureScript_CST_Range_TokenList.cons(v2.value0)(tokensOf(tokensOfNonEmptyArray(tokensOfInstanceBinding(dictTokensOf)))(v2.value1));
                })(v.body)))));
            }));
        }
    };
};
var tokensOfDecl = function (dictTokensOf) {
    return {
        tokensOf: function (v) {
            if (v instanceof PureScript_CST_Types.DeclData) {
                return PureScript_CST_Range_TokenList.cons(v.value0.keyword)(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfName)(v.value0.name))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfArray(tokensOfTypeVarBinding(dictTokensOf)))(v.value0.vars))(Data_Foldable.foldMap(Data_Foldable.foldableMaybe)(PureScript_CST_Range_TokenList.monoidTokenList)(function (v2) {
                        return PureScript_CST_Range_TokenList.cons(v2.value0)(tokensOf(tokensOfSeparated(tokensOfDataCtor(dictTokensOf)))(v2.value1));
                    })(v.value1)));
                }));
            };
            if (v instanceof PureScript_CST_Types.DeclType) {
                return PureScript_CST_Range_TokenList.cons(v.value0.keyword)(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfName)(v.value0.name))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfArray(tokensOfTypeVarBinding(dictTokensOf)))(v.value0.vars))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(PureScript_CST_Range_TokenList.singleton(v.value1))(tokensOf(tokensOfType(dictTokensOf))(v.value2))));
                }));
            };
            if (v instanceof PureScript_CST_Types.DeclNewtype) {
                return PureScript_CST_Range_TokenList.cons(v.value0.keyword)(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfName)(v.value0.name))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfArray(tokensOfTypeVarBinding(dictTokensOf)))(v.value0.vars))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(PureScript_CST_Range_TokenList.singleton(v.value1))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfName)(v.value2))(tokensOf(tokensOfType(dictTokensOf))(v.value3)))));
                }));
            };
            if (v instanceof PureScript_CST_Types.DeclClass) {
                return PureScript_CST_Range_TokenList.cons(v.value0.keyword)(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(Data_Foldable.foldMap(Data_Foldable.foldableMaybe)(PureScript_CST_Range_TokenList.monoidTokenList)(function (v2) {
                        return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfOneOrDelimited(tokensOfType(dictTokensOf)))(v2.value0))(PureScript_CST_Range_TokenList.singleton(v2.value1));
                    })(v["value0"]["super"]))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfName)(v.value0.name))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfArray(tokensOfTypeVarBinding(dictTokensOf)))(v.value0.vars))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(Data_Foldable.foldMap(Data_Foldable.foldableMaybe)(PureScript_CST_Range_TokenList.monoidTokenList)(function (v2) {
                        return PureScript_CST_Range_TokenList.cons(v2.value0)(tokensOf(tokensOfSeparated(tokensOfClassFundep))(v2.value1));
                    })(v.value0.fundeps))(Data_Foldable.foldMap(Data_Foldable.foldableMaybe)(PureScript_CST_Range_TokenList.monoidTokenList)(function (v2) {
                        return PureScript_CST_Range_TokenList.cons(v2.value0)(tokensOf(tokensOfNonEmptyArray(tokensOfLabeled(tokensOfName)(tokensOfType(dictTokensOf))))(v2.value1));
                    })(v.value1)))));
                }));
            };
            if (v instanceof PureScript_CST_Types.DeclInstanceChain) {
                return tokensOf(tokensOfSeparated(tokensOfInstance(dictTokensOf)))(v.value0);
            };
            if (v instanceof PureScript_CST_Types.DeclDerive) {
                return PureScript_CST_Range_TokenList.cons(v.value0)(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(Data_Foldable.foldMap(Data_Foldable.foldableMaybe)(PureScript_CST_Range_TokenList.monoidTokenList)(PureScript_CST_Range_TokenList.singleton)(v.value1))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(PureScript_CST_Range_TokenList.singleton(v.value2.keyword))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(Data_Foldable.foldMap(Data_Foldable.foldableMaybe)(PureScript_CST_Range_TokenList.monoidTokenList)(function (v2) {
                        return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfName)(v2.value0))(PureScript_CST_Range_TokenList.singleton(v2.value1));
                    })(v.value2.name))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(Data_Foldable.foldMap(Data_Foldable.foldableMaybe)(PureScript_CST_Range_TokenList.monoidTokenList)(function (v2) {
                        return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfOneOrDelimited(tokensOfType(dictTokensOf)))(v2.value0))(PureScript_CST_Range_TokenList.singleton(v2.value1));
                    })(v.value2.constraints))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfQualifiedName)(v.value2.className))(tokensOf(tokensOfArray(tokensOfType(dictTokensOf)))(v.value2.types))))));
                }));
            };
            if (v instanceof PureScript_CST_Types.DeclKindSignature) {
                return PureScript_CST_Range_TokenList.cons(v.value0)(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return tokensOf(tokensOfLabeled(tokensOfName)(tokensOfType(dictTokensOf)))(v.value1);
                }));
            };
            if (v instanceof PureScript_CST_Types.DeclSignature) {
                return tokensOf(tokensOfLabeled(tokensOfName)(tokensOfType(dictTokensOf)))(v.value0);
            };
            if (v instanceof PureScript_CST_Types.DeclValue) {
                return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfName)(v.value0.name))(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfArray(tokensOfBinder(dictTokensOf)))(v.value0.binders))(tokensOf(tokensOfGuarded(dictTokensOf))(v.value0.guarded));
                }));
            };
            if (v instanceof PureScript_CST_Types.DeclFixity) {
                return PureScript_CST_Range_TokenList.cons(v.value0.keyword.value0)(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return PureScript_CST_Range_TokenList.cons(v.value0.prec.value0)(tokensOf(tokensOfFixityOp)(v.value0.operator));
                }));
            };
            if (v instanceof PureScript_CST_Types.DeclForeign) {
                return PureScript_CST_Range_TokenList.cons(v.value0)(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return PureScript_CST_Range_TokenList.cons(v.value1)(tokensOf(tokensOfForeign(dictTokensOf))(v.value2));
                }));
            };
            if (v instanceof PureScript_CST_Types.DeclRole) {
                return PureScript_CST_Range_TokenList.cons(v.value0)(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                    return Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(PureScript_CST_Range_TokenList.singleton(v.value1))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfName)(v.value2))(Data_Foldable.foldMap(Data_Array_NonEmpty_Internal.foldableNonEmptyArray)(PureScript_CST_Range_TokenList.monoidTokenList)(function (v2) {
                        return PureScript_CST_Range_TokenList.singleton(v2.value0);
                    })(v.value3)));
                }));
            };
            if (v instanceof PureScript_CST_Types.DeclError) {
                return tokensOf(dictTokensOf)(v.value0);
            };
            throw new Error("Failed pattern match at PureScript.CST.Range (line 484, column 14 - line 540, column 17): " + [ v.constructor.name ]);
        }
    };
};
var tokensOfModule = function (dictTokensOf) {
    return {
        tokensOf: function (v) {
            return PureScript_CST_Range_TokenList.cons(v.header.keyword)(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(tokensOf(tokensOfName)(v.header.name))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                return Data_Foldable.foldMap(Data_Foldable.foldableMaybe)(PureScript_CST_Range_TokenList.monoidTokenList)(tokensOf(tokensOfWrapped(tokensOfSeparated(tokensOfExport(dictTokensOf)))))(v.header.exports);
            }))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(PureScript_CST_Range_TokenList.singleton(v.header.where))(Data_Semigroup.append(PureScript_CST_Range_TokenList.semigroupTokenList)(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                return Data_Foldable.foldMap(Data_Foldable.foldableArray)(PureScript_CST_Range_TokenList.monoidTokenList)(tokensOf(tokensOfImportDecl(dictTokensOf)))(v.header.imports);
            }))(Control_Lazy.defer(PureScript_CST_Range_TokenList.lazyTokenList)(function (v1) {
                return Data_Foldable.foldMap(Data_Foldable.foldableArray)(PureScript_CST_Range_TokenList.monoidTokenList)(tokensOf(tokensOfDecl(dictTokensOf)))(v.body.decls);
            }))))));
        }
    };
};
var rangeOfWrapped = {
    rangeOf: function (v) {
        return {
            start: v.open.range.start,
            end: v.close.range.end
        };
    }
};
var rangeOfVoid = {
    rangeOf: Data_Void.absurd
};
var rangeOfRecoveredError = {
    rangeOf: function (v) {
        var v1 = Data_Array_NonEmpty.fromArray(v.tokens);
        if (v1 instanceof Data_Maybe.Just) {
            return {
                start: (Data_Array_NonEmpty.head(v1.value0)).range.start,
                end: (Data_Array_NonEmpty.last(v1.value0)).range.end
            };
        };
        if (v1 instanceof Data_Maybe.Nothing) {
            return {
                start: v.position,
                end: v.position
            };
        };
        throw new Error("Failed pattern match at PureScript.CST.Range (line 49, column 5 - line 57, column 10): " + [ v1.constructor.name ]);
    }
};
var rangeOfQualifiedName = {
    rangeOf: function (v) {
        return v.token.range;
    }
};
var rangeOfName = {
    rangeOf: function (v) {
        return v.token.range;
    }
};
var rangeOfModule = {
    rangeOf: function (v) {
        return {
            start: v.header.keyword.range.start,
            end: v.body.end
        };
    }
};
var rangeOf = function (dict) {
    return dict.rangeOf;
};
var rangeOfClassFundep = {
    rangeOf: function (v) {
        if (v instanceof PureScript_CST_Types.FundepDetermined) {
            return {
                start: v.value0.range.start,
                end: (rangeOf(rangeOfName)(Data_Array_NonEmpty.last(v.value1))).end
            };
        };
        if (v instanceof PureScript_CST_Types.FundepDetermines) {
            return {
                start: (rangeOf(rangeOfName)(Data_Array_NonEmpty.head(v.value0))).start,
                end: (rangeOf(rangeOfName)(Data_Array_NonEmpty.last(v.value2))).end
            };
        };
        throw new Error("Failed pattern match at PureScript.CST.Range (line 543, column 13 - line 551, column 8): " + [ v.constructor.name ]);
    }
};
var rangeOfDataMembers = {
    rangeOf: function (v) {
        if (v instanceof PureScript_CST_Types.DataAll) {
            return v.value0.range;
        };
        if (v instanceof PureScript_CST_Types.DataEnumerated) {
            return rangeOf(rangeOfWrapped)(v.value0);
        };
        throw new Error("Failed pattern match at PureScript.CST.Range (line 306, column 13 - line 310, column 16): " + [ v.constructor.name ]);
    }
};
var rangeOfExport = function (dictRangeOf) {
    return {
        rangeOf: function (v) {
            if (v instanceof PureScript_CST_Types.ExportValue) {
                return rangeOf(rangeOfName)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.ExportOp) {
                return rangeOf(rangeOfName)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.ExportType) {
                if (v.value1 instanceof Data_Maybe.Nothing) {
                    return rangeOf(rangeOfName)(v.value0);
                };
                if (v.value1 instanceof Data_Maybe.Just) {
                    return {
                        start: (rangeOf(rangeOfName)(v.value0)).start,
                        end: (rangeOf(rangeOfDataMembers)(v.value1.value0)).end
                    };
                };
                throw new Error("Failed pattern match at PureScript.CST.Range (line 260, column 7 - line 266, column 12): " + [ v.value1.constructor.name ]);
            };
            if (v instanceof PureScript_CST_Types.ExportTypeOp) {
                return {
                    start: v.value0.range.start,
                    end: (rangeOf(rangeOfName)(v.value1)).end
                };
            };
            if (v instanceof PureScript_CST_Types.ExportClass) {
                return {
                    start: v.value0.range.start,
                    end: (rangeOf(rangeOfName)(v.value1)).end
                };
            };
            if (v instanceof PureScript_CST_Types.ExportKind) {
                return {
                    start: v.value0.range.start,
                    end: (rangeOf(rangeOfName)(v.value1)).end
                };
            };
            if (v instanceof PureScript_CST_Types.ExportModule) {
                return {
                    start: v.value0.range.start,
                    end: (rangeOf(rangeOfName)(v.value1)).end
                };
            };
            if (v instanceof PureScript_CST_Types.ExportError) {
                return rangeOf(dictRangeOf)(v.value0);
            };
            throw new Error("Failed pattern match at PureScript.CST.Range (line 254, column 13 - line 284, column 16): " + [ v.constructor.name ]);
        }
    };
};
var rangeOfFixityOp = {
    rangeOf: function (v) {
        if (v instanceof PureScript_CST_Types.FixityValue) {
            return {
                start: (rangeOf(rangeOfQualifiedName)(v.value0)).start,
                end: (rangeOf(rangeOfName)(v.value2)).end
            };
        };
        if (v instanceof PureScript_CST_Types.FixityType) {
            return {
                start: v.value0.range.start,
                end: (rangeOf(rangeOfName)(v.value3)).end
            };
        };
        throw new Error("Failed pattern match at PureScript.CST.Range (line 622, column 13 - line 630, column 8): " + [ v.constructor.name ]);
    }
};
var rangeOfImport = function (dictRangeOf) {
    return {
        rangeOf: function (v) {
            if (v instanceof PureScript_CST_Types.ImportValue) {
                return rangeOf(rangeOfName)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.ImportOp) {
                return rangeOf(rangeOfName)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.ImportType) {
                if (v.value1 instanceof Data_Maybe.Nothing) {
                    return rangeOf(rangeOfName)(v.value0);
                };
                if (v.value1 instanceof Data_Maybe.Just) {
                    return {
                        start: (rangeOf(rangeOfName)(v.value0)).start,
                        end: (rangeOf(rangeOfDataMembers)(v.value1.value0)).end
                    };
                };
                throw new Error("Failed pattern match at PureScript.CST.Range (line 349, column 7 - line 355, column 12): " + [ v.value1.constructor.name ]);
            };
            if (v instanceof PureScript_CST_Types.ImportTypeOp) {
                return {
                    start: v.value0.range.start,
                    end: (rangeOf(rangeOfName)(v.value1)).end
                };
            };
            if (v instanceof PureScript_CST_Types.ImportClass) {
                return {
                    start: v.value0.range.start,
                    end: (rangeOf(rangeOfName)(v.value1)).end
                };
            };
            if (v instanceof PureScript_CST_Types.ImportKind) {
                return {
                    start: v.value0.range.start,
                    end: (rangeOf(rangeOfName)(v.value1)).end
                };
            };
            if (v instanceof PureScript_CST_Types.ImportError) {
                return rangeOf(dictRangeOf)(v.value0);
            };
            throw new Error("Failed pattern match at PureScript.CST.Range (line 343, column 13 - line 369, column 16): " + [ v.constructor.name ]);
        }
    };
};
var rangeOfImportDecl = {
    rangeOf: function (v) {
        var v1 = (function () {
            if (v.qualified instanceof Data_Maybe.Nothing) {
                if (v.names instanceof Data_Maybe.Nothing) {
                    return rangeOf(rangeOfName)(v.module);
                };
                if (v.names instanceof Data_Maybe.Just) {
                    return rangeOf(rangeOfWrapped)(v.names.value0.value1);
                };
                throw new Error("Failed pattern match at PureScript.CST.Range (line 324, column 11 - line 328, column 30): " + [ v.names.constructor.name ]);
            };
            if (v.qualified instanceof Data_Maybe.Just) {
                return rangeOf(rangeOfName)(v.qualified.value0.value1);
            };
            throw new Error("Failed pattern match at PureScript.CST.Range (line 322, column 17 - line 330, column 20): " + [ v.qualified.constructor.name ]);
        })();
        return {
            start: v.keyword.range.start,
            end: v1.end
        };
    }
};
var rangeOfLabeled = function (dictRangeOf) {
    return function (dictRangeOf1) {
        return {
            rangeOf: function (v) {
                return {
                    start: (rangeOf(dictRangeOf)(v.label)).start,
                    end: (rangeOf(dictRangeOf1)(v.value)).end
                };
            }
        };
    };
};
var rangeOfOneOrDelimited = function (dictRangeOf) {
    return {
        rangeOf: function (v) {
            if (v instanceof PureScript_CST_Types.One) {
                return rangeOf(dictRangeOf)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.Many) {
                return rangeOf(rangeOfWrapped)(v.value0);
            };
            throw new Error("Failed pattern match at PureScript.CST.Range (line 125, column 13 - line 127, column 26): " + [ v.constructor.name ]);
        }
    };
};
var rangeOfSeparated = function (dictRangeOf) {
    return {
        rangeOf: function (v) {
            var v1 = Data_Array.last(v.tail);
            if (v1 instanceof Data_Maybe.Just) {
                return {
                    start: (rangeOf(dictRangeOf)(v.head)).start,
                    end: (rangeOf(dictRangeOf)(v1.value0.value1)).end
                };
            };
            if (v1 instanceof Data_Maybe.Nothing) {
                return rangeOf(dictRangeOf)(v.head);
            };
            throw new Error("Failed pattern match at PureScript.CST.Range (line 101, column 5 - line 107, column 21): " + [ v1.constructor.name ]);
        }
    };
};
var rangeOfType = function (dictRangeOf) {
    return {
        rangeOf: function (v) {
            if (v instanceof PureScript_CST_Types.TypeVar) {
                return rangeOf(rangeOfName)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.TypeConstructor) {
                return rangeOf(rangeOfQualifiedName)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.TypeWildcard) {
                return v.value0.range;
            };
            if (v instanceof PureScript_CST_Types.TypeHole) {
                return rangeOf(rangeOfName)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.TypeString) {
                return v.value0.range;
            };
            if (v instanceof PureScript_CST_Types.TypeRow) {
                return rangeOf(rangeOfWrapped)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.TypeRecord) {
                return rangeOf(rangeOfWrapped)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.TypeForall) {
                return {
                    start: v.value0.range.start,
                    end: (rangeOf(rangeOfType(dictRangeOf))(v.value3)).end
                };
            };
            if (v instanceof PureScript_CST_Types.TypeKinded) {
                return {
                    start: (rangeOf(rangeOfType(dictRangeOf))(v.value0)).start,
                    end: (rangeOf(rangeOfType(dictRangeOf))(v.value2)).end
                };
            };
            if (v instanceof PureScript_CST_Types.TypeApp) {
                return {
                    start: (rangeOf(rangeOfType(dictRangeOf))(v.value0)).start,
                    end: (rangeOf(rangeOfType(dictRangeOf))(Data_Array_NonEmpty.last(v.value1))).end
                };
            };
            if (v instanceof PureScript_CST_Types.TypeOp) {
                return {
                    start: (rangeOf(rangeOfType(dictRangeOf))(v.value0)).start,
                    end: (rangeOf(rangeOfType(dictRangeOf))(Data_Tuple.snd(Data_Array_NonEmpty.last(v.value1)))).end
                };
            };
            if (v instanceof PureScript_CST_Types.TypeOpName) {
                return rangeOf(rangeOfQualifiedName)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.TypeArrow) {
                return {
                    start: (rangeOf(rangeOfType(dictRangeOf))(v.value0)).start,
                    end: (rangeOf(rangeOfType(dictRangeOf))(v.value2)).end
                };
            };
            if (v instanceof PureScript_CST_Types.TypeArrowName) {
                return v.value0.range;
            };
            if (v instanceof PureScript_CST_Types.TypeConstrained) {
                return {
                    start: (rangeOf(rangeOfType(dictRangeOf))(v.value0)).start,
                    end: (rangeOf(rangeOfType(dictRangeOf))(v.value2)).end
                };
            };
            if (v instanceof PureScript_CST_Types.TypeParens) {
                return rangeOf(rangeOfWrapped)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.TypeUnaryRow) {
                return {
                    start: v.value0.range.start,
                    end: (rangeOf(rangeOfType(dictRangeOf))(v.value1)).end
                };
            };
            if (v instanceof PureScript_CST_Types["TypeError"]) {
                return rangeOf(dictRangeOf)(v.value0);
            };
            throw new Error("Failed pattern match at PureScript.CST.Range (line 135, column 13 - line 185, column 16): " + [ v.constructor.name ]);
        }
    };
};
var rangeOfBinder = function (dictRangeOf) {
    return {
        rangeOf: function (v) {
            if (v instanceof PureScript_CST_Types.BinderWildcard) {
                return v.value0.range;
            };
            if (v instanceof PureScript_CST_Types.BinderVar) {
                return rangeOf(rangeOfName)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.BinderNamed) {
                return {
                    start: (rangeOf(rangeOfName)(v.value0)).start,
                    end: (rangeOf(rangeOfBinder(dictRangeOf))(v.value2)).end
                };
            };
            if (v instanceof PureScript_CST_Types.BinderConstructor) {
                var v1 = Data_Array.last(v.value1);
                if (v1 instanceof Data_Maybe.Nothing) {
                    return rangeOf(rangeOfQualifiedName)(v.value0);
                };
                if (v1 instanceof Data_Maybe.Just) {
                    return {
                        start: (rangeOf(rangeOfQualifiedName)(v.value0)).start,
                        end: (rangeOf(rangeOfBinder(dictRangeOf))(v1.value0)).end
                    };
                };
                throw new Error("Failed pattern match at PureScript.CST.Range (line 906, column 7 - line 912, column 12): " + [ v1.constructor.name ]);
            };
            if (v instanceof PureScript_CST_Types.BinderBoolean) {
                return v.value0.range;
            };
            if (v instanceof PureScript_CST_Types.BinderChar) {
                return v.value0.range;
            };
            if (v instanceof PureScript_CST_Types.BinderString) {
                return v.value0.range;
            };
            if (v instanceof PureScript_CST_Types.BinderInt) {
                if (v.value0 instanceof Data_Maybe.Nothing) {
                    return v.value1.range;
                };
                if (v.value0 instanceof Data_Maybe.Just) {
                    return {
                        start: v.value0.value0.range.start,
                        end: v.value1.range.end
                    };
                };
                throw new Error("Failed pattern match at PureScript.CST.Range (line 920, column 7 - line 926, column 12): " + [ v.value0.constructor.name ]);
            };
            if (v instanceof PureScript_CST_Types.BinderNumber) {
                if (v.value0 instanceof Data_Maybe.Nothing) {
                    return v.value1.range;
                };
                if (v.value0 instanceof Data_Maybe.Just) {
                    return {
                        start: v.value0.value0.range.start,
                        end: v.value1.range.end
                    };
                };
                throw new Error("Failed pattern match at PureScript.CST.Range (line 928, column 7 - line 934, column 12): " + [ v.value0.constructor.name ]);
            };
            if (v instanceof PureScript_CST_Types.BinderArray) {
                return rangeOf(rangeOfWrapped)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.BinderRecord) {
                return rangeOf(rangeOfWrapped)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.BinderParens) {
                return rangeOf(rangeOfWrapped)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.BinderTyped) {
                return {
                    start: (rangeOf(rangeOfBinder(dictRangeOf))(v.value0)).start,
                    end: (rangeOf(rangeOfType(dictRangeOf))(v.value2)).end
                };
            };
            if (v instanceof PureScript_CST_Types.BinderOp) {
                return {
                    start: (rangeOf(rangeOfBinder(dictRangeOf))(v.value0)).start,
                    end: (rangeOf(rangeOfBinder(dictRangeOf))(Data_Tuple.snd(Data_Array_NonEmpty.last(v.value1)))).end
                };
            };
            if (v instanceof PureScript_CST_Types.BinderError) {
                return rangeOf(dictRangeOf)(v.value0);
            };
            throw new Error("Failed pattern match at PureScript.CST.Range (line 896, column 13 - line 950, column 16): " + [ v.constructor.name ]);
        }
    };
};
var rangeOfDataCtor = function (dictRangeOf) {
    return {
        rangeOf: function (v) {
            var v1 = (function () {
                var v2 = Data_Array.last(v.fields);
                if (v2 instanceof Data_Maybe.Nothing) {
                    return rangeOf(rangeOfName)(v.name);
                };
                if (v2 instanceof Data_Maybe.Just) {
                    return rangeOf(rangeOfType(dictRangeOf))(v2.value0);
                };
                throw new Error("Failed pattern match at PureScript.CST.Range (line 391, column 17 - line 395, column 21): " + [ v2.constructor.name ]);
            })();
            return {
                start: (rangeOf(rangeOfName)(v.name)).start,
                end: v1.end
            };
        }
    };
};
var rangeOfForeign = function (dictRangeOf) {
    return {
        rangeOf: function (v) {
            if (v instanceof PureScript_CST_Types.ForeignValue) {
                return rangeOf(rangeOfLabeled(rangeOfName)(rangeOfType(dictRangeOf)))(v.value0);
            };
            if (v instanceof PureScript_CST_Types.ForeignData) {
                return {
                    start: v.value0.range.start,
                    end: (rangeOf(rangeOfLabeled(rangeOfName)(rangeOfType(dictRangeOf)))(v.value1)).end
                };
            };
            if (v instanceof PureScript_CST_Types.ForeignKind) {
                return {
                    start: v.value0.range.start,
                    end: (rangeOf(rangeOfName)(v.value1)).end
                };
            };
            throw new Error("Failed pattern match at PureScript.CST.Range (line 640, column 13 - line 650, column 8): " + [ v.constructor.name ]);
        }
    };
};
var rangeOfTypeVarBinding = {
    rangeOf: function (v) {
        if (v instanceof PureScript_CST_Types.TypeVarKinded) {
            return rangeOf(rangeOfWrapped)(v.value0);
        };
        if (v instanceof PureScript_CST_Types.TypeVarName) {
            return rangeOf(rangeOfName)(v.value0);
        };
        throw new Error("Failed pattern match at PureScript.CST.Range (line 240, column 13 - line 244, column 16): " + [ v.constructor.name ]);
    }
};
var rangeOfWhere = function (dictRangeOf) {
    return {
        rangeOf: function (v) {
            if (v.bindings instanceof Data_Maybe.Nothing) {
                return rangeOf(rangeOfExpr(dictRangeOf))(v.expr);
            };
            if (v.bindings instanceof Data_Maybe.Just) {
                return {
                    start: (rangeOf(rangeOfExpr(dictRangeOf))(v.expr)).start,
                    end: (rangeOf(rangeOfLetBinding(dictRangeOf))(Data_Array_NonEmpty.last(v.bindings.value0.value1))).end
                };
            };
            throw new Error("Failed pattern match at PureScript.CST.Range (line 993, column 40 - line 999, column 8): " + [ v.bindings.constructor.name ]);
        }
    };
};
var rangeOfLetBinding = function (dictRangeOf) {
    return {
        rangeOf: function (v) {
            if (v instanceof PureScript_CST_Types.LetBindingSignature) {
                return rangeOf(rangeOfLabeled(rangeOfName)(rangeOfType(dictRangeOf)))(v.value0);
            };
            if (v instanceof PureScript_CST_Types.LetBindingName) {
                return {
                    start: (rangeOf(rangeOfName)(v.value0.name)).start,
                    end: (rangeOf(rangeOfGuarded(dictRangeOf))(v.value0.guarded)).end
                };
            };
            if (v instanceof PureScript_CST_Types.LetBindingPattern) {
                return {
                    start: (rangeOf(rangeOfBinder(dictRangeOf))(v.value0)).start,
                    end: (rangeOf(rangeOfWhere(dictRangeOf))(v.value2)).end
                };
            };
            if (v instanceof PureScript_CST_Types.LetBindingError) {
                return rangeOf(dictRangeOf)(v.value0);
            };
            throw new Error("Failed pattern match at PureScript.CST.Range (line 870, column 13 - line 882, column 16): " + [ v.constructor.name ]);
        }
    };
};
var rangeOfGuardedExpr = function (dictRangeOf) {
    return {
        rangeOf: function (v) {
            return {
                start: v.bar.range.start,
                end: (rangeOf(rangeOfWhere(dictRangeOf))(v.where)).end
            };
        }
    };
};
var rangeOfGuarded = function (dictRangeOf) {
    return {
        rangeOf: function (v) {
            if (v instanceof PureScript_CST_Types.Unconditional) {
                return {
                    start: v.value0.range.start,
                    end: (rangeOf(rangeOfWhere(dictRangeOf))(v.value1)).end
                };
            };
            if (v instanceof PureScript_CST_Types.Guarded) {
                return {
                    start: (rangeOf(rangeOfGuardedExpr(dictRangeOf))(Data_Array_NonEmpty.head(v.value0))).start,
                    end: (rangeOf(rangeOfGuardedExpr(dictRangeOf))(Data_Array_NonEmpty.last(v.value0))).end
                };
            };
            throw new Error("Failed pattern match at PureScript.CST.Range (line 586, column 13 - line 594, column 8): " + [ v.constructor.name ]);
        }
    };
};
var rangeOfExpr = function (dictRangeOf) {
    return {
        rangeOf: function (v) {
            if (v instanceof PureScript_CST_Types.ExprHole) {
                return rangeOf(rangeOfName)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.ExprSection) {
                return v.value0.range;
            };
            if (v instanceof PureScript_CST_Types.ExprIdent) {
                return rangeOf(rangeOfQualifiedName)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.ExprConstructor) {
                return rangeOf(rangeOfQualifiedName)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.ExprBoolean) {
                return v.value0.range;
            };
            if (v instanceof PureScript_CST_Types.ExprChar) {
                return v.value0.range;
            };
            if (v instanceof PureScript_CST_Types.ExprString) {
                return v.value0.range;
            };
            if (v instanceof PureScript_CST_Types.ExprInt) {
                return v.value0.range;
            };
            if (v instanceof PureScript_CST_Types.ExprNumber) {
                return v.value0.range;
            };
            if (v instanceof PureScript_CST_Types.ExprArray) {
                return rangeOf(rangeOfWrapped)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.ExprRecord) {
                return rangeOf(rangeOfWrapped)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.ExprParens) {
                return rangeOf(rangeOfWrapped)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.ExprTyped) {
                return {
                    start: (rangeOf(rangeOfExpr(dictRangeOf))(v.value0)).start,
                    end: (rangeOf(rangeOfType(dictRangeOf))(v.value2)).end
                };
            };
            if (v instanceof PureScript_CST_Types.ExprInfix) {
                return {
                    start: (rangeOf(rangeOfExpr(dictRangeOf))(v.value0)).start,
                    end: (rangeOf(rangeOfExpr(dictRangeOf))(Data_Tuple.snd(Data_Array_NonEmpty.last(v.value1)))).end
                };
            };
            if (v instanceof PureScript_CST_Types.ExprOp) {
                return {
                    start: (rangeOf(rangeOfExpr(dictRangeOf))(v.value0)).start,
                    end: (rangeOf(rangeOfExpr(dictRangeOf))(Data_Tuple.snd(Data_Array_NonEmpty.last(v.value1)))).end
                };
            };
            if (v instanceof PureScript_CST_Types.ExprOpName) {
                return rangeOf(rangeOfQualifiedName)(v.value0);
            };
            if (v instanceof PureScript_CST_Types.ExprNegate) {
                return {
                    start: v.value0.range.start,
                    end: (rangeOf(rangeOfExpr(dictRangeOf))(v.value1)).end
                };
            };
            if (v instanceof PureScript_CST_Types.ExprRecordAccessor) {
                return {
                    start: (rangeOf(rangeOfExpr(dictRangeOf))(v.value0.expr)).start,
                    end: (rangeOf(rangeOfSeparated(rangeOfName))(v.value0.path)).end
                };
            };
            if (v instanceof PureScript_CST_Types.ExprRecordUpdate) {
                return {
                    start: (rangeOf(rangeOfExpr(dictRangeOf))(v.value0)).start,
                    end: (rangeOf(rangeOfWrapped)(v.value1)).end
                };
            };
            if (v instanceof PureScript_CST_Types.ExprApp) {
                return {
                    start: (rangeOf(rangeOfExpr(dictRangeOf))(v.value0)).start,
                    end: (rangeOf(rangeOfExpr(dictRangeOf))(Data_Array_NonEmpty.last(v.value1))).end
                };
            };
            if (v instanceof PureScript_CST_Types.ExprLambda) {
                return {
                    start: v.value0.symbol.range.start,
                    end: (rangeOf(rangeOfExpr(dictRangeOf))(v.value0.body)).end
                };
            };
            if (v instanceof PureScript_CST_Types.ExprIf) {
                return {
                    start: v.value0.keyword.range.start,
                    end: (rangeOf(rangeOfExpr(dictRangeOf))(v["value0"]["false"])).end
                };
            };
            if (v instanceof PureScript_CST_Types.ExprCase) {
                return {
                    start: v.value0.keyword.range.start,
                    end: (rangeOf(rangeOfGuarded(dictRangeOf))(Data_Tuple.snd(Data_Array_NonEmpty.last(v.value0.branches)))).end
                };
            };
            if (v instanceof PureScript_CST_Types.ExprLet) {
                return {
                    start: v.value0.keyword.range.start,
                    end: (rangeOf(rangeOfExpr(dictRangeOf))(v.value0.body)).end
                };
            };
            if (v instanceof PureScript_CST_Types.ExprDo) {
                return {
                    start: v.value0.keyword.range.start,
                    end: (rangeOf(rangeOfDoStatement(dictRangeOf))(Data_Array_NonEmpty.last(v.value0.statements))).end
                };
            };
            if (v instanceof PureScript_CST_Types.ExprAdo) {
                return {
                    start: v.value0.keyword.range.start,
                    end: (rangeOf(rangeOfExpr(dictRangeOf))(v.value0.result)).end
                };
            };
            if (v instanceof PureScript_CST_Types.ExprError) {
                return rangeOf(dictRangeOf)(v.value0);
            };
            throw new Error("Failed pattern match at PureScript.CST.Range (line 680, column 13 - line 760, column 16): " + [ v.constructor.name ]);
        }
    };
};
var rangeOfDoStatement = function (dictRangeOf) {
    return {
        rangeOf: function (v) {
            if (v instanceof PureScript_CST_Types.DoLet) {
                return {
                    start: v.value0.range.start,
                    end: (rangeOf(rangeOfLetBinding(dictRangeOf))(Data_Array_NonEmpty.last(v.value1))).end
                };
            };
            if (v instanceof PureScript_CST_Types.DoDiscard) {
                return rangeOf(rangeOfExpr(dictRangeOf))(v.value0);
            };
            if (v instanceof PureScript_CST_Types.DoBind) {
                return {
                    start: (rangeOf(rangeOfBinder(dictRangeOf))(v.value0)).start,
                    end: (rangeOf(rangeOfExpr(dictRangeOf))(v.value2)).end
                };
            };
            if (v instanceof PureScript_CST_Types.DoError) {
                return rangeOf(dictRangeOf)(v.value0);
            };
            throw new Error("Failed pattern match at PureScript.CST.Range (line 844, column 13 - line 856, column 16): " + [ v.constructor.name ]);
        }
    };
};
var rangeOfInstanceBinding = function (dictRangeOf) {
    return {
        rangeOf: function (v) {
            if (v instanceof PureScript_CST_Types.InstanceBindingSignature) {
                return rangeOf(rangeOfLabeled(rangeOfName)(rangeOfType(dictRangeOf)))(v.value0);
            };
            if (v instanceof PureScript_CST_Types.InstanceBindingName) {
                return {
                    start: (rangeOf(rangeOfName)(v.value0.name)).start,
                    end: (rangeOf(rangeOfGuarded(dictRangeOf))(v.value0.guarded)).end
                };
            };
            throw new Error("Failed pattern match at PureScript.CST.Range (line 662, column 13 - line 668, column 8): " + [ v.constructor.name ]);
        }
    };
};
var rangeOfInstance = function (dictRangeOf) {
    return {
        rangeOf: function (v) {
            var v1 = (function () {
                if (v.body instanceof Data_Maybe.Nothing) {
                    var v2 = Data_Array.last(v.head.types);
                    if (v2 instanceof Data_Maybe.Nothing) {
                        return rangeOf(rangeOfQualifiedName)(v.head.className);
                    };
                    if (v2 instanceof Data_Maybe.Just) {
                        return rangeOf(rangeOfType(dictRangeOf))(v2.value0);
                    };
                    throw new Error("Failed pattern match at PureScript.CST.Range (line 565, column 11 - line 569, column 25): " + [ v2.constructor.name ]);
                };
                if (v.body instanceof Data_Maybe.Just) {
                    return rangeOf(rangeOfInstanceBinding(dictRangeOf))(Data_Array_NonEmpty.last(v.body.value0.value1));
                };
                throw new Error("Failed pattern match at PureScript.CST.Range (line 563, column 17 - line 571, column 42): " + [ v.body.constructor.name ]);
            })();
            return {
                start: v.head.keyword.range.start,
                end: v1.end
            };
        }
    };
};
var rangeOfDecl = function (dictRangeOf) {
    return {
        rangeOf: function (v) {
            if (v instanceof PureScript_CST_Types.DeclData) {
                var v1 = (function () {
                    if (v.value1 instanceof Data_Maybe.Nothing) {
                        var v2 = Data_Array.last(v.value0.vars);
                        if (v2 instanceof Data_Maybe.Nothing) {
                            return rangeOf(rangeOfName)(v.value0.name);
                        };
                        if (v2 instanceof Data_Maybe.Just) {
                            return rangeOf(rangeOfTypeVarBinding)(v2.value0);
                        };
                        throw new Error("Failed pattern match at PureScript.CST.Range (line 410, column 13 - line 414, column 28): " + [ v2.constructor.name ]);
                    };
                    if (v.value1 instanceof Data_Maybe.Just) {
                        return rangeOf(rangeOfDataCtor(dictRangeOf))(Data_Maybe.maybe(v.value1.value0.value1.head)(Data_Tuple.snd)(Data_Array.last(v.value1.value0.value1.tail)));
                    };
                    throw new Error("Failed pattern match at PureScript.CST.Range (line 408, column 19 - line 416, column 55): " + [ v.value1.constructor.name ]);
                })();
                return {
                    start: v.value0.keyword.range.start,
                    end: v1.end
                };
            };
            if (v instanceof PureScript_CST_Types.DeclType) {
                return {
                    start: v.value0.keyword.range.start,
                    end: (rangeOf(rangeOfType(dictRangeOf))(v.value2)).end
                };
            };
            if (v instanceof PureScript_CST_Types.DeclNewtype) {
                return {
                    start: v.value0.keyword.range.start,
                    end: (rangeOf(rangeOfType(dictRangeOf))(v.value3)).end
                };
            };
            if (v instanceof PureScript_CST_Types.DeclClass) {
                var v1 = (function () {
                    if (v.value1 instanceof Data_Maybe.Nothing) {
                        if (v.value0.fundeps instanceof Data_Maybe.Nothing) {
                            var v2 = Data_Array.last(v.value0.vars);
                            if (v2 instanceof Data_Maybe.Nothing) {
                                return rangeOf(rangeOfName)(v.value0.name);
                            };
                            if (v2 instanceof Data_Maybe.Just) {
                                return rangeOf(rangeOfTypeVarBinding)(v2.value0);
                            };
                            throw new Error("Failed pattern match at PureScript.CST.Range (line 434, column 17 - line 438, column 32): " + [ v2.constructor.name ]);
                        };
                        if (v.value0.fundeps instanceof Data_Maybe.Just) {
                            return rangeOf(rangeOfSeparated(rangeOfClassFundep))(v.value0.fundeps.value0.value1);
                        };
                        throw new Error("Failed pattern match at PureScript.CST.Range (line 432, column 13 - line 440, column 32): " + [ v.value0.fundeps.constructor.name ]);
                    };
                    if (v.value1 instanceof Data_Maybe.Just) {
                        return rangeOf(rangeOfLabeled(rangeOfName)(rangeOfType(dictRangeOf)))(Data_Array_NonEmpty.last(v.value1.value0.value1));
                    };
                    throw new Error("Failed pattern match at PureScript.CST.Range (line 430, column 19 - line 442, column 44): " + [ v.value1.constructor.name ]);
                })();
                return {
                    start: v.value0.keyword.range.start,
                    end: v1.end
                };
            };
            if (v instanceof PureScript_CST_Types.DeclInstanceChain) {
                return rangeOf(rangeOfSeparated(rangeOfInstance(dictRangeOf)))(v.value0);
            };
            if (v instanceof PureScript_CST_Types.DeclDerive) {
                var v1 = (function () {
                    var v2 = Data_Array.last(v.value2.types);
                    if (v2 instanceof Data_Maybe.Nothing) {
                        return rangeOf(rangeOfQualifiedName)(v.value2.className);
                    };
                    if (v2 instanceof Data_Maybe.Just) {
                        return rangeOf(rangeOfType(dictRangeOf))(v2.value0);
                    };
                    throw new Error("Failed pattern match at PureScript.CST.Range (line 450, column 19 - line 454, column 23): " + [ v2.constructor.name ]);
                })();
                return {
                    start: v.value0.range.start,
                    end: v1.end
                };
            };
            if (v instanceof PureScript_CST_Types.DeclKindSignature) {
                return {
                    start: v.value0.range.start,
                    end: (rangeOf(rangeOfLabeled(rangeOfName)(rangeOfType(dictRangeOf)))(v.value1)).end
                };
            };
            if (v instanceof PureScript_CST_Types.DeclSignature) {
                return rangeOf(rangeOfLabeled(rangeOfName)(rangeOfType(dictRangeOf)))(v.value0);
            };
            if (v instanceof PureScript_CST_Types.DeclValue) {
                return {
                    start: (rangeOf(rangeOfName)(v.value0.name)).start,
                    end: (rangeOf(rangeOfGuarded(dictRangeOf))(v.value0.guarded)).end
                };
            };
            if (v instanceof PureScript_CST_Types.DeclFixity) {
                return {
                    start: v.value0.keyword.value0.range.start,
                    end: (rangeOf(rangeOfFixityOp)(v.value0.operator)).end
                };
            };
            if (v instanceof PureScript_CST_Types.DeclForeign) {
                return {
                    start: v.value0.range.start,
                    end: (rangeOf(rangeOfForeign(dictRangeOf))(v.value2)).end
                };
            };
            if (v instanceof PureScript_CST_Types.DeclRole) {
                return {
                    start: v.value0.range.start,
                    end: (Data_Tuple.fst(Data_Array_NonEmpty.last(v.value3))).range.end
                };
            };
            if (v instanceof PureScript_CST_Types.DeclError) {
                return rangeOf(dictRangeOf)(v.value0);
            };
            throw new Error("Failed pattern match at PureScript.CST.Range (line 405, column 13 - line 481, column 16): " + [ v.constructor.name ]);
        }
    };
};
module.exports = {
    rangeOf: rangeOf,
    tokensOf: tokensOf,
    tokensOfTuple: tokensOfTuple,
    tokensOfMaybe: tokensOfMaybe,
    tokensOfArray: tokensOfArray,
    tokensOfNonEmptyArray: tokensOfNonEmptyArray,
    rangeOfVoid: rangeOfVoid,
    tokensOfVoid: tokensOfVoid,
    rangeOfRecoveredError: rangeOfRecoveredError,
    tokensOfRecoveredError: tokensOfRecoveredError,
    rangeOfModule: rangeOfModule,
    tokensOfModule: tokensOfModule,
    rangeOfName: rangeOfName,
    tokensOfName: tokensOfName,
    rangeOfQualifiedName: rangeOfQualifiedName,
    tokensOfQualifiedName: tokensOfQualifiedName,
    rangeOfWrapped: rangeOfWrapped,
    tokensOfWrapped: tokensOfWrapped,
    rangeOfSeparated: rangeOfSeparated,
    tokensOfSeparated: tokensOfSeparated,
    rangeOfLabeled: rangeOfLabeled,
    tokensOfLabeled: tokensOfLabeled,
    rangeOfOneOrDelimited: rangeOfOneOrDelimited,
    tokensOfOneOrDelimited: tokensOfOneOrDelimited,
    rangeOfType: rangeOfType,
    tokensOfType: tokensOfType,
    tokensOfRow: tokensOfRow,
    rangeOfTypeVarBinding: rangeOfTypeVarBinding,
    tokensOfTypeVarBinding: tokensOfTypeVarBinding,
    rangeOfExport: rangeOfExport,
    tokensOfExport: tokensOfExport,
    rangeOfDataMembers: rangeOfDataMembers,
    tokensOfDataMembers: tokensOfDataMembers,
    rangeOfImportDecl: rangeOfImportDecl,
    tokensOfImportDecl: tokensOfImportDecl,
    rangeOfImport: rangeOfImport,
    tokensOfImport: tokensOfImport,
    rangeOfDataCtor: rangeOfDataCtor,
    tokensOfDataCtor: tokensOfDataCtor,
    rangeOfDecl: rangeOfDecl,
    tokensOfDecl: tokensOfDecl,
    rangeOfClassFundep: rangeOfClassFundep,
    tokensOfClassFundep: tokensOfClassFundep,
    rangeOfInstance: rangeOfInstance,
    tokensOfInstance: tokensOfInstance,
    rangeOfGuarded: rangeOfGuarded,
    tokensOfGuarded: tokensOfGuarded,
    rangeOfGuardedExpr: rangeOfGuardedExpr,
    tokensOfGuardedExpr: tokensOfGuardedExpr,
    tokensOfPatternGuard: tokensOfPatternGuard,
    rangeOfFixityOp: rangeOfFixityOp,
    tokensOfFixityOp: tokensOfFixityOp,
    rangeOfForeign: rangeOfForeign,
    tokensOfForeign: tokensOfForeign,
    rangeOfInstanceBinding: rangeOfInstanceBinding,
    tokensOfInstanceBinding: tokensOfInstanceBinding,
    rangeOfExpr: rangeOfExpr,
    tokensOfExpr: tokensOfExpr,
    tokensOfRecordUpdate: tokensOfRecordUpdate,
    rangeOfDoStatement: rangeOfDoStatement,
    tokensOfDoStatement: tokensOfDoStatement,
    rangeOfLetBinding: rangeOfLetBinding,
    tokensOfLetBinding: tokensOfLetBinding,
    rangeOfBinder: rangeOfBinder,
    tokensOfBinder: tokensOfBinder,
    tokensOfRecordLabeled: tokensOfRecordLabeled,
    rangeOfWhere: rangeOfWhere,
    tokensOfWhere: tokensOfWhere
};
