// Data.Ord.Down/index.js
"use strict";
var Data_Bounded = require("../Data.Bounded/index.js");
var Data_Ord = require("../Data.Ord/index.js");
var Data_Ordering = require("../Data.Ordering/index.js");
var Data_Show = require("../Data.Show/index.js");

// | A newtype wrapper which provides a reversed `Ord` instance. For example:
// |
// |     sortBy (comparing Down) [1,2,3] = [3,2,1]
// |
var Down = function (x) {
    return x;
};
var showDown = function (dictShow) {
    return {
        show: function (v) {
            return "(Down " + (Data_Show.show(dictShow)(v) + ")");
        }
    };
};
var newtypeDown = {
    Coercible0: function () {
        return undefined;
    }
};
var eqDown = function (dictEq) {
    return dictEq;
};
var ordDown = function (dictOrd) {
    return {
        compare: function (v) {
            return function (v1) {
                return Data_Ordering.invert(Data_Ord.compare(dictOrd)(v)(v1));
            };
        },
        Eq0: function () {
            return eqDown(dictOrd.Eq0());
        }
    };
};
var boundedDown = function (dictBounded) {
    return {
        top: Data_Bounded.bottom(dictBounded),
        bottom: Data_Bounded.top(dictBounded),
        Ord0: function () {
            return ordDown(dictBounded.Ord0());
        }
    };
};
module.exports = {
    Down: Down,
    newtypeDown: newtypeDown,
    eqDown: eqDown,
    ordDown: ordDown,
    boundedDown: boundedDown,
    showDown: showDown
};
