// Type.Data.Symbol/index.js
"use strict";
var Data_Symbol = require("../Data.Symbol/index.js");
var Type_Proxy = require("../Type.Proxy/index.js");
var uncons = function () {
    return function (v) {
        return {
            head: Type_Proxy["Proxy"].value,
            tail: Type_Proxy["Proxy"].value
        };
    };
};
var equalsSymbol = function () {
    return function () {
        return {};
    };
};
var equals = function () {
    return function (v) {
        return function (v1) {
            return Type_Proxy["Proxy"].value;
        };
    };
};
var compare = function () {
    return function (v) {
        return function (v1) {
            return Type_Proxy["Proxy"].value;
        };
    };
};
var append = function () {
    return function (v) {
        return function (v1) {
            return Type_Proxy["Proxy"].value;
        };
    };
};
module.exports = {
    append: append,
    compare: compare,
    uncons: uncons,
    equals: equals,
    equalsSymbol: equalsSymbol,
    SProxy: Data_Symbol.SProxy,
    reflectSymbol: Data_Symbol.reflectSymbol,
    reifySymbol: Data_Symbol.reifySymbol
};
