// Dodo.Internal.Buffer/index.js
"use strict";
var Data_Foldable = require("../Data.Foldable/index.js");
var Data_Function = require("../Data.Function/index.js");
var Data_List = require("../Data.List/index.js");
var Data_List_Types = require("../Data.List.Types/index.js");
var Buffer = function (x) {
    return x;
};
var revert = function (v) {
    return {
        buffer: v.buffer,
        queue: Data_List.drop(1)(v.queue)
    };
};
var $$new = function (buffer) {
    return {
        buffer: buffer,
        queue: Data_List_Types.Nil.value
    };
};
var modify = function (f) {
    return function (v) {
        if (v.queue instanceof Data_List_Types.Cons) {
            return {
                buffer: v.buffer,
                queue: new Data_List_Types.Cons(new Data_List_Types.Cons(f, v.queue.value0), v.queue.value1)
            };
        };
        return {
            buffer: f(v.buffer),
            queue: v.queue
        };
    };
};
var isBranching = function (v) {
    return !Data_List["null"](v.queue);
};
var commit = function (v) {
    return {
        buffer: Data_Foldable.foldr(Data_List_Types.foldableList)(Data_Function.flip(Data_Foldable.foldr(Data_List_Types.foldableList)(Data_Function.apply)))(v.buffer)(v.queue),
        queue: Data_List_Types.Nil.value
    };
};
var get = function ($20) {
    return (function (v) {
        return v.buffer;
    })(commit($20));
};
var branch = function (v) {
    return {
        buffer: v.buffer,
        queue: new Data_List_Types.Cons(Data_List_Types.Nil.value, v.queue)
    };
};
module.exports = {
    Buffer: Buffer,
    branch: branch,
    commit: commit,
    revert: revert,
    modify: modify,
    get: get,
    "new": $$new,
    isBranching: isBranching
};
