// Data.Argonaut.Core/index.js

// | This module defines a data type and various functions for creating and
// | manipulating JSON values. The README contains additional documentation
// | for this module.
"use strict";
var $foreign = require("./foreign.js");
var Data_Eq = require("../Data.Eq/index.js");
var Data_Function = require("../Data.Function/index.js");
var Data_Maybe = require("../Data.Maybe/index.js");
var Data_Ord = require("../Data.Ord/index.js");
var Data_Ordering = require("../Data.Ordering/index.js");
var Foreign_Object = require("../Foreign.Object/index.js");
var verbJsonType = function (def) {
    return function (f) {
        return function (g) {
            return g(def)(f);
        };
    };
};

// Decoding
var toJsonType = verbJsonType(Data_Maybe.Nothing.value)(Data_Maybe.Just.create);

// | The number zero represented as `Json`
var jsonZero = $foreign.fromNumber(0.0);

// | The true boolean value represented as `Json`
var jsonTrue = $foreign.fromBoolean(true);

// | Constructs a `Json` object value containing only the provided key and value
var jsonSingletonObject = function (key) {
    return function (val) {
        return $foreign.fromObject(Foreign_Object.singleton(key)(val));
    };
};

// | Constructs a `Json` array value containing only the provided value
var jsonSingletonArray = function (j) {
    return $foreign.fromArray([ j ]);
};

// | The false boolean value represented as `Json`
var jsonFalse = $foreign.fromBoolean(false);

// | An empty string represented as `Json`
var jsonEmptyString = $foreign.fromString("");

// | An empty object represented as `Json`
var jsonEmptyObject = $foreign.fromObject(Foreign_Object.empty);

// | An empty array represented as `Json`
var jsonEmptyArray = $foreign.fromArray([  ]);

// Tests
var isJsonType = verbJsonType(false)(Data_Function["const"](true));
var ordJson = {
    compare: function (a) {
        return function (b) {
            return $foreign["_compare"](Data_Ordering.EQ.value, Data_Ordering.GT.value, Data_Ordering.LT.value, a, b);
        };
    },
    Eq0: function () {
        return eqJson;
    }
};
var eqJson = {
    eq: function (j1) {
        return function (j2) {
            return Data_Eq.eq(Data_Ordering.eqOrdering)(Data_Ord.compare(ordJson)(j1)(j2))(Data_Ordering.EQ.value);
        };
    }
};
var eqJNull = {
    eq: function (v) {
        return function (v1) {
            return true;
        };
    }
};
var ordJNull = {
    compare: function (v) {
        return function (v1) {
            return Data_Ordering.EQ.value;
        };
    },
    Eq0: function () {
        return eqJNull;
    }
};

// | A simpler version of `caseJson` which accepts a callback for when the
// | `Json` argument was a `String`, and a default value for all other cases.
var caseJsonString = function (d) {
    return function (f) {
        return function (j) {
            return $foreign["_caseJson"](Data_Function["const"](d), Data_Function["const"](d), Data_Function["const"](d), f, Data_Function["const"](d), Data_Function["const"](d), j);
        };
    };
};

// | Check if the provided `Json` is a `String`
var isString = isJsonType(caseJsonString);

// | Convert `Json` to a `String` value, if the `Json` is a string. To write a
// | `Json` value to a JSON string, see `stringify`.
var toString = toJsonType(caseJsonString);

// | A simpler version of `caseJson` which accepts a callback for when the
// | `Json` argument was an `Object`, and a default value for all other cases.
var caseJsonObject = function (d) {
    return function (f) {
        return function (j) {
            return $foreign["_caseJson"](Data_Function["const"](d), Data_Function["const"](d), Data_Function["const"](d), Data_Function["const"](d), Data_Function["const"](d), f, j);
        };
    };
};

// | Check if the provided `Json` is an `Object`
var isObject = isJsonType(caseJsonObject);

// | Convert `Json` to an `Object` of `Json` values, if the `Json` is an object.
var toObject = toJsonType(caseJsonObject);

// | A simpler version of `caseJson` which accepts a callback for when the
// | `Json` argument was a `Number`, and a default value for all other cases.
var caseJsonNumber = function (d) {
    return function (f) {
        return function (j) {
            return $foreign["_caseJson"](Data_Function["const"](d), Data_Function["const"](d), f, Data_Function["const"](d), Data_Function["const"](d), Data_Function["const"](d), j);
        };
    };
};

// | Check if the provided `Json` is a `Number`
var isNumber = isJsonType(caseJsonNumber);

// | Convert `Json` to a `Number` value, if the `Json` is a number.
var toNumber = toJsonType(caseJsonNumber);

// | A simpler version of `caseJson` which accepts a callback for when the
// | `Json` argument was null, and a default value for all other cases.
var caseJsonNull = function (d) {
    return function (f) {
        return function (j) {
            return $foreign["_caseJson"](f, Data_Function["const"](d), Data_Function["const"](d), Data_Function["const"](d), Data_Function["const"](d), Data_Function["const"](d), j);
        };
    };
};

// | Check if the provided `Json` is the `null` value
var isNull = isJsonType(caseJsonNull);

// | Convert `Json` to the `Unit` value if the `Json` is the null value
var toNull = toJsonType(caseJsonNull);

// | A simpler version of `caseJson` which accepts a callback for when the
// | `Json` argument was a `Boolean`, and a default value for all other cases.
var caseJsonBoolean = function (d) {
    return function (f) {
        return function (j) {
            return $foreign["_caseJson"](Data_Function["const"](d), f, Data_Function["const"](d), Data_Function["const"](d), Data_Function["const"](d), Data_Function["const"](d), j);
        };
    };
};

// | Check if the provided `Json` is a `Boolean`
var isBoolean = isJsonType(caseJsonBoolean);

// | Convert `Json` to a `Boolean` value, if the `Json` is a boolean.
var toBoolean = toJsonType(caseJsonBoolean);

// | A simpler version of `caseJson` which accepts a callback for when the
// | `Json` argument was a `Array Json`, and a default value for all other cases.
var caseJsonArray = function (d) {
    return function (f) {
        return function (j) {
            return $foreign["_caseJson"](Data_Function["const"](d), Data_Function["const"](d), Data_Function["const"](d), Data_Function["const"](d), f, Data_Function["const"](d), j);
        };
    };
};

// | Check if the provided `Json` is an `Array`
var isArray = isJsonType(caseJsonArray);

// | Convert `Json` to an `Array` of `Json` values, if the `Json` is an array.
var toArray = toJsonType(caseJsonArray);

// | Case analysis for `Json` values. See the README for more information.
var caseJson = function (a) {
    return function (b) {
        return function (c) {
            return function (d) {
                return function (e) {
                    return function (f) {
                        return function (json) {
                            return $foreign["_caseJson"](a, b, c, d, e, f, json);
                        };
                    };
                };
            };
        };
    };
};
module.exports = {
    caseJson: caseJson,
    caseJsonNull: caseJsonNull,
    caseJsonBoolean: caseJsonBoolean,
    caseJsonNumber: caseJsonNumber,
    caseJsonString: caseJsonString,
    caseJsonArray: caseJsonArray,
    caseJsonObject: caseJsonObject,
    isNull: isNull,
    isBoolean: isBoolean,
    isNumber: isNumber,
    isString: isString,
    isArray: isArray,
    isObject: isObject,
    toNull: toNull,
    toBoolean: toBoolean,
    toNumber: toNumber,
    toString: toString,
    toArray: toArray,
    toObject: toObject,
    jsonTrue: jsonTrue,
    jsonFalse: jsonFalse,
    jsonZero: jsonZero,
    jsonEmptyString: jsonEmptyString,
    jsonEmptyArray: jsonEmptyArray,
    jsonSingletonArray: jsonSingletonArray,
    jsonEmptyObject: jsonEmptyObject,
    jsonSingletonObject: jsonSingletonObject,
    eqJson: eqJson,
    ordJson: ordJson,
    fromBoolean: $foreign.fromBoolean,
    fromNumber: $foreign.fromNumber,
    fromString: $foreign.fromString,
    fromArray: $foreign.fromArray,
    fromObject: $foreign.fromObject,
    jsonNull: $foreign.jsonNull,
    stringify: $foreign.stringify,
    stringifyWithIndent: $foreign.stringifyWithIndent
};
