// Effect.Aff.Compat/index.js

// | This module provides compatability functions for constructing `Aff`s which
// | are defined via the FFI.
"use strict";
var Data_Either = require("../Data.Either/index.js");
var Effect_Aff = require("../Effect.Aff/index.js");
var Effect_Uncurried = require("../Effect.Uncurried/index.js");
var EffectFnCanceler = function (x) {
    return x;
};
var EffectFnAff = function (x) {
    return x;
};

// | Lift a FFI definition into an `Aff`. `EffectFnAff` makes use of `EffectFn` so
// | `Effect` thunks are unnecessary. A definition might follow this example:
// |
// | ```javascript
// | exports._myAff = function (onError, onSuccess) {
// |   var cancel = doSomethingAsync(function (err, res) {
// |     if (err) {
// |       onError(err);
// |     } else {
// |       onSuccess(res);
// |     }
// |   });
// |   return function (cancelError, onCancelerError, onCancelerSuccess) {
// |     cancel();
// |     onCancelerSuccess();
// |   };
// | };
// | ```
// |
// | ```purescript
// | foreign import _myAff :: EffectFnAff String
// |
// | myAff :: Aff String
// | myAff = fromEffectFnAff _myAff
// | ````
var fromEffectFnAff = function (v) {
    return Effect_Aff.makeAff(function (k) {
        return function __do() {
            var v1 = v(function ($4) {
                return k(Data_Either.Left.create($4))();
            }, function ($5) {
                return k(Data_Either.Right.create($5))();
            });
            return function (e) {
                return Effect_Aff.makeAff(function (k2) {
                    return function __do() {
                        v1(e, function ($6) {
                            return k2(Data_Either.Left.create($6))();
                        }, function ($7) {
                            return k2(Data_Either.Right.create($7))();
                        });
                        return Effect_Aff.nonCanceler;
                    };
                });
            };
        };
    });
};
module.exports = {
    EffectFnAff: EffectFnAff,
    EffectFnCanceler: EffectFnCanceler,
    fromEffectFnAff: fromEffectFnAff,
    mkEffectFn1: Effect_Uncurried.mkEffectFn1,
    mkEffectFn2: Effect_Uncurried.mkEffectFn2,
    mkEffectFn3: Effect_Uncurried.mkEffectFn3,
    runEffectFn1: Effect_Uncurried.runEffectFn1,
    runEffectFn2: Effect_Uncurried.runEffectFn2,
    runEffectFn3: Effect_Uncurried.runEffectFn3
};
