// Text.Parsing.Parser.Pos/index.js
"use strict";
var Data_Ord = require("../Data.Ord/index.js");
var Data_Ordering = require("../Data.Ordering/index.js");
var Data_Show = require("../Data.Show/index.js");

// | `Position` represents the position of the parser in the input.
// |
// | - `line` is the current line in the input
// | - `column` is the column of the next character in the current line that will be parsed
var Position = function (x) {
    return x;
};
var showPosition = {
    show: function (v) {
        return "(Position { line: " + (Data_Show.show(Data_Show.showInt)(v.line) + (", column: " + (Data_Show.show(Data_Show.showInt)(v.column) + " })")));
    }
};

// | The `Position` before any input has been parsed.
var initialPos = {
    line: 1,
    column: 1
};
var genericPosition = {
    to: function (x) {
        return x;
    },
    from: function (x) {
        return x;
    }
};
var eqPosition = {
    eq: function (x) {
        return function (y) {
            return x.column === y.column && x.line === y.line;
        };
    }
};
var ordPosition = {
    compare: function (x) {
        return function (y) {
            var v = Data_Ord.compare(Data_Ord.ordInt)(x.column)(y.column);
            if (v instanceof Data_Ordering.LT) {
                return Data_Ordering.LT.value;
            };
            if (v instanceof Data_Ordering.GT) {
                return Data_Ordering.GT.value;
            };
            return Data_Ord.compare(Data_Ord.ordInt)(x.line)(y.line);
        };
    },
    Eq0: function () {
        return eqPosition;
    }
};
module.exports = {
    Position: Position,
    initialPos: initialPos,
    genericPosition: genericPosition,
    showPosition: showPosition,
    eqPosition: eqPosition,
    ordPosition: ordPosition
};
