// Unsafe.Reference/index.js
"use strict";
var $foreign = require("./foreign.js");
var Data_Eq = require("../Data.Eq/index.js");

// | The `Eq` instance first checks `unsafeRefEq`, if `false` falls back to
// | the underlying `Eq` instance.
var UnsafeRefEqFallback = function (x) {
    return x;
};

// | The `Eq` instance is defined by `unsafeRefEq`.
var UnsafeRefEq = function (x) {
    return x;
};

// | Compares two values of the same type using strict (`===`) equality.
var unsafeRefEq = $foreign.reallyUnsafeRefEq;
var eqUnsafeRefEqFallback = function (dictEq) {
    return {
        eq: function (v) {
            return function (v1) {
                return unsafeRefEq(v)(v1) || Data_Eq.eq(dictEq)(v)(v1);
            };
        }
    };
};
var eqUnsafeRefEq = {
    eq: function (v) {
        return function (v1) {
            return unsafeRefEq(v)(v1);
        };
    }
};
module.exports = {
    unsafeRefEq: unsafeRefEq,
    UnsafeRefEq: UnsafeRefEq,
    UnsafeRefEqFallback: UnsafeRefEqFallback,
    eqUnsafeRefEq: eqUnsafeRefEq,
    eqUnsafeRefEqFallback: eqUnsafeRefEqFallback,
    reallyUnsafeRefEq: $foreign.reallyUnsafeRefEq
};
