// URI.Host.IPv6Address/index.js
"use strict";
var Control_Alt = require("../Control.Alt/index.js");
var Control_Apply = require("../Control.Apply/index.js");
var Data_Array = require("../Data.Array/index.js");
var Data_Eq = require("../Data.Eq/index.js");
var Data_Functor = require("../Data.Functor/index.js");
var Data_Identity = require("../Data.Identity/index.js");
var Data_Ord = require("../Data.Ord/index.js");
var Data_Show = require("../Data.Show/index.js");
var Data_String_CodeUnits = require("../Data.String.CodeUnits/index.js");
var Text_Parsing_Parser = require("../Text.Parsing.Parser/index.js");
var Text_Parsing_Parser_Combinators = require("../Text.Parsing.Parser.Combinators/index.js");
var Text_Parsing_Parser_String = require("../Text.Parsing.Parser.String/index.js");
var Text_Parsing_Parser_String_Basic = require("../Text.Parsing.Parser.String.Basic/index.js");

// | This type and parser are much too forgiving currently, allowing almost
// | anything through that looks vaguely IPv6ish.
var IPv6Address = function (x) {
    return x;
};
var unsafeToString = function (v) {
    return "[" + (v + "]");
};
var unsafeFromString = IPv6Address;
var showIPv6Address = {
    show: function (v) {
        return "(IPv6Address.unsafeFromString " + (Data_Show.show(Data_Show.showString)(v) + ")");
    }
};
var parser = (function () {
    var ipv6Char = Control_Alt.alt(Text_Parsing_Parser.altParserT(Data_Identity.monadIdentity))(Control_Alt.alt(Text_Parsing_Parser.altParserT(Data_Identity.monadIdentity))(Text_Parsing_Parser_String_Basic.hexDigit(Data_Identity.monadIdentity))(Text_Parsing_Parser_String["char"](Data_Identity.monadIdentity)(":")))(Text_Parsing_Parser_String["char"](Data_Identity.monadIdentity)("."));
    return Text_Parsing_Parser_Combinators.withErrorMessage(Data_Identity.monadIdentity)(Data_Functor.map(Text_Parsing_Parser.functorParserT(Data_Identity.functorIdentity))(IPv6Address)(Control_Apply.applyFirst(Text_Parsing_Parser.applyParserT(Data_Identity.monadIdentity))(Control_Apply.applySecond(Text_Parsing_Parser.applyParserT(Data_Identity.monadIdentity))(Text_Parsing_Parser_String["char"](Data_Identity.monadIdentity)("["))(Data_Functor.map(Text_Parsing_Parser.functorParserT(Data_Identity.functorIdentity))(Data_String_CodeUnits.fromCharArray)(Data_Array.some(Text_Parsing_Parser.alternativeParserT(Data_Identity.monadIdentity))(Text_Parsing_Parser.lazyParserT)(ipv6Char))))(Text_Parsing_Parser_String["char"](Data_Identity.monadIdentity)("]"))))("IPv6 address");
})();
var ordIPv6Address = Data_Ord.ordString;
var eqIPv6Address = Data_Eq.eqString;
module.exports = {
    unsafeFromString: unsafeFromString,
    unsafeToString: unsafeToString,
    parser: parser,
    eqIPv6Address: eqIPv6Address,
    ordIPv6Address: ordIPv6Address,
    showIPv6Address: showIPv6Address
};
