// Data.Lens.Internal.Bazaar/index.js
"use strict";
var Control_Applicative = require("../Control.Applicative/index.js");
var Data_Bitraversable = require("../Data.Bitraversable/index.js");
var Data_Function = require("../Data.Function/index.js");
var Data_Functor = require("../Data.Functor/index.js");
var Data_Traversable = require("../Data.Traversable/index.js");
var Data_Tuple = require("../Data.Tuple/index.js");
var Bazaar = function (x) {
    return x;
};
var runBazaar = function (v) {
    return function (dictApplicative) {
        return v(dictApplicative);
    };
};
var profunctorBazaar = {
    dimap: function (f) {
        return function (g) {
            return function (v) {
                return function (dictApplicative) {
                    return function (pafb) {
                        return function (s) {
                            return Data_Functor.map((dictApplicative.Apply0()).Functor0())(g)(v(dictApplicative)(pafb)(f(s)));
                        };
                    };
                };
            };
        };
    }
};
var strongBazaar = {
    first: function (v) {
        return function (dictApplicative) {
            return function (pafb) {
                return function (v1) {
                    return Data_Functor.map((dictApplicative.Apply0()).Functor0())(Data_Function.flip(Data_Tuple.Tuple.create)(v1.value1))(v(dictApplicative)(pafb)(v1.value0));
                };
            };
        };
    },
    second: function (v) {
        return function (dictApplicative) {
            return function (pafb) {
                return function (v1) {
                    return Data_Functor.map((dictApplicative.Apply0()).Functor0())(Data_Tuple.Tuple.create(v1.value0))(v(dictApplicative)(pafb)(v1.value1));
                };
            };
        };
    },
    Profunctor0: function () {
        return profunctorBazaar;
    }
};
var choiceBazaar = {
    left: function (v) {
        return function (dictApplicative) {
            return function (pafb) {
                return function (e) {
                    return Data_Bitraversable.bitraverse(Data_Bitraversable.bitraversableEither)(dictApplicative)(v(dictApplicative)(pafb))(Control_Applicative.pure(dictApplicative))(e);
                };
            };
        };
    },
    right: function (v) {
        return function (dictApplicative) {
            return function (pafb) {
                return function (e) {
                    return Data_Traversable.traverse(Data_Traversable.traversableEither)(dictApplicative)(v(dictApplicative)(pafb))(e);
                };
            };
        };
    },
    Profunctor0: function () {
        return profunctorBazaar;
    }
};
var wanderBazaar = {
    wander: function (w) {
        return function (v) {
            return function (dictApplicative) {
                return function (pafb) {
                    return function (s) {
                        return w(dictApplicative)(v(dictApplicative)(pafb))(s);
                    };
                };
            };
        };
    },
    Strong0: function () {
        return strongBazaar;
    },
    Choice1: function () {
        return choiceBazaar;
    }
};
module.exports = {
    Bazaar: Bazaar,
    runBazaar: runBazaar,
    profunctorBazaar: profunctorBazaar,
    strongBazaar: strongBazaar,
    choiceBazaar: choiceBazaar,
    wanderBazaar: wanderBazaar
};
