// Main/index.js
"use strict";
var Control_Applicative = require("../Control.Applicative/index.js");
var Control_Bind = require("../Control.Bind/index.js");
var Data_Functor = require("../Data.Functor/index.js");
var Data_Maybe = require("../Data.Maybe/index.js");
var Data_Unit = require("../Data.Unit/index.js");
var Effect = require("../Effect/index.js");
var Effect_Aff = require("../Effect.Aff/index.js");
var Effect_Class = require("../Effect.Class/index.js");
var Effect_Class_Console = require("../Effect.Class.Console/index.js");
var Halogen_Subscription = require("../Halogen.Subscription/index.js");
var Purescriptify_AppM = require("../Purescriptify.AppM/index.js");
var Purescriptify_Component_Routing = require("../Purescriptify.Component.Routing/index.js");
var Purescriptify_Config = require("../Purescriptify.Config/index.js");
var Purescriptify_Root = require("../Purescriptify.Root/index.js");
var React_Basic = require("../React.Basic/index.js");
var React_Basic_DOM = require("../React.Basic.DOM/index.js");
var React_Basic_Hooks = require("../React.Basic.Hooks/index.js");
var React_Basic_Hooks_Internal = require("../React.Basic.Hooks.Internal/index.js");
var Type_Equality = require("../Type.Equality/index.js");
var Web_DOM_NonElementParentNode = require("../Web.DOM.NonElementParentNode/index.js");
var Web_HTML = require("../Web.HTML/index.js");
var Web_HTML_HTMLDocument = require("../Web.HTML.HTMLDocument/index.js");
var Web_HTML_Window = require("../Web.HTML.Window/index.js");
var mkGlobalIO = (function () {
    var create = function __do() {
        var v = Halogen_Subscription.create();
        var v1 = Halogen_Subscription.create();
        return {
            eventsEmitter: v1.emitter,
            eventsListener: v1.listener,
            actionsEmitter: v.emitter,
            actionsListener: v.listener
        };
    };
    return create;
})();
var main = function __do() {
    var rootMay = Control_Bind.bindFlipped(Effect.bindEffect)(Web_DOM_NonElementParentNode.getElementById("root"))(Data_Functor.map(Effect.functorEffect)(Web_HTML_HTMLDocument.toNonElementParentNode)(Control_Bind.bindFlipped(Effect.bindEffect)(Web_HTML_Window.document)(Web_HTML.window)))();
    if (rootMay instanceof Data_Maybe.Just) {
        var v = Purescriptify_Component_Routing.mkRoutingManager();
        var logLevel = (function () {
            if (Purescriptify_Config.nodeEnv === "production") {
                return Purescriptify_AppM.Prod.value;
            };
            if (Purescriptify_Config.nodeEnv === "development") {
                return Purescriptify_AppM.Dev.value;
            };
            return Purescriptify_AppM.Dev.value;
        })();
        var globalIO = mkGlobalIO();
        return Effect_Aff.launchAff_(Control_Bind.bind(Effect_Aff.bindAff)(Purescriptify_AppM.runAppM({
            routing: v.value0,
            logLevel: logLevel,
            globalIO: globalIO
        })(Purescriptify_Root.mkRoot(Purescriptify_AppM.monadAffAppM)(Purescriptify_AppM.monadGlobalAppM)(Purescriptify_AppM.monadRoutingAppM)(Purescriptify_AppM.monadHaloAppM)(Purescriptify_AppM.monadUsersAppM)(Purescriptify_AppM.nowAppM)(Purescriptify_AppM.logMessagesAppM)))(function (rootComponent) {
            return Control_Bind.bind(Effect_Aff.bindAff)(Effect_Class.liftEffect(Effect_Aff.monadEffectAff)(React_Basic_Hooks.component("Main")(function (v1) {
                return Control_Applicative.pure(React_Basic_Hooks_Internal.applicativeRender(Type_Equality.refl))(React_Basic.fragment([ v.value1, rootComponent({}) ]));
            })))(function (mainComponent) {
                return Effect_Class.liftEffect(Effect_Aff.monadEffectAff)(React_Basic_DOM.render(mainComponent(Data_Unit.unit))(rootMay.value0));
            });
        }))();
    };
    if (rootMay instanceof Data_Maybe.Nothing) {
        return Effect_Class_Console.log(Effect_Class.monadEffectEffect)("No root element.")();
    };
    throw new Error("Failed pattern match at Main (line 40, column 3 - line 63, column 38): " + [ rootMay.constructor.name ]);
};
module.exports = {
    main: main,
    mkGlobalIO: mkGlobalIO
};
