// Data.DateTime.Gen/index.js
"use strict";
var Control_Apply = require("../Control.Apply/index.js");
var Data_Date_Gen = require("../Data.Date.Gen/index.js");
var Data_DateTime = require("../Data.DateTime/index.js");
var Data_Functor = require("../Data.Functor/index.js");
var Data_Time_Gen = require("../Data.Time.Gen/index.js");

// | Generates a random `DateTime` between 1st Jan 1900 00:00:00 and
// | 31st Dec 2100 23:59:59, inclusive.
var genDateTime = function (dictMonadGen) {
    return Control_Apply.apply(((dictMonadGen.Monad0()).Bind1()).Apply0())(Data_Functor.map((((dictMonadGen.Monad0()).Bind1()).Apply0()).Functor0())(Data_DateTime.DateTime.create)(Data_Date_Gen.genDate(dictMonadGen)))(Data_Time_Gen.genTime(dictMonadGen));
};
module.exports = {
    genDateTime: genDateTime,
    genDate: Data_Date_Gen.genDate,
    genDay: Data_Date_Gen.genDay,
    genMonth: Data_Date_Gen.genMonth,
    genWeekday: Data_Date_Gen.genWeekday,
    genYear: Data_Date_Gen.genYear,
    genHour: Data_Time_Gen.genHour,
    genMillisecond: Data_Time_Gen.genMillisecond,
    genMinute: Data_Time_Gen.genMinute,
    genSecond: Data_Time_Gen.genSecond,
    genTime: Data_Time_Gen.genTime
};
