// Node.Process/index.js

// | Bindings to the global `process` object in Node.js. See also [the Node API documentation](https://nodejs.org/api/process.html)
"use strict";
var $foreign = require("./foreign.js");
var Data_Posix_Signal = require("../Data.Posix.Signal/index.js");
var Foreign_Object = require("../Foreign.Object/index.js");
var Node_Platform = require("../Node.Platform/index.js");
var Unsafe_Coerce = require("../Unsafe.Coerce/index.js");

// | Get the Node.js version.
var version = $foreign.process.version;

// | Check whether the standard output stream appears to be attached to a TTY.
// | It is a good idea to check this before printing ANSI codes to stdout
// | (e.g. for coloured text in the terminal).
var stdoutIsTTY = $foreign.process.stdout.isTTY;

// | The standard output stream. Note that this stream cannot be closed; calling
// | `end` will result in an exception being thrown.
var stdout = $foreign.process.stdout;

// | Check whether the standard input stream appears to be attached to a TTY.
// | It is a good idea to check this before processing the input data from stdin.
var stdinIsTTY = $foreign.process.stdin.isTTY;

// | The standard input stream. Note that this stream will never emit an `end`
// | event, so any handlers attached via `onEnd` will never be called.
var stdin = $foreign.process.stdin;

// | Check whether the standard error stream appears to be attached to a TTY.
// | It is a good idea to check this before printing ANSI codes to stderr
// | (e.g. for coloured text in the terminal).
var stderrIsTTY = $foreign.process.stderr.isTTY;

// | The standard error stream. Note that this stream cannot be closed; calling
// | `end` will result in an exception being thrown.
var stderr = $foreign.process.stderr;
var platformStr = $foreign.process.platform;
var platform = Node_Platform.fromString(platformStr);
var pid = $foreign.process.pid;

// | Install a handler for a particular signal.
var onSignal = function (sig) {
    return $foreign.onSignalImpl(Data_Posix_Signal.toString(sig));
};
var mkEffect = Unsafe_Coerce.unsafeCoerce;

// | Register a callback to run as soon as the current event loop runs to
// | completion.
var nextTick = function (callback) {
    return mkEffect(function (v) {
        return $foreign.process.nextTick(callback);
    });
};
var lookupMutableObject = function (k) {
    return function (o) {
        return mkEffect(function (v) {
            return Foreign_Object.lookup(k)(o);
        });
    };
};

// | Lookup a particular environment variable.
var lookupEnv = function (k) {
    return lookupMutableObject(k)($foreign.process.env);
};

// | Get a copy of the current environment.
var getEnv = $foreign.copyObject($foreign.process.env);

// | The absolute pathname of the `node` executable that started the
// | process.
var execPath = mkEffect(function (v) {
    return $foreign.process.execPath;
});

// | Node-specific options passed to the `node` executable.
var execArgv = $foreign.copyArray($foreign.process.execArgv);

// | Get the current working directory of the process.
var cwd = $foreign.process.cwd;

// | Get an array containing the command line arguments.
var argv = $foreign.copyArray($foreign.process.argv);
module.exports = {
    onSignal: onSignal,
    nextTick: nextTick,
    argv: argv,
    execArgv: execArgv,
    execPath: execPath,
    cwd: cwd,
    getEnv: getEnv,
    lookupEnv: lookupEnv,
    pid: pid,
    platform: platform,
    stdin: stdin,
    stdout: stdout,
    stderr: stderr,
    stdinIsTTY: stdinIsTTY,
    stdoutIsTTY: stdoutIsTTY,
    stderrIsTTY: stderrIsTTY,
    version: version,
    onBeforeExit: $foreign.onBeforeExit,
    onExit: $foreign.onExit,
    onUncaughtException: $foreign.onUncaughtException,
    onUnhandledRejection: $foreign.onUnhandledRejection,
    chdir: $foreign.chdir,
    setEnv: $foreign.setEnv,
    unsetEnv: $foreign.unsetEnv,
    exit: $foreign.exit
};
