// Test.Spec.Speed/index.js
"use strict";
var Data_Eq_Generic = require("../Data.Eq.Generic/index.js");
var Data_Generic_Rep = require("../Data.Generic.Rep/index.js");
var Data_Ord = require("../Data.Ord/index.js");
var Data_Show_Generic = require("../Data.Show.Generic/index.js");
var Data_Time_Duration = require("../Data.Time.Duration/index.js");
var Test_Spec_Style = require("../Test.Spec.Style/index.js");
var Fast = (function () {
    function Fast() {

    };
    Fast.value = new Fast();
    return Fast;
})();
var Medium = (function () {
    function Medium() {

    };
    Medium.value = new Medium();
    return Medium;
})();
var Slow = (function () {
    function Slow() {

    };
    Slow.value = new Slow();
    return Slow;
})();
var toStyle = function (v) {
    if (v instanceof Fast) {
        return Test_Spec_Style.dim;
    };
    if (v instanceof Medium) {
        return Test_Spec_Style.yellow;
    };
    if (v instanceof Slow) {
        return Test_Spec_Style.red;
    };
    throw new Error("Failed pattern match at Test.Spec.Speed (line 23, column 1 - line 23, column 26): " + [ v.constructor.name ]);
};
var speedOf = function (v) {
    return function (v1) {
        if (Data_Ord.greaterThan(Data_Time_Duration.ordMilliseconds)(v1)(v)) {
            return Slow.value;
        };
        if (v1 > v / 2.0) {
            return Medium.value;
        };
        return Fast.value;
    };
};
var genericSpeed = {
    to: function (x) {
        if (x instanceof Data_Generic_Rep.Inl) {
            return Fast.value;
        };
        if (x instanceof Data_Generic_Rep.Inr && x.value0 instanceof Data_Generic_Rep.Inl) {
            return Medium.value;
        };
        if (x instanceof Data_Generic_Rep.Inr && x.value0 instanceof Data_Generic_Rep.Inr) {
            return Slow.value;
        };
        throw new Error("Failed pattern match at Test.Spec.Speed (line 14, column 1 - line 14, column 48): " + [ x.constructor.name ]);
    },
    from: function (x) {
        if (x instanceof Fast) {
            return new Data_Generic_Rep.Inl(Data_Generic_Rep.NoArguments.value);
        };
        if (x instanceof Medium) {
            return new Data_Generic_Rep.Inr(new Data_Generic_Rep.Inl(Data_Generic_Rep.NoArguments.value));
        };
        if (x instanceof Slow) {
            return new Data_Generic_Rep.Inr(new Data_Generic_Rep.Inr(Data_Generic_Rep.NoArguments.value));
        };
        throw new Error("Failed pattern match at Test.Spec.Speed (line 14, column 1 - line 14, column 48): " + [ x.constructor.name ]);
    }
};
var showEq = {
    eq: Data_Eq_Generic.genericEq(genericSpeed)(Data_Eq_Generic.genericEqSum(Data_Eq_Generic.genericEqConstructor(Data_Eq_Generic.genericEqNoArguments))(Data_Eq_Generic.genericEqSum(Data_Eq_Generic.genericEqConstructor(Data_Eq_Generic.genericEqNoArguments))(Data_Eq_Generic.genericEqConstructor(Data_Eq_Generic.genericEqNoArguments))))
};
var showSpeed = {
    show: Data_Show_Generic.genericShow(genericSpeed)(Data_Show_Generic.genericShowSum(Data_Show_Generic.genericShowConstructor(Data_Show_Generic.genericShowArgsNoArguments)({
        reflectSymbol: function () {
            return "Fast";
        }
    }))(Data_Show_Generic.genericShowSum(Data_Show_Generic.genericShowConstructor(Data_Show_Generic.genericShowArgsNoArguments)({
        reflectSymbol: function () {
            return "Medium";
        }
    }))(Data_Show_Generic.genericShowConstructor(Data_Show_Generic.genericShowArgsNoArguments)({
        reflectSymbol: function () {
            return "Slow";
        }
    }))))
};
module.exports = {
    Fast: Fast,
    Medium: Medium,
    Slow: Slow,
    speedOf: speedOf,
    toStyle: toStyle,
    genericSpeed: genericSpeed,
    showSpeed: showSpeed,
    showEq: showEq
};
