// Control.Comonad.Store/index.js

// | This module defines the `Store` comonad.
"use strict";
var Control_Comonad_Store_Class = require("../Control.Comonad.Store.Class/index.js");
var Control_Comonad_Store_Trans = require("../Control.Comonad.Store.Trans/index.js");
var Data_Functor = require("../Data.Functor/index.js");
var Data_Newtype = require("../Data.Newtype/index.js");
var Data_Tuple = require("../Data.Tuple/index.js");

// | Create a value in context in the `Store` comonad.
var store = function (f) {
    return function (x) {
        return new Data_Tuple.Tuple(f, x);
    };
};

// | Unwrap a value in the `Store` comonad.
var runStore = function (v) {
    return Data_Tuple.swap(Data_Functor.map(Data_Tuple.functorTuple)(Data_Newtype.unwrap())(Data_Tuple.swap(v)));
};
module.exports = {
    runStore: runStore,
    store: store,
    experiment: Control_Comonad_Store_Class.experiment,
    peek: Control_Comonad_Store_Class.peek,
    peeks: Control_Comonad_Store_Class.peeks,
    pos: Control_Comonad_Store_Class.pos,
    seek: Control_Comonad_Store_Class.seek,
    seeks: Control_Comonad_Store_Class.seeks,
    StoreT: Control_Comonad_Store_Trans.StoreT,
    runStoreT: Control_Comonad_Store_Trans.runStoreT
};
