// Node.FS.Async/index.js
"use strict";
var $foreign = require("./foreign.js");
var Data_DateTime_Instant = require("../Data.DateTime.Instant/index.js");
var Data_Either = require("../Data.Either/index.js");
var Data_EuclideanRing = require("../Data.EuclideanRing/index.js");
var Data_Functor = require("../Data.Functor/index.js");
var Data_Int = require("../Data.Int/index.js");
var Data_Maybe = require("../Data.Maybe/index.js");
var Data_Nullable = require("../Data.Nullable/index.js");
var Data_Show = require("../Data.Show/index.js");
var Effect_Unsafe = require("../Effect.Unsafe/index.js");
var Node_Buffer = require("../Node.Buffer/index.js");
var Node_Buffer_Class = require("../Node.Buffer.Class/index.js");
var Node_Encoding = require("../Node.Encoding/index.js");
var Node_FS = require("../Node.FS/index.js");
var Node_FS_Internal = require("../Node.FS.Internal/index.js");
var Node_FS_Perms = require("../Node.FS.Perms/index.js");
var Node_FS_Stats = require("../Node.FS.Stats/index.js");
var handleCallback = function (cb) {
    return $foreign.handleCallbackImpl(Data_Either.Left.create, Data_Either.Right.create, cb);
};
var fs = Node_FS_Internal.unsafeRequireFS;

// | Creates a link to an existing file.
var link = function (src) {
    return function (dst) {
        return function (cb) {
            return Node_FS_Internal.mkEffect(function (v) {
                return fs.link(src, dst, handleCallback(cb));
            });
        };
    };
};

// | Makes a new directory with the specified permissions.
var mkdir$prime = function (file) {
    return function (perms) {
        return function (cb) {
            return Node_FS_Internal.mkEffect(function (v) {
                return fs.mkdir(file, {
                    recursive: false,
                    mode: Node_FS_Perms.permsToString(perms)
                }, handleCallback(cb));
            });
        };
    };
};

// | Makes a new directory.
var mkdir = function (path) {
    return mkdir$prime(path)(Node_FS_Perms.mkPerms(Node_FS_Perms.all)(Node_FS_Perms.all)(Node_FS_Perms.all));
};

// | Makes a new directory (and any directories that don't exist
// | in the path) with the specified permissions.
var mkdirRecursive$prime = function (file) {
    return function (perms) {
        return function (cb) {
            return Node_FS_Internal.mkEffect(function (v) {
                return fs.mkdir(file, {
                    recursive: true,
                    mode: Node_FS_Perms.permsToString(perms)
                }, handleCallback(cb));
            });
        };
    };
};

// | Makes a new directory and any directories that don't exist
// | in the path. Similar to `mkdir -p`.
var mkdirRecursive = function (path) {
    return mkdirRecursive$prime(path)(Node_FS_Perms.mkPerms(Node_FS_Perms.all)(Node_FS_Perms.all)(Node_FS_Perms.all));
};

// | Reads the entire contents of a file returning the result as a raw buffer.
var readFile = function (file) {
    return function (cb) {
        return Node_FS_Internal.mkEffect(function (v) {
            return fs.readFile(file, {}, handleCallback(cb));
        });
    };
};

// | Reads the entire contents of a text file with the specified encoding.
var readTextFile = function (encoding) {
    return function (file) {
        return function (cb) {
            return Node_FS_Internal.mkEffect(function (v) {
                return fs.readFile(file, {
                    encoding: Data_Show.show(Node_Encoding.showEncoding)(encoding)
                }, handleCallback(cb));
            });
        };
    };
};

// | Reads the contents of a directory.
var readdir = function (file) {
    return function (cb) {
        return Node_FS_Internal.mkEffect(function (v) {
            return fs.readdir(file, handleCallback(cb));
        });
    };
};

// | Reads the value of a symlink.
var readlink = function (path) {
    return function (cb) {
        return Node_FS_Internal.mkEffect(function (v) {
            return fs.readlink(path, handleCallback(cb));
        });
    };
};

// | Find the canonicalized absolute location for a path.
var realpath = function (path) {
    return function (cb) {
        return Node_FS_Internal.mkEffect(function (v) {
            return fs.realpath(path, {}, handleCallback(cb));
        });
    };
};

// | Find the canonicalized absolute location for a path using a cache object
// | for already resolved paths.
var realpath$prime = function (path) {
    return function (cache) {
        return function (cb) {
            return Node_FS_Internal.mkEffect(function (v) {
                return fs.realpath(path, cache, handleCallback(cb));
            });
        };
    };
};

// | Renames a file.
var rename = function (oldFile) {
    return function (newFile) {
        return function (cb) {
            return Node_FS_Internal.mkEffect(function (v) {
                return fs.rename(oldFile, newFile, handleCallback(cb));
            });
        };
    };
};

// | Deletes a directory.
var rmdir = function (file) {
    return function (cb) {
        return Node_FS_Internal.mkEffect(function (v) {
            return fs.rmdir(file, handleCallback(cb));
        });
    };
};

// | Gets file statistics.
var stat = function (file) {
    return function (cb) {
        return Node_FS_Internal.mkEffect(function (v) {
            return fs.stat(file, handleCallback((function () {
                var $29 = Data_Functor.map(Data_Either.functorEither)(Node_FS_Stats.Stats.create);
                return function ($30) {
                    return cb($29($30));
                };
            })()));
        });
    };
};

// | Creates a symlink.
var symlink = function (src) {
    return function (dest) {
        return function (ty) {
            return function (cb) {
                return Node_FS_Internal.mkEffect(function (v) {
                    return fs.symlink(src, dest, Node_FS.symlinkTypeToNode(ty), handleCallback(cb));
                });
            };
        };
    };
};

// | Truncates a file to the specified length.
var truncate = function (file) {
    return function (len) {
        return function (cb) {
            return Node_FS_Internal.mkEffect(function (v) {
                return fs.truncate(file, len, handleCallback(cb));
            });
        };
    };
};

// | Deletes a file.
var unlink = function (file) {
    return function (cb) {
        return Node_FS_Internal.mkEffect(function (v) {
            return fs.unlink(file, handleCallback(cb));
        });
    };
};

// | Sets the accessed and modified times for the specified file.
var utimes = function (file) {
    return function (atime) {
        return function (mtime) {
            return function (cb) {
                var toEpochMilliseconds = function ($31) {
                    return Data_DateTime_Instant.unInstant(Data_DateTime_Instant.fromDateTime($31));
                };
                var ms = function (v) {
                    return Data_Int.round(v);
                };
                var fromDate = function (date) {
                    return Data_EuclideanRing.div(Data_EuclideanRing.euclideanRingInt)(ms(toEpochMilliseconds(date)))(1000);
                };
                return Node_FS_Internal.mkEffect(function (v) {
                    return fs.utimes(file, fromDate(atime), fromDate(mtime), handleCallback(cb));
                });
            };
        };
    };
};

// | Writes a buffer to a file.
var writeFile = function (file) {
    return function (buff) {
        return function (cb) {
            return Node_FS_Internal.mkEffect(function (v) {
                return fs.writeFile(file, buff, {}, handleCallback(cb));
            });
        };
    };
};

// | Writes text to a file using the specified encoding.
var writeTextFile = function (encoding) {
    return function (file) {
        return function (buff) {
            return function (cb) {
                return Node_FS_Internal.mkEffect(function (v) {
                    return fs.writeFile(file, buff, {
                        encoding: Data_Show.show(Node_Encoding.showEncoding)(encoding)
                    }, handleCallback(cb));
                });
            };
        };
    };
};

// | Write to a file asynchronously. See the [Node Documentation](https://nodejs.org/api/fs.html#fs_fs_write_fd_buffer_offset_length_position_callback)
// | for details.
var fdWrite = function (fd) {
    return function (buff) {
        return function (off) {
            return function (len) {
                return function (pos) {
                    return function (cb) {
                        return Node_FS_Internal.mkEffect(function (v) {
                            return fs.write(fd, buff, off, len, Data_Nullable.toNullable(pos), handleCallback(cb));
                        });
                    };
                };
            };
        };
    };
};

// | Read from a file asynchronously. See the [Node Documentation](https://nodejs.org/api/fs.html#fs_fs_read_fd_buffer_offset_length_position_callback)
// | for details.
var fdRead = function (fd) {
    return function (buff) {
        return function (off) {
            return function (len) {
                return function (pos) {
                    return function (cb) {
                        return Node_FS_Internal.mkEffect(function (v) {
                            return fs.read(fd, buff, off, len, Data_Nullable.toNullable(pos), handleCallback(cb));
                        });
                    };
                };
            };
        };
    };
};

// | Open a file asynchronously. See the [Node Documentation](https://nodejs.org/api/fs.html#fs_fs_open_path_flags_mode_callback)
// | for details.
var fdOpen = function (file) {
    return function (flags) {
        return function (mode) {
            return function (cb) {
                return Node_FS_Internal.mkEffect(function (v) {
                    return fs.open(file, Node_FS.fileFlagsToNode(flags), Data_Nullable.toNullable(mode), handleCallback(cb));
                });
            };
        };
    };
};

// | Convenience function to fill the whole buffer from the current
// | file position.
var fdNext = function (fd) {
    return function (buff) {
        return function (cb) {
            return function __do() {
                var sz = Node_Buffer_Class.size(Node_Buffer.mutableBufferEffect)(buff)();
                return fdRead(fd)(buff)(0)(sz)(Data_Maybe.Nothing.value)(cb)();
            };
        };
    };
};

// | Close a file asynchronously. See the [Node Documentation](https://nodejs.org/api/fs.html#fs_fs_close_fd_callback)
// | for details.
var fdClose = function (fd) {
    return function (cb) {
        return Node_FS_Internal.mkEffect(function (v) {
            return fs.close(fd, handleCallback(cb));
        });
    };
};

// | Convenience function to append the whole buffer to the current
// | file position.
var fdAppend = function (fd) {
    return function (buff) {
        return function (cb) {
            return function __do() {
                var sz = Node_Buffer_Class.size(Node_Buffer.mutableBufferEffect)(buff)();
                return fdWrite(fd)(buff)(0)(sz)(Data_Maybe.Nothing.value)(cb)();
            };
        };
    };
};

// | Check if the path exists.
var exists = function (file) {
    return function (cb) {
        return Node_FS_Internal.mkEffect(function (v) {
            return fs.exists(file, function (b) {
                return Effect_Unsafe.unsafePerformEffect(cb(b));
            });
        });
    };
};

// | Changes the ownership of a file.
var chown = function (file) {
    return function (uid) {
        return function (gid) {
            return function (cb) {
                return Node_FS_Internal.mkEffect(function (v) {
                    return fs.chown(file, uid, gid, handleCallback(cb));
                });
            };
        };
    };
};

// | Changes the permissions of a file.
var chmod = function (file) {
    return function (perms) {
        return function (cb) {
            return Node_FS_Internal.mkEffect(function (v) {
                return fs.chmod(file, Node_FS_Perms.permsToString(perms), handleCallback(cb));
            });
        };
    };
};

// | Appends text to a file using the specified encoding.
var appendTextFile = function (encoding) {
    return function (file) {
        return function (buff) {
            return function (cb) {
                return Node_FS_Internal.mkEffect(function (v) {
                    return fs.appendFile(file, buff, {
                        encoding: Data_Show.show(Node_Encoding.showEncoding)(encoding)
                    }, handleCallback(cb));
                });
            };
        };
    };
};

// | Appends the contents of a buffer to a file.
var appendFile = function (file) {
    return function (buff) {
        return function (cb) {
            return Node_FS_Internal.mkEffect(function (v) {
                return fs.appendFile(file, buff, {}, handleCallback(cb));
            });
        };
    };
};
module.exports = {
    rename: rename,
    truncate: truncate,
    chown: chown,
    chmod: chmod,
    stat: stat,
    link: link,
    symlink: symlink,
    readlink: readlink,
    realpath: realpath,
    "realpath'": realpath$prime,
    unlink: unlink,
    rmdir: rmdir,
    mkdir: mkdir,
    mkdirRecursive: mkdirRecursive,
    "mkdir'": mkdir$prime,
    readdir: readdir,
    utimes: utimes,
    readFile: readFile,
    readTextFile: readTextFile,
    writeFile: writeFile,
    writeTextFile: writeTextFile,
    appendFile: appendFile,
    appendTextFile: appendTextFile,
    exists: exists,
    fdOpen: fdOpen,
    fdRead: fdRead,
    fdNext: fdNext,
    fdWrite: fdWrite,
    fdAppend: fdAppend,
    fdClose: fdClose
};
