// Data.Predicate/index.js
"use strict";
var Data_BooleanAlgebra = require("../Data.BooleanAlgebra/index.js");
var Data_HeytingAlgebra = require("../Data.HeytingAlgebra/index.js");

// | An adaptor allowing `>$<` to map over the inputs of a predicate.
var Predicate = function (x) {
    return x;
};
var newtypePredicate = {
    Coercible0: function () {
        return undefined;
    }
};
var heytingAlgebraPredicate = Data_HeytingAlgebra.heytingAlgebraFunction(Data_HeytingAlgebra.heytingAlgebraBoolean);
var contravariantPredicate = {
    cmap: function (f) {
        return function (v) {
            return function ($3) {
                return v(f($3));
            };
        };
    }
};
var booleanAlgebraPredicate = Data_BooleanAlgebra.booleanAlgebraFn(Data_BooleanAlgebra.booleanAlgebraBoolean);
module.exports = {
    Predicate: Predicate,
    newtypePredicate: newtypePredicate,
    heytingAlgebraPredicate: heytingAlgebraPredicate,
    booleanAlgebraPredicate: booleanAlgebraPredicate,
    contravariantPredicate: contravariantPredicate
};
