// Test.Spec.Result/index.js
"use strict";
var Data_Eq = require("../Data.Eq/index.js");
var Data_Function = require("../Data.Function/index.js");
var Data_Maybe = require("../Data.Maybe/index.js");
var Data_Show = require("../Data.Show/index.js");
var Data_Time_Duration = require("../Data.Time.Duration/index.js");
var Effect_Exception = require("../Effect.Exception/index.js");
var Test_Spec_Speed = require("../Test.Spec.Speed/index.js");
var Success = (function () {
    function Success(value0, value1) {
        this.value0 = value0;
        this.value1 = value1;
    };
    Success.create = function (value0) {
        return function (value1) {
            return new Success(value0, value1);
        };
    };
    return Success;
})();
var Failure = (function () {
    function Failure(value0) {
        this.value0 = value0;
    };
    Failure.create = function (value0) {
        return new Failure(value0);
    };
    return Failure;
})();
var showResult = {
    show: function (v) {
        if (v instanceof Success) {
            return "Success (" + (Data_Show.show(Test_Spec_Speed.showSpeed)(v.value0) + (" " + (Data_Show.show(Data_Time_Duration.showMilliseconds)(v.value1) + ")")));
        };
        if (v instanceof Failure) {
            return "Failure (Error " + (Effect_Exception.message(v.value0) + ")");
        };
        throw new Error("Failed pattern match at Test.Spec.Result (line 15, column 1 - line 17, column 69): " + [ v.constructor.name ]);
    }
};
var eqResult = {
    eq: function (v) {
        return function (v1) {
            if (v instanceof Success && v1 instanceof Success) {
                return Data_Eq.eq(Test_Spec_Speed.showEq)(v.value0)(v1.value0) && Data_Eq.eq(Data_Time_Duration.eqMilliseconds)(v.value1)(v1.value1);
            };
            if (v instanceof Failure && v1 instanceof Failure) {
                return Data_Function.on(Data_Eq.eq(Data_Eq.eqString))(Effect_Exception.message)(v.value0)(v1.value0) && Data_Function.on(Data_Eq.eq(Data_Maybe.eqMaybe(Data_Eq.eqString)))(Effect_Exception.stack)(v.value0)(v1.value0);
            };
            return false;
        };
    }
};
module.exports = {
    Success: Success,
    Failure: Failure,
    showResult: showResult,
    eqResult: eqResult
};
