// Data.Lens.Internal.Focusing/index.js

// | This module defines the `Focusing` functor
"use strict";
var Control_Applicative = require("../Control.Applicative/index.js");
var Control_Apply = require("../Control.Apply/index.js");
var Data_Functor = require("../Data.Functor/index.js");
var Data_Tuple = require("../Data.Tuple/index.js");

// | The functor used to zoom into `StateT`.
var Focusing = function (x) {
    return x;
};
var newtypeFocusing = {
    Coercible0: function () {
        return undefined;
    }
};
var focusingFunctor = function (dictFunctor) {
    return {
        map: function (f) {
            return function (v) {
                return Data_Functor.map(dictFunctor)(Data_Functor.map(Data_Tuple.functorTuple)(f))(v);
            };
        }
    };
};
var focusingApply = function (dictApply) {
    return function (dictSemigroup) {
        return {
            apply: function (v) {
                return function (v1) {
                    return Control_Apply.apply(dictApply)(Data_Functor.map(dictApply.Functor0())(Control_Apply.apply(Data_Tuple.applyTuple(dictSemigroup)))(v))(v1);
                };
            },
            Functor0: function () {
                return focusingFunctor(dictApply.Functor0());
            }
        };
    };
};
var focusingApplicative = function (dictApplicative) {
    return function (dictMonoid) {
        return {
            pure: (function () {
                var $12 = Control_Applicative.pure(dictApplicative);
                var $13 = Control_Applicative.pure(Data_Tuple.applicativeTuple(dictMonoid));
                return function ($14) {
                    return Focusing($12($13($14)));
                };
            })(),
            Apply0: function () {
                return focusingApply(dictApplicative.Apply0())(dictMonoid.Semigroup0());
            }
        };
    };
};
module.exports = {
    Focusing: Focusing,
    newtypeFocusing: newtypeFocusing,
    focusingFunctor: focusingFunctor,
    focusingApply: focusingApply,
    focusingApplicative: focusingApplicative
};
