// Purescriptify.Component.GlobalContext/index.js
"use strict";
var Control_Applicative = require("../Control.Applicative/index.js");
var Control_Bind = require("../Control.Bind/index.js");
var Data_Maybe = require("../Data.Maybe/index.js");
var Data_Unit = require("../Data.Unit/index.js");
var Foreign_Toast = require("../Foreign.Toast/index.js");
var Purescriptify_API_Error = require("../Purescriptify.API.Error/index.js");
var Purescriptify_Capability_Global = require("../Purescriptify.Capability.Global/index.js");
var Purescriptify_Capability_Halo = require("../Purescriptify.Capability.Halo/index.js");
var React_Basic_DOM_Generated = require("../React.Basic.DOM.Generated/index.js");
var React_Basic_Hooks_Internal = require("../React.Basic.Hooks.Internal/index.js");
var React_Halo_Internal_Control = require("../React.Halo.Internal.Control/index.js");
var React_Halo_Internal_Eval = require("../React.Halo.Internal.Eval/index.js");
var Type_Equality = require("../Type.Equality/index.js");
var Initialize = (function () {
    function Initialize() {

    };
    Initialize.value = new Initialize();
    return Initialize;
})();
var HandleGlobalAction = (function () {
    function HandleGlobalAction(value0) {
        this.value0 = value0;
    };
    HandleGlobalAction.create = function (value0) {
        return new HandleGlobalAction(value0);
    };
    return HandleGlobalAction;
})();
var HandleGlobalEvent = (function () {
    function HandleGlobalEvent(value0) {
        this.value0 = value0;
    };
    HandleGlobalEvent.create = function (value0) {
        return new HandleGlobalEvent(value0);
    };
    return HandleGlobalEvent;
})();
var mkGlobalContext = function (dictMonadGlobal) {
    return function (dictMonadHalo) {
        return function (dictMonadEffect) {
            var render = function (v) {
                return React_Basic_DOM_Generated.div()({});
            };
            var initialState = function (v) {
                return function (v1) {
                    return Data_Unit.unit;
                };
            };
            var handleAction = function (a) {
                if (a instanceof Initialize) {
                    return Control_Bind.discard(Control_Bind.discardUnit)(React_Halo_Internal_Control.bindHaloM)(Purescriptify_Capability_Global.subscribeForGlobalActions(dictMonadGlobal)(HandleGlobalAction.create))(function () {
                        return Purescriptify_Capability_Global.subscribeForGlobalEvents(dictMonadGlobal)(HandleGlobalEvent.create);
                    });
                };
                if (a instanceof HandleGlobalAction) {
                    return Control_Applicative.pure(React_Halo_Internal_Control.applicativeHaloM)(Data_Unit.unit);
                };
                if (a instanceof HandleGlobalEvent) {
                    return Control_Bind.discard(Control_Bind.discardUnit)(React_Halo_Internal_Control.bindHaloM)(Purescriptify_API_Error.withErrorContents(React_Halo_Internal_Control.monadHaloM)(a.value0.value0)(function (err) {
                        if (err instanceof Purescriptify_API_Error.ValidationError) {
                            return Foreign_Toast.toastError(React_Halo_Internal_Control.monadEffectHaloM(dictMonadEffect))(err.value0.error);
                        };
                        if (err instanceof Purescriptify_API_Error.UnknownError) {
                            return Foreign_Toast.toastError(React_Halo_Internal_Control.monadEffectHaloM(dictMonadEffect))("An unknown error occurred!");
                        };
                        throw new Error("Failed pattern match at Purescriptify.Component.GlobalContext (line 49, column 21 - line 51, column 72): " + [ err.constructor.name ]);
                    }))(function () {
                        return Control_Applicative.pure(React_Halo_Internal_Control.applicativeHaloM)(Data_Unit.unit);
                    });
                };
                throw new Error("Failed pattern match at Purescriptify.Component.GlobalContext (line 40, column 20 - line 52, column 18): " + [ a.constructor.name ]);
            };
            var $$eval = React_Halo_Internal_Eval.mkEval(function (v) {
                return {
                    onInitialize: function (v1) {
                        return new Data_Maybe.Just(Initialize.value);
                    },
                    onUpdate: v.onUpdate,
                    onAction: handleAction,
                    onFinalize: v.onFinalize
                };
            });
            var context = function (v) {
                return Control_Applicative.pure(React_Basic_Hooks_Internal.applicativeRender(Type_Equality.refl))(Data_Unit.unit);
            };
            return Purescriptify_Capability_Halo.component(dictMonadHalo)("GlobalContext")({
                context: context,
                initialState: initialState,
                "eval": $$eval,
                render: render
            });
        };
    };
};
module.exports = {
    Initialize: Initialize,
    HandleGlobalAction: HandleGlobalAction,
    HandleGlobalEvent: HandleGlobalEvent,
    mkGlobalContext: mkGlobalContext
};
