// Node.Buffer.Immutable/index.js

// | Immutable buffers and associated operations.
"use strict";
var $foreign = require("./foreign.js");
var Data_Maybe = require("../Data.Maybe/index.js");
var Data_Ordering = require("../Data.Ordering/index.js");
var Data_Show = require("../Data.Show/index.js");
var Node_Buffer_Types = require("../Node.Buffer.Types/index.js");
var Node_Encoding = require("../Node.Encoding/index.js");

// | Reads the buffer as a string with the specified encoding.
var toString = function ($3) {
    return $foreign.toStringImpl(Node_Encoding.encodingToNode($3));
};
var showBuffer = {
    show: $foreign.showImpl
};

// | Reads a section of a buffer as a string with the specified encoding.
var readString = function ($4) {
    return $foreign.readStringImpl(Node_Encoding.encodingToNode($4));
};

// | Reads a numeric value from a buffer at the specified offset.
var read = (function () {
    var $5 = Data_Show.show(Node_Buffer_Types.showBufferValueType);
    return function ($6) {
        return $foreign.readImpl($5($6));
    };
})();

// | Reads an octet from a buffer at the specified offset.
var getAtOffset = $foreign.getAtOffsetImpl(Data_Maybe.Just.create)(Data_Maybe.Nothing.value);

// | Creates a new buffer from a string with the specified encoding, sized to match the string.
var fromString = function (str) {
    var $7 = $foreign.fromStringImpl(str);
    return function ($8) {
        return $7(Node_Encoding.encodingToNode($8));
    };
};
var eqBuffer = {
    eq: $foreign.eqImpl
};
var ordBuffer = {
    compare: function (a) {
        return function (b) {
            var v = $foreign.compareImpl(a)(b);
            if (v < 0) {
                return Data_Ordering.LT.value;
            };
            if (v > 0) {
                return Data_Ordering.GT.value;
            };
            return Data_Ordering.EQ.value;
        };
    },
    Eq0: function () {
        return eqBuffer;
    }
};

// | Concatenates a list of buffers, combining them into a new buffer of the
// | specified length.
var concat$prime = $foreign.concatToLength;
module.exports = {
    fromString: fromString,
    read: read,
    readString: readString,
    toString: toString,
    getAtOffset: getAtOffset,
    "concat'": concat$prime,
    showBuffer: showBuffer,
    eqBuffer: eqBuffer,
    ordBuffer: ordBuffer,
    create: $foreign.create,
    fromArray: $foreign.fromArray,
    fromArrayBuffer: $foreign.fromArrayBuffer,
    toArray: $foreign.toArray,
    toArrayBuffer: $foreign.toArrayBuffer,
    concat: $foreign.concat,
    slice: $foreign.slice,
    size: $foreign.size
};
