// Dodo.Common/index.js
"use strict";
var Data_Semigroup = require("../Data.Semigroup/index.js");
var Dodo = require("../Dodo/index.js");
var Dodo_Internal = require("../Dodo.Internal/index.js");
var trailingComma = Dodo.flexAlt(Dodo.text(", "))(Data_Semigroup.append(Dodo_Internal.semigroupDoc)(Dodo.text(","))(Dodo["break"]));
var pursSquares = (function () {
    var open = Dodo.text("[ ");
    var close = Dodo.flexAlt(Dodo.text(" ]"))(Data_Semigroup.append(Dodo_Internal.semigroupDoc)(Dodo["break"])(Dodo.text("]")));
    var $0 = Dodo.encloseEmptyAlt(open)(close)(Dodo.text("[]"));
    return function ($1) {
        return Dodo.flexGroup($0($1));
    };
})();
var pursParensExpr = (function () {
    var open = Dodo.text("(");
    var close = Dodo.text(")");
    var $2 = Dodo.enclose(open)(close);
    return function ($3) {
        return Dodo.flexGroup($2(Dodo.indent($3)));
    };
})();
var pursParens = (function () {
    var open = Dodo.flexAlt(Dodo.text("("))(Dodo.text("( "));
    var close = Dodo.flexAlt(Dodo.text(")"))(Data_Semigroup.append(Dodo_Internal.semigroupDoc)(Dodo["break"])(Dodo.text(")")));
    var $4 = Dodo.encloseEmptyAlt(open)(close)(Dodo.text("()"));
    return function ($5) {
        return Dodo.flexGroup($4($5));
    };
})();
var pursCurlies = (function () {
    var open = Dodo.text("{ ");
    var close = Dodo.flexAlt(Dodo.text(" }"))(Data_Semigroup.append(Dodo_Internal.semigroupDoc)(Dodo["break"])(Dodo.text("}")));
    var $6 = Dodo.encloseEmptyAlt(open)(close)(Dodo.text("{}"));
    return function ($7) {
        return Dodo.flexGroup($6($7));
    };
})();
var leadingComma = Dodo.flexAlt(Dodo.text(", "))(Data_Semigroup.append(Dodo_Internal.semigroupDoc)(Dodo["break"])(Dodo.text(", ")));
var jsSquares = (function () {
    var open = Dodo.flexAlt(Dodo.text("["))(Data_Semigroup.append(Dodo_Internal.semigroupDoc)(Dodo.text("["))(Dodo["break"]));
    var close = Dodo.flexAlt(Dodo.text("]"))(Data_Semigroup.append(Dodo_Internal.semigroupDoc)(Dodo["break"])(Dodo.text("]")));
    var $8 = Dodo.encloseEmptyAlt(open)(close)(Dodo.text("{}"));
    return function ($9) {
        return Dodo.flexGroup($8(Dodo.indent($9)));
    };
})();
var jsParens = (function () {
    var open = Dodo.flexAlt(Dodo.text("("))(Data_Semigroup.append(Dodo_Internal.semigroupDoc)(Dodo.text("("))(Dodo["break"]));
    var close = Dodo.flexAlt(Dodo.text(")"))(Data_Semigroup.append(Dodo_Internal.semigroupDoc)(Dodo["break"])(Dodo.text(")")));
    var $10 = Dodo.encloseEmptyAlt(open)(close)(Dodo.text("()"));
    return function ($11) {
        return Dodo.flexGroup($10(Dodo.indent($11)));
    };
})();
var jsCurlies = (function () {
    var open = Dodo.flexAlt(Dodo.text("{"))(Data_Semigroup.append(Dodo_Internal.semigroupDoc)(Dodo.text("{"))(Dodo["break"]));
    var close = Dodo.flexAlt(Dodo.text("}"))(Data_Semigroup.append(Dodo_Internal.semigroupDoc)(Dodo["break"])(Dodo.text("}")));
    var $12 = Dodo.encloseEmptyAlt(open)(close)(Dodo.text("{}"));
    return function ($13) {
        return Dodo.flexGroup($12(Dodo.indent($13)));
    };
})();
module.exports = {
    leadingComma: leadingComma,
    trailingComma: trailingComma,
    pursCurlies: pursCurlies,
    pursSquares: pursSquares,
    pursParens: pursParens,
    pursParensExpr: pursParensExpr,
    jsCurlies: jsCurlies,
    jsSquares: jsSquares,
    jsParens: jsParens
};
