// React.Halo.Internal.Types/index.js
"use strict";
var Data_Eq = require("../Data.Eq/index.js");
var Data_Ord = require("../Data.Ord/index.js");
var SubscriptionId = function (x) {
    return x;
};

// | The Halo lifecycle events.
// |
// | - `Initialize` contains the initial props. It occurs when the component mounts, and only once per component.
// | - `Update` contains the previous and new props. It occurs when the component re-renders and the props have changes.
// | - `Action` contains the dispatched action. It occurs each time an action is dispatched to be eval'd, up until the
// |   `Finalize` event.
// | - `Finalize` occurs when the component unmounts.
var Initialize = (function () {
    function Initialize(value0) {
        this.value0 = value0;
    };
    Initialize.create = function (value0) {
        return new Initialize(value0);
    };
    return Initialize;
})();

// | The Halo lifecycle events.
// |
// | - `Initialize` contains the initial props. It occurs when the component mounts, and only once per component.
// | - `Update` contains the previous and new props. It occurs when the component re-renders and the props have changes.
// | - `Action` contains the dispatched action. It occurs each time an action is dispatched to be eval'd, up until the
// |   `Finalize` event.
// | - `Finalize` occurs when the component unmounts.
var Update = (function () {
    function Update(value0, value1) {
        this.value0 = value0;
        this.value1 = value1;
    };
    Update.create = function (value0) {
        return function (value1) {
            return new Update(value0, value1);
        };
    };
    return Update;
})();

// | The Halo lifecycle events.
// |
// | - `Initialize` contains the initial props. It occurs when the component mounts, and only once per component.
// | - `Update` contains the previous and new props. It occurs when the component re-renders and the props have changes.
// | - `Action` contains the dispatched action. It occurs each time an action is dispatched to be eval'd, up until the
// |   `Finalize` event.
// | - `Finalize` occurs when the component unmounts.
var Action = (function () {
    function Action(value0) {
        this.value0 = value0;
    };
    Action.create = function (value0) {
        return new Action(value0);
    };
    return Action;
})();

// | The Halo lifecycle events.
// |
// | - `Initialize` contains the initial props. It occurs when the component mounts, and only once per component.
// | - `Update` contains the previous and new props. It occurs when the component re-renders and the props have changes.
// | - `Action` contains the dispatched action. It occurs each time an action is dispatched to be eval'd, up until the
// |   `Finalize` event.
// | - `Finalize` occurs when the component unmounts.
var Finalize = (function () {
    function Finalize() {

    };
    Finalize.value = new Finalize();
    return Finalize;
})();
var ForkId = function (x) {
    return x;
};
var ordSubscriptionId = Data_Ord.ordInt;
var ordForkId = Data_Ord.ordInt;
var eqSubscriptionId = Data_Eq.eqInt;
var eqForkId = Data_Eq.eqInt;
module.exports = {
    Initialize: Initialize,
    Update: Update,
    Action: Action,
    Finalize: Finalize,
    SubscriptionId: SubscriptionId,
    ForkId: ForkId,
    eqSubscriptionId: eqSubscriptionId,
    ordSubscriptionId: ordSubscriptionId,
    eqForkId: eqForkId,
    ordForkId: ordForkId
};
