// Data.Posix/index.js
"use strict";
var Data_Eq = require("../Data.Eq/index.js");
var Data_Ord = require("../Data.Ord/index.js");
var Data_Show = require("../Data.Show/index.js");

// | A user ID (for a process or a file).
var Uid = function (x) {
    return x;
};

// | A process ID.
var Pid = function (x) {
    return x;
};

// | A group ID (for a process or a file).
var Gid = function (x) {
    return x;
};
var showUid = {
    show: function (v) {
        return "(Uid " + (Data_Show.show(Data_Show.showInt)(v) + ")");
    }
};
var showPid = {
    show: function (v) {
        return "(Pid " + (Data_Show.show(Data_Show.showInt)(v) + ")");
    }
};
var showGid = {
    show: function (v) {
        return "(Gid " + (Data_Show.show(Data_Show.showInt)(v) + ")");
    }
};
var ordUid = Data_Ord.ordInt;
var ordPid = Data_Ord.ordInt;
var ordGid = Data_Ord.ordInt;
var newtypeUid = {
    Coercible0: function () {
        return undefined;
    }
};
var newtypePid = {
    Coercible0: function () {
        return undefined;
    }
};
var newtypeGid = {
    Coercible0: function () {
        return undefined;
    }
};
var eqUid = Data_Eq.eqInt;
var eqPid = Data_Eq.eqInt;
var eqGid = Data_Eq.eqInt;
module.exports = {
    Pid: Pid,
    Gid: Gid,
    Uid: Uid,
    newtypePid: newtypePid,
    eqPid: eqPid,
    ordPid: ordPid,
    showPid: showPid,
    newtypeGid: newtypeGid,
    eqGid: eqGid,
    ordGid: ordGid,
    showGid: showGid,
    newtypeUid: newtypeUid,
    eqUid: eqUid,
    ordUid: ordUid,
    showUid: showUid
};
