// ServerAPI/index.js

// File auto generated by servant-purescript! --
"use strict";
var Affjax_RequestHeader = require("../Affjax.RequestHeader/index.js");
var Data_Argonaut_Decode_Aeson = require("../Data.Argonaut.Decode.Aeson/index.js");
var Data_Argonaut_Decode_Class = require("../Data.Argonaut.Decode.Class/index.js");
var Data_Argonaut_Encode_Aeson = require("../Data.Argonaut.Encode.Aeson/index.js");
var Data_Array = require("../Data.Array/index.js");
var Data_Either = require("../Data.Either/index.js");
var Data_Functor = require("../Data.Functor/index.js");
var Data_HTTP_Method = require("../Data.HTTP.Method/index.js");
var Data_Maybe = require("../Data.Maybe/index.js");
var Purescriptify_API_Auth = require("../Purescriptify.API.Auth/index.js");
var Purescriptify_API_Types = require("../Purescriptify.API.Types/index.js");
var Servant_PureScript = require("../Servant.PureScript/index.js");
var URI_RelativePart = require("../URI.RelativePart/index.js");
var URI_RelativeRef = require("../URI.RelativeRef/index.js");
var Api = (function () {
    function Api() {

    };
    Api.value = new Api();
    return Api;
})();
var putUserByUserId = function (dictMonadAjax) {
    return function (reqBody) {
        return function (userId) {
            var relativePart = new URI_RelativePart.RelativePartNoAuth(new Data_Maybe.Just([ "user", Servant_PureScript.toPathSegment(Purescriptify_API_Types.toPathSegmentUserId)(userId) ]));
            var uri = new URI_RelativeRef.RelativeRef(relativePart, Data_Maybe.Nothing.value, Data_Maybe.Nothing.value);
            var method = new Data_Either.Left(Data_HTTP_Method.PUT.value);
            var headers = Data_Array.catMaybes([  ]);
            var encode = Data_Argonaut_Encode_Aeson.encode(Data_Argonaut_Encode_Aeson.value(Purescriptify_API_Types.encodeJsonUpdateUserReque));
            var decode = Data_Argonaut_Decode_Aeson.decode(Data_Argonaut_Decode_Aeson.unit);
            var content = new Data_Maybe.Just(reqBody);
            var req = {
                method: method,
                uri: uri,
                headers: headers,
                content: content,
                encode: encode,
                decode: decode
            };
            return Servant_PureScript.request(dictMonadAjax)(Servant_PureScript.contentTypeJson)(Servant_PureScript.contentTypeJson)(Api.value)(req);
        };
    };
};
var postUser = function (dictMonadAjax) {
    return function (reqBody) {
        var relativePart = new URI_RelativePart.RelativePartNoAuth(new Data_Maybe.Just([ "user" ]));
        var uri = new URI_RelativeRef.RelativeRef(relativePart, Data_Maybe.Nothing.value, Data_Maybe.Nothing.value);
        var method = new Data_Either.Left(Data_HTTP_Method.POST.value);
        var headers = Data_Array.catMaybes([  ]);
        var encode = Data_Argonaut_Encode_Aeson.encode(Data_Argonaut_Encode_Aeson.value(Purescriptify_API_Types.encodeJsonCreateUserReque));
        var decode = Data_Argonaut_Decode_Aeson.decode(Data_Argonaut_Decode_Aeson.value(Purescriptify_API_Types.decodeJsonUser));
        var content = new Data_Maybe.Just(reqBody);
        var req = {
            method: method,
            uri: uri,
            headers: headers,
            content: content,
            encode: encode,
            decode: decode
        };
        return Servant_PureScript.request(dictMonadAjax)(Servant_PureScript.contentTypeJson)(Servant_PureScript.contentTypeJson)(Api.value)(req);
    };
};
var getUsers = function (dictMonadAjax) {
    var relativePart = new URI_RelativePart.RelativePartNoAuth(new Data_Maybe.Just([ "users" ]));
    var uri = new URI_RelativeRef.RelativeRef(relativePart, Data_Maybe.Nothing.value, Data_Maybe.Nothing.value);
    var method = new Data_Either.Left(Data_HTTP_Method.GET.value);
    var headers = Data_Array.catMaybes([  ]);
    var encode = Data_Argonaut_Encode_Aeson.encode(Data_Argonaut_Encode_Aeson["null"]);
    var decode = Data_Argonaut_Decode_Aeson.decode(Data_Argonaut_Decode_Aeson.value(Data_Argonaut_Decode_Class.decodeArray(Purescriptify_API_Types.decodeJsonUser)));
    var req = {
        method: method,
        uri: uri,
        headers: headers,
        content: Data_Maybe.Nothing.value,
        encode: encode,
        decode: decode
    };
    return Servant_PureScript.request(dictMonadAjax)(Servant_PureScript.contentTypeJson)(Servant_PureScript.contentTypeJson)(Api.value)(req);
};
var getUserByUserId = function (dictMonadAjax) {
    return function (userId) {
        var relativePart = new URI_RelativePart.RelativePartNoAuth(new Data_Maybe.Just([ "user", Servant_PureScript.toPathSegment(Purescriptify_API_Types.toPathSegmentUserId)(userId) ]));
        var uri = new URI_RelativeRef.RelativeRef(relativePart, Data_Maybe.Nothing.value, Data_Maybe.Nothing.value);
        var method = new Data_Either.Left(Data_HTTP_Method.GET.value);
        var headers = Data_Array.catMaybes([  ]);
        var encode = Data_Argonaut_Encode_Aeson.encode(Data_Argonaut_Encode_Aeson["null"]);
        var decode = Data_Argonaut_Decode_Aeson.decode(Data_Argonaut_Decode_Aeson.value(Purescriptify_API_Types.decodeJsonUser));
        var req = {
            method: method,
            uri: uri,
            headers: headers,
            content: Data_Maybe.Nothing.value,
            encode: encode,
            decode: decode
        };
        return Servant_PureScript.request(dictMonadAjax)(Servant_PureScript.contentTypeJson)(Servant_PureScript.contentTypeJson)(Api.value)(req);
    };
};
var deleteUserByUserId = function (dictMonadAjax) {
    return function (authorization) {
        return function (userId) {
            var relativePart = new URI_RelativePart.RelativePartNoAuth(new Data_Maybe.Just([ "user", Servant_PureScript.toPathSegment(Purescriptify_API_Types.toPathSegmentUserId)(userId) ]));
            var uri = new URI_RelativeRef.RelativeRef(relativePart, Data_Maybe.Nothing.value, Data_Maybe.Nothing.value);
            var method = new Data_Either.Left(Data_HTTP_Method.DELETE.value);
            var headers = Data_Array.catMaybes([ Data_Functor.map(Data_Maybe.functorMaybe)((function () {
                var $5 = Affjax_RequestHeader.RequestHeader.create("Authorization");
                var $6 = Servant_PureScript.toHeader(Purescriptify_API_Auth.toHeaderAuthorizationHead);
                return function ($7) {
                    return $5($6($7));
                };
            })())(authorization) ]);
            var encode = Data_Argonaut_Encode_Aeson.encode(Data_Argonaut_Encode_Aeson["null"]);
            var decode = Data_Argonaut_Decode_Aeson.decode(Data_Argonaut_Decode_Aeson.unit);
            var req = {
                method: method,
                uri: uri,
                headers: headers,
                content: Data_Maybe.Nothing.value,
                encode: encode,
                decode: decode
            };
            return Servant_PureScript.request(dictMonadAjax)(Servant_PureScript.contentTypeJson)(Servant_PureScript.contentTypeJson)(Api.value)(req);
        };
    };
};
module.exports = {
    Api: Api,
    getUsers: getUsers,
    getUserByUserId: getUserByUserId,
    postUser: postUser,
    putUserByUserId: putUserByUserId,
    deleteUserByUserId: deleteUserByUserId
};
