// Ansi.Output/index.js

// | Convenience functions to simplify outputting ANSI escape codes to
// | terminals.
"use strict";
var Ansi_Codes = require("../Ansi.Codes/index.js");
var Control_Applicative = require("../Control.Applicative/index.js");
var Data_List_Types = require("../Data.List.Types/index.js");

// | Wrap the given text in escape codes corresponding to the given parameters.
// | For example:
// |
// | ```purescript
// | Console.log $ withGraphics (bold <> underline <> foreground BrightRed) "hello world"
// | ```
// |
// | would print "hello world" to the terminal, bold, underlined, and in bright
// | red, and then reset (so that further logging to the console uses the
// | normal color and style).
// |
// | This function simply wraps the given text in an escape code and a reset
// | code, so that it is a little more comfortable to use than the functions
// | in `Ansi.Codes`.
var withGraphics = function (params) {
    return function (text) {
        return Ansi_Codes.escapeCodeToString(new Ansi_Codes.Graphics(params)) + (text + Ansi_Codes.escapeCodeToString(new Ansi_Codes.Graphics(Control_Applicative.pure(Data_List_Types.applicativeNonEmptyList)(Ansi_Codes.Reset.value))));
    };
};
var underline = Control_Applicative.pure(Data_List_Types.applicativeNonEmptyList)(new Ansi_Codes.PMode(Ansi_Codes.Underline.value));
var strikethrough = Control_Applicative.pure(Data_List_Types.applicativeNonEmptyList)(new Ansi_Codes.PMode(Ansi_Codes.Strikethrough.value));
var italic = Control_Applicative.pure(Data_List_Types.applicativeNonEmptyList)(new Ansi_Codes.PMode(Ansi_Codes.Italic.value));
var inverse = Control_Applicative.pure(Data_List_Types.applicativeNonEmptyList)(new Ansi_Codes.PMode(Ansi_Codes.Inverse.value));
var foreground = function (c) {
    return Control_Applicative.pure(Data_List_Types.applicativeNonEmptyList)(new Ansi_Codes.PForeground(c));
};
var dim = Control_Applicative.pure(Data_List_Types.applicativeNonEmptyList)(new Ansi_Codes.PMode(Ansi_Codes.Dim.value));
var bold = Control_Applicative.pure(Data_List_Types.applicativeNonEmptyList)(new Ansi_Codes.PMode(Ansi_Codes.Bold.value));
var background = function (c) {
    return Control_Applicative.pure(Data_List_Types.applicativeNonEmptyList)(new Ansi_Codes.PBackground(c));
};
module.exports = {
    withGraphics: withGraphics,
    bold: bold,
    dim: dim,
    italic: italic,
    underline: underline,
    inverse: inverse,
    strikethrough: strikethrough,
    foreground: foreground,
    background: background
};
