// URI.Scheme.Common/index.js

// | Common URI schemes, taken from the list of permanent URI schemes
// | [assigned by IANA](https://www.iana.org/assignments/uri-schemes/uri-schemes.xhtml).
"use strict";
var URI_Scheme = require("../URI.Scheme/index.js");

// | Z39.50 Session ([RFC2056](https://tools.ietf.org/html/rfc2056))
var z3950s = URI_Scheme.unsafeFromString("z39.50s");

// | Z39.50 Retrieval ([RFC2056](https://tools.ietf.org/html/rfc2056))
var z3950r = URI_Scheme.unsafeFromString("z39.50r");

// | Extensible Messaging and Presence ([RFC5122](https://tools.ietf.org/html/rfc5122))
var xmpp = URI_Scheme.unsafeFromString("xmpp");

// | xmlrpc.beeps ([RFC3529](https://tools.ietf.org/html/rfc3529))
var xmlrpcbeeps = URI_Scheme.unsafeFromString("xmlrpc.beeps");

// | xmlrpc.beep ([RFC3529](https://tools.ietf.org/html/rfc3529))
var xmlrpcbeep = URI_Scheme.unsafeFromString("xmlrpc.beep");

// | xcon-userid ([RFC6501](https://tools.ietf.org/html/rfc6501))
var xconuserid = URI_Scheme.unsafeFromString("xcon-userid");

// | xcon ([RFC6501](https://tools.ietf.org/html/rfc6501))
var xcon = URI_Scheme.unsafeFromString("xcon");

// | Encrypted WebSocket connections ([RFC6455](https://tools.ietf.org/html/rfc6455))
var wss = URI_Scheme.unsafeFromString("wss");

// | WebSocket connections ([RFC6455](https://tools.ietf.org/html/rfc6455))
var ws = URI_Scheme.unsafeFromString("ws");

// | Remote Framebuffer Protocol ([RFC7869](https://tools.ietf.org/html/rfc7869))
var vnc = URI_Scheme.unsafeFromString("vnc");

// | versatile multimedia interface ([RFC2122](https://tools.ietf.org/html/rfc2122))
var vemmi = URI_Scheme.unsafeFromString("vemmi");

// | Uniform Resource Names ([RFC8141](https://tools.ietf.org/html/rfc8141))
var urn = URI_Scheme.unsafeFromString("urn");

// | TV Broadcasts ([RFC2838](https://tools.ietf.org/html/rfc2838))
var tv = URI_Scheme.unsafeFromString("tv");

// | turns ([RFC7065](https://tools.ietf.org/html/rfc7065))
var turns = URI_Scheme.unsafeFromString("turns");

// | turn ([RFC7065](https://tools.ietf.org/html/rfc7065))
var turn = URI_Scheme.unsafeFromString("turn");

// | Interactive 3270 emulation sessions ([RFC6270](https://tools.ietf.org/html/rfc6270))
var tn3270 = URI_Scheme.unsafeFromString("tn3270");

// | Transaction Internet Protocol ([RFC2371](https://tools.ietf.org/html/rfc2371))
var tip = URI_Scheme.unsafeFromString("tip");

// | multipart/related relative reference ([RFC2557](https://tools.ietf.org/html/rfc2557))
var thismessage = URI_Scheme.unsafeFromString("thismessage");

// | Trivial File Transfer Protocol ([RFC3617](https://tools.ietf.org/html/rfc3617))
var tftp = URI_Scheme.unsafeFromString("tftp");

// | Reference to interactive sessions ([RFC4248](https://tools.ietf.org/html/rfc4248))
var telnet = URI_Scheme.unsafeFromString("telnet");

// | telephone ([RFC3966](https://tools.ietf.org/html/rfc3966))
var tel = URI_Scheme.unsafeFromString("tel");

// | tag ([RFC4151](https://tools.ietf.org/html/rfc4151))
var tag = URI_Scheme.unsafeFromString("tag");

// | stuns ([RFC7064](https://tools.ietf.org/html/rfc7064))
var stuns = URI_Scheme.unsafeFromString("stuns");

// | stun ([RFC7064](https://tools.ietf.org/html/rfc7064))
var stun = URI_Scheme.unsafeFromString("stun");

// | soap.beeps ([RFC4227](https://tools.ietf.org/html/rfc4227))
var soapbeeps = URI_Scheme.unsafeFromString("soap.beeps");

// | soap.beep ([RFC4227](https://tools.ietf.org/html/rfc4227))
var soapbeep = URI_Scheme.unsafeFromString("soap.beep");

// | Simple Network Management Protocol ([RFC4088](https://tools.ietf.org/html/rfc4088))
var snmp = URI_Scheme.unsafeFromString("snmp");

// | Short Message Service ([RFC5724](https://tools.ietf.org/html/rfc5724))
var sms = URI_Scheme.unsafeFromString("sms");

// | secure session initiation protocol ([RFC3261](https://tools.ietf.org/html/rfc3261))
var sips = URI_Scheme.unsafeFromString("sips");

// | session initiation protocol ([RFC3261](https://tools.ietf.org/html/rfc3261))
var sip = URI_Scheme.unsafeFromString("sip");

// | ManageSieve Protocol ([RFC5804](https://tools.ietf.org/html/rfc5804))
var sieve = URI_Scheme.unsafeFromString("sieve");

// | Secure Hypertext Transfer Protocol ([RFC2660](https://tools.ietf.org/html/rfc2660))
var shttp = URI_Scheme.unsafeFromString("shttp");

// | session ([RFC6787](https://tools.ietf.org/html/rfc6787))
var session = URI_Scheme.unsafeFromString("session");

// | service location ([RFC2609](https://tools.ietf.org/html/rfc2609))
var service = URI_Scheme.unsafeFromString("service");

// | Real-Time Streaming Protocol (RTSP) ([RFC2326](https://tools.ietf.org/html/rfc2326))
var rtspu = URI_Scheme.unsafeFromString("rtspu");

// | Real-Time Streaming Protocol (RTSP) ([RFC2326](https://tools.ietf.org/html/rfc2326), [RFC7826](https://tools.ietf.org/html/rfc7826))
var rtsps = URI_Scheme.unsafeFromString("rtsps");

// | Real-Time Streaming Protocol (RTSP) ([RFC2326](https://tools.ietf.org/html/rfc2326), [RFC7826](https://tools.ietf.org/html/rfc7826))
var rtsp = URI_Scheme.unsafeFromString("rtsp");

// | reload ([RFC6940](https://tools.ietf.org/html/rfc6940))
var reload = URI_Scheme.unsafeFromString("reload");

// | Presence ([RFC3859](https://tools.ietf.org/html/rfc3859))
var pres = URI_Scheme.unsafeFromString("pres");

// | Post Office Protocol v3 ([RFC2384](https://tools.ietf.org/html/rfc2384))
var pop = URI_Scheme.unsafeFromString("pop");

// | PKCS#11 ([RFC7512](https://tools.ietf.org/html/rfc7512))
var pkcs11 = URI_Scheme.unsafeFromString("pkcs11");

// | opaquelocktokent ([RFC4918](https://tools.ietf.org/html/rfc4918))
var opaquelocktoken = URI_Scheme.unsafeFromString("opaquelocktoken");

// | USENET news using NNTP access ([RFC5538](https://tools.ietf.org/html/rfc5538))
var nntp = URI_Scheme.unsafeFromString("nntp");

// | nih ([RFC6920](https://tools.ietf.org/html/rfc6920))
var nih = URI_Scheme.unsafeFromString("nih");

// | ni ([RFC6920](https://tools.ietf.org/html/rfc6920))
var ni = URI_Scheme.unsafeFromString("ni");

// | network file system protocol ([RFC2224](https://tools.ietf.org/html/rfc2224))
var nfs = URI_Scheme.unsafeFromString("nfs");

// | USENET news ([RFC5538](https://tools.ietf.org/html/rfc5538))
var news = URI_Scheme.unsafeFromString("news");

// | Mailbox Update (MUPDATE) Protocol ([RFC3656](https://tools.ietf.org/html/rfc3656))
var mupdate = URI_Scheme.unsafeFromString("mupdate");

// | Message Tracking Query Protocol ([RFC3887](https://tools.ietf.org/html/rfc3887))
var mtqp = URI_Scheme.unsafeFromString("mtqp");

// | Message Session Relay Protocol ([RFC4975](https://tools.ietf.org/html/rfc4975))
var msrps = URI_Scheme.unsafeFromString("msrps");

// | Message Session Relay Protocol ([RFC4975](https://tools.ietf.org/html/rfc4975))
var msrp = URI_Scheme.unsafeFromString("msrp");

// | message identifier ([RFC2392](https://tools.ietf.org/html/rfc2392))
var mid = URI_Scheme.unsafeFromString("mid");

// | Electronic mail address ([RFC6068](https://tools.ietf.org/html/rfc6068))
var mailto = URI_Scheme.unsafeFromString("mailto");

// | Lightweight Directory Access ([RFC4516](https://tools.ietf.org/html/rfc4516))
var ldap = URI_Scheme.unsafeFromString("ldap");

// | iris.xpcs ([RFC4992](https://tools.ietf.org/html/rfc4992))
var irisxpcs = URI_Scheme.unsafeFromString("iris.xpcs");

// | iris.xpc ([RFC4992](https://tools.ietf.org/html/rfc4992))
var irisxpc = URI_Scheme.unsafeFromString("iris.xpc");

// | iris.lwz ([RFC4993](https://tools.ietf.org/html/rfc4993))
var irislwz = URI_Scheme.unsafeFromString("iris.lwz");

// | iris.beep ([RFC3983](https://tools.ietf.org/html/rfc3983))
var irisbeep = URI_Scheme.unsafeFromString("iris.beep");

// | Internet Registry Information ([RFC3981](https://tools.ietf.org/html/rfc3981))
var iris = URI_Scheme.unsafeFromString("iris");

// | Internet Printing Protocol over ([RFC7472](https://tools.ietf.org/html/rfc7472))
var ipps = URI_Scheme.unsafeFromString("ipps");

// | Internet Printing Protocol ([RFC3510](https://tools.ietf.org/html/rfc3510))
var ipp = URI_Scheme.unsafeFromString("ipp");

// | registry of public namespaces, which ([RFC4452](https://tools.ietf.org/html/rfc4452))
var info = URI_Scheme.unsafeFromString("info");

// | internet message access protocol ([RFC5092](https://tools.ietf.org/html/rfc5092))
var imap = URI_Scheme.unsafeFromString("imap");

// | Instant Messaging ([RFC3860](https://tools.ietf.org/html/rfc3860))
var im = URI_Scheme.unsafeFromString("im");

// | Internet Content Adaptation Protocol ([RFC3507](https://tools.ietf.org/html/rfc3507))
var icap = URI_Scheme.unsafeFromString("icap");

// | Inter-Asterisk eXchange Version 2 ([RFC5456](https://tools.ietf.org/html/rfc5456))
var iax = URI_Scheme.unsafeFromString("iax");

// | Hypertext Transfer Protocol Secure ([RFC7230, Section 2.7.2](https://tools.ietf.org/html/rfc7230#section-2.7.2))
var https = URI_Scheme.unsafeFromString("https");

// | Hypertext Transfer Protocol ([RFC7230, Section 2.7.1](https://tools.ietf.org/html/rfc7230#section-2.7.1))
var http = URI_Scheme.unsafeFromString("http");

// | H.323 ([RFC3508](https://tools.ietf.org/html/rfc3508))
var h323 = URI_Scheme.unsafeFromString("h323");

// | The Gopher Protocol ([RFC4266](https://tools.ietf.org/html/rfc4266))
var gopher = URI_Scheme.unsafeFromString("gopher");

// | go ([RFC3368](https://tools.ietf.org/html/rfc3368))
var go = URI_Scheme.unsafeFromString("go");

// | Geographic Locations ([RFC5870](https://tools.ietf.org/html/rfc5870))
var geo = URI_Scheme.unsafeFromString("geo");

// | File Transfer Protocol ([RFC1738](https://tools.ietf.org/html/rfc1738))
var ftp = URI_Scheme.unsafeFromString("ftp");

// | Host-specific file names ([RFC8089](https://tools.ietf.org/html/rfc8089))
var file = URI_Scheme.unsafeFromString("file");

// | example ([RFC7595](https://tools.ietf.org/html/rfc7595))
var example = URI_Scheme.unsafeFromString("example");

// | Domain Name System ([RFC4501](https://tools.ietf.org/html/rfc4501))
var dns = URI_Scheme.unsafeFromString("dns");

// | dictionary service protocol ([RFC2229](https://tools.ietf.org/html/rfc2229))
var dict = URI_Scheme.unsafeFromString("dict");

// | dav ([RFC4918](https://tools.ietf.org/html/rfc4918))
var dav = URI_Scheme.unsafeFromString("dav");

// | data ([RFC2397](https://tools.ietf.org/html/rfc2397))
var data_ = URI_Scheme.unsafeFromString("data");

// | TV-Anytime Content Reference ([RFC4078](https://tools.ietf.org/html/rfc4078))
var crid = URI_Scheme.unsafeFromString("crid");

// | coap+ws ([RFC8323](https://tools.ietf.org/html/rfc8323))
var coapws = URI_Scheme.unsafeFromString("coap+ws");

// | coap+tcp ([RFC8323](https://tools.ietf.org/html/rfc8323))
var coaptcp = URI_Scheme.unsafeFromString("coap+tcp");

// | coaps+ws ([RFC8323](https://tools.ietf.org/html/rfc8323))
var coapsws = URI_Scheme.unsafeFromString("coaps+ws");

// | coaps+tcp ([RFC8323](https://tools.ietf.org/html/rfc8323))
var coapstcp = URI_Scheme.unsafeFromString("coaps+tcp");

// | coaps ([RFC7252](https://tools.ietf.org/html/rfc7252))
var coaps = URI_Scheme.unsafeFromString("coaps");

// | coap ([RFC7252](https://tools.ietf.org/html/rfc7252))
var coap = URI_Scheme.unsafeFromString("coap");

// | content identifier ([RFC2392](https://tools.ietf.org/html/rfc2392))
var cid = URI_Scheme.unsafeFromString("cid");

// | Calendar Access Protocol ([RFC4324](https://tools.ietf.org/html/rfc4324))
var cap = URI_Scheme.unsafeFromString("cap");

// | acct ([RFC7565](https://tools.ietf.org/html/rfc7565))
var acct = URI_Scheme.unsafeFromString("acct");

// | application configuration access ([RFC2244](https://tools.ietf.org/html/rfc2244))
var acap = URI_Scheme.unsafeFromString("acap");

// | about ([RFC6694](https://tools.ietf.org/html/rfc6694))
var about = URI_Scheme.unsafeFromString("about");

// | Diameter Protocol with Secure ([RFC6733](https://tools.ietf.org/html/rfc6733))
var aaas = URI_Scheme.unsafeFromString("aaas");

// | Diameter Protocol ([RFC6733](https://tools.ietf.org/html/rfc6733))
var aaa = URI_Scheme.unsafeFromString("aaa");
module.exports = {
    aaa: aaa,
    aaas: aaas,
    about: about,
    acap: acap,
    acct: acct,
    cap: cap,
    cid: cid,
    coap: coap,
    coaptcp: coaptcp,
    coapws: coapws,
    coaps: coaps,
    coapstcp: coapstcp,
    coapsws: coapsws,
    crid: crid,
    data_: data_,
    dav: dav,
    dict: dict,
    dns: dns,
    example: example,
    file: file,
    ftp: ftp,
    geo: geo,
    go: go,
    gopher: gopher,
    h323: h323,
    http: http,
    https: https,
    iax: iax,
    icap: icap,
    im: im,
    imap: imap,
    info: info,
    ipp: ipp,
    ipps: ipps,
    iris: iris,
    irisbeep: irisbeep,
    irislwz: irislwz,
    irisxpc: irisxpc,
    irisxpcs: irisxpcs,
    ldap: ldap,
    mailto: mailto,
    mid: mid,
    msrp: msrp,
    msrps: msrps,
    mtqp: mtqp,
    mupdate: mupdate,
    news: news,
    nfs: nfs,
    ni: ni,
    nih: nih,
    nntp: nntp,
    opaquelocktoken: opaquelocktoken,
    pkcs11: pkcs11,
    pop: pop,
    pres: pres,
    reload: reload,
    rtsp: rtsp,
    rtsps: rtsps,
    rtspu: rtspu,
    service: service,
    session: session,
    shttp: shttp,
    sieve: sieve,
    sip: sip,
    sips: sips,
    sms: sms,
    snmp: snmp,
    soapbeep: soapbeep,
    soapbeeps: soapbeeps,
    stun: stun,
    stuns: stuns,
    tag: tag,
    tel: tel,
    telnet: telnet,
    tftp: tftp,
    thismessage: thismessage,
    tip: tip,
    tn3270: tn3270,
    turn: turn,
    turns: turns,
    tv: tv,
    urn: urn,
    vemmi: vemmi,
    vnc: vnc,
    ws: ws,
    wss: wss,
    xcon: xcon,
    xconuserid: xconuserid,
    xmlrpcbeep: xmlrpcbeep,
    xmlrpcbeeps: xmlrpcbeeps,
    xmpp: xmpp,
    z3950r: z3950r,
    z3950s: z3950s
};
