// Purescriptify.API.Types/index.js

// File auto generated by purescript-bridge! --
"use strict";
var Control_Lazy = require("../Control.Lazy/index.js");
var Control_Monad_Reader_Trans = require("../Control.Monad.Reader.Trans/index.js");
var Data_Argonaut_Decode_Aeson = require("../Data.Argonaut.Decode.Aeson/index.js");
var Data_Argonaut_Decode_Class = require("../Data.Argonaut.Decode.Class/index.js");
var Data_Argonaut_Encode_Aeson = require("../Data.Argonaut.Encode.Aeson/index.js");
var Data_Argonaut_Encode_Class = require("../Data.Argonaut.Encode.Class/index.js");
var Data_Argonaut_JSONDateTime = require("../Data.Argonaut.JSONDateTime/index.js");
var Data_Either = require("../Data.Either/index.js");
var Data_Eq = require("../Data.Eq/index.js");
var Data_Functor = require("../Data.Functor/index.js");
var Data_Functor_Contravariant = require("../Data.Functor.Contravariant/index.js");
var Data_Lens_Iso_Newtype = require("../Data.Lens.Iso.Newtype/index.js");
var Data_Maybe = require("../Data.Maybe/index.js");
var Data_Newtype = require("../Data.Newtype/index.js");
var Data_Op = require("../Data.Op/index.js");
var Data_Ord = require("../Data.Ord/index.js");
var Data_Ordering = require("../Data.Ordering/index.js");
var Data_Show = require("../Data.Show/index.js");
var Data_Show_Generic = require("../Data.Show.Generic/index.js");
var Data_UUID_Argonaut = require("../Data.UUID.Argonaut/index.js");
var Purescriptify_Utilities_Email = require("../Purescriptify.Utilities.Email/index.js");

//------------------------------------------------------------------------------
var Username = function (x) {
    return x;
};

//------------------------------------------------------------------------------
var UserId = function (x) {
    return x;
};

//------------------------------------------------------------------------------
var UpdateUserRequest = function (x) {
    return x;
};

//------------------------------------------------------------------------------
var $$Error = function (x) {
    return x;
};

//------------------------------------------------------------------------------
var CreatedAt = function (x) {
    return x;
};

//------------------------------------------------------------------------------
var UserData = function (x) {
    return x;
};

//------------------------------------------------------------------------------
var User = function (x) {
    return x;
};
var CreateUserRequest = function (x) {
    return x;
};
var toPathSegmentUserId = Data_UUID_Argonaut.toPathSegmentUUID;
var newtypeUsername_ = {
    Coercible0: function () {
        return undefined;
    }
};
var newtypeUser_ = {
    Coercible0: function () {
        return undefined;
    }
};
var newtypeUserId_ = {
    Coercible0: function () {
        return undefined;
    }
};
var newtypeUserData_ = {
    Coercible0: function () {
        return undefined;
    }
};
var newtypeUpdateUserRequest_ = {
    Coercible0: function () {
        return undefined;
    }
};
var newtypeError_ = {
    Coercible0: function () {
        return undefined;
    }
};
var newtypeCreatedAt_ = {
    Coercible0: function () {
        return undefined;
    }
};
var newtypeCreateUserRequest_ = {
    Coercible0: function () {
        return undefined;
    }
};
var genericUsername_ = {
    to: function (x) {
        return x;
    },
    from: function (x) {
        return x;
    }
};
var showUsername = {
    show: function (a) {
        return Data_Show_Generic.genericShow(genericUsername_)(Data_Show_Generic.genericShowConstructor(Data_Show_Generic.genericShowArgsArgument(Data_Show.showString))({
            reflectSymbol: function () {
                return "Username";
            }
        }))(a);
    }
};
var genericUser_ = {
    to: function (x) {
        return x;
    },
    from: function (x) {
        return x;
    }
};
var genericUserId_ = {
    to: function (x) {
        return x;
    },
    from: function (x) {
        return x;
    }
};
var showUserId = {
    show: function (a) {
        return Data_Show_Generic.genericShow(genericUserId_)(Data_Show_Generic.genericShowConstructor(Data_Show_Generic.genericShowArgsArgument(Data_UUID_Argonaut.showUUID))({
            reflectSymbol: function () {
                return "UserId";
            }
        }))(a);
    }
};
var genericUserData_ = {
    to: function (x) {
        return x;
    },
    from: function (x) {
        return x;
    }
};
var genericUpdateUserRequest_ = {
    to: function (x) {
        return x;
    },
    from: function (x) {
        return x;
    }
};
var showUpdateUserRequest = {
    show: function (a) {
        return Data_Show_Generic.genericShow(genericUpdateUserRequest_)(Data_Show_Generic.genericShowConstructor(Data_Show_Generic.genericShowArgsArgument(Data_Show.showRecord()(Data_Show.showRecordFieldsCons({
            reflectSymbol: function () {
                return "newEmail";
            }
        })(Data_Show.showRecordFieldsCons({
            reflectSymbol: function () {
                return "newUsername";
            }
        })(Data_Show.showRecordFieldsNil)(Data_Maybe.showMaybe(showUsername)))(Data_Maybe.showMaybe(Purescriptify_Utilities_Email.showEmail)))))({
            reflectSymbol: function () {
                return "UpdateUserRequest";
            }
        }))(a);
    }
};
var genericError_ = {
    to: function (x) {
        return x;
    },
    from: function (x) {
        return x;
    }
};
var showError = {
    show: function (a) {
        return Data_Show_Generic.genericShow(genericError_)(Data_Show_Generic.genericShowConstructor(Data_Show_Generic.genericShowArgsArgument(Data_Show.showRecord()(Data_Show.showRecordFieldsCons({
            reflectSymbol: function () {
                return "error";
            }
        })(Data_Show.showRecordFieldsCons({
            reflectSymbol: function () {
                return "status";
            }
        })(Data_Show.showRecordFieldsNil)(Data_Show.showInt))(Data_Show.showString))))({
            reflectSymbol: function () {
                return "Error";
            }
        }))(a);
    }
};
var genericCreatedAt_ = {
    to: function (x) {
        return x;
    },
    from: function (x) {
        return x;
    }
};
var showCreatedAt = {
    show: function (a) {
        return Data_Show_Generic.genericShow(genericCreatedAt_)(Data_Show_Generic.genericShowConstructor(Data_Show_Generic.genericShowArgsArgument(Data_Argonaut_JSONDateTime.showJSONDateTime))({
            reflectSymbol: function () {
                return "CreatedAt";
            }
        }))(a);
    }
};
var showUserData = {
    show: function (a) {
        return Data_Show_Generic.genericShow(genericUserData_)(Data_Show_Generic.genericShowConstructor(Data_Show_Generic.genericShowArgsArgument(Data_Show.showRecord()(Data_Show.showRecordFieldsCons({
            reflectSymbol: function () {
                return "created";
            }
        })(Data_Show.showRecordFieldsCons({
            reflectSymbol: function () {
                return "email";
            }
        })(Data_Show.showRecordFieldsCons({
            reflectSymbol: function () {
                return "username";
            }
        })(Data_Show.showRecordFieldsNil)(showUsername))(Purescriptify_Utilities_Email.showEmail))(showCreatedAt))))({
            reflectSymbol: function () {
                return "UserData";
            }
        }))(a);
    }
};
var showUser = {
    show: function (a) {
        return Data_Show_Generic.genericShow(genericUser_)(Data_Show_Generic.genericShowConstructor(Data_Show_Generic.genericShowArgsArgument(Data_Show.showRecord()(Data_Show.showRecordFieldsCons({
            reflectSymbol: function () {
                return "id";
            }
        })(Data_Show.showRecordFieldsCons({
            reflectSymbol: function () {
                return "info";
            }
        })(Data_Show.showRecordFieldsNil)(showUserData))(showUserId))))({
            reflectSymbol: function () {
                return "User";
            }
        }))(a);
    }
};
var genericCreateUserRequest_ = {
    to: function (x) {
        return x;
    },
    from: function (x) {
        return x;
    }
};
var showCreateUserRequest = {
    show: function (a) {
        return Data_Show_Generic.genericShow(genericCreateUserRequest_)(Data_Show_Generic.genericShowConstructor(Data_Show_Generic.genericShowArgsArgument(Data_Show.showRecord()(Data_Show.showRecordFieldsCons({
            reflectSymbol: function () {
                return "email";
            }
        })(Data_Show.showRecordFieldsCons({
            reflectSymbol: function () {
                return "username";
            }
        })(Data_Show.showRecordFieldsNil)(showUsername))(Purescriptify_Utilities_Email.showEmail))))({
            reflectSymbol: function () {
                return "CreateUserRequest";
            }
        }))(a);
    }
};
var eqUsername = {
    eq: function (x) {
        return function (y) {
            return x === y;
        };
    }
};
var ordUsername = {
    compare: function (x) {
        return function (y) {
            return Data_Ord.compare(Data_Ord.ordString)(x)(y);
        };
    },
    Eq0: function () {
        return eqUsername;
    }
};
var eqUserId = {
    eq: function (x) {
        return function (y) {
            return Data_Eq.eq(Data_UUID_Argonaut.eqUUID)(x)(y);
        };
    }
};
var ordUserId = {
    compare: function (x) {
        return function (y) {
            return Data_Ord.compare(Data_UUID_Argonaut.ordUUID)(x)(y);
        };
    },
    Eq0: function () {
        return eqUserId;
    }
};
var eqUpdateUserRequest = {
    eq: function (x) {
        return function (y) {
            return Data_Eq.eq(Data_Maybe.eqMaybe(Purescriptify_Utilities_Email.eqEmail))(x.newEmail)(y.newEmail) && Data_Eq.eq(Data_Maybe.eqMaybe(eqUsername))(x.newUsername)(y.newUsername);
        };
    }
};
var eqError = {
    eq: function (x) {
        return function (y) {
            return x.error === y.error && x.status === y.status;
        };
    }
};
var ordError = {
    compare: function (x) {
        return function (y) {
            var v = Data_Ord.compare(Data_Ord.ordString)(x.error)(y.error);
            if (v instanceof Data_Ordering.LT) {
                return Data_Ordering.LT.value;
            };
            if (v instanceof Data_Ordering.GT) {
                return Data_Ordering.GT.value;
            };
            return Data_Ord.compare(Data_Ord.ordInt)(x.status)(y.status);
        };
    },
    Eq0: function () {
        return eqError;
    }
};
var eqCreatedAt = {
    eq: function (x) {
        return function (y) {
            return Data_Eq.eq(Data_Argonaut_JSONDateTime.eqJSONDateTime)(x)(y);
        };
    }
};
var eqUserData = {
    eq: function (x) {
        return function (y) {
            return Data_Eq.eq(eqCreatedAt)(x.created)(y.created) && Data_Eq.eq(Purescriptify_Utilities_Email.eqEmail)(x.email)(y.email) && Data_Eq.eq(eqUsername)(x.username)(y.username);
        };
    }
};
var eqUser = {
    eq: function (x) {
        return function (y) {
            return Data_Eq.eq(eqUserId)(x.id)(y.id) && Data_Eq.eq(eqUserData)(x.info)(y.info);
        };
    }
};
var ordCreatedAt = {
    compare: function (x) {
        return function (y) {
            return Data_Ord.compare(Data_Argonaut_JSONDateTime.ordJSONDateTime)(x)(y);
        };
    },
    Eq0: function () {
        return eqCreatedAt;
    }
};
var eqCreateUserRequest = {
    eq: function (x) {
        return function (y) {
            return Data_Eq.eq(Purescriptify_Utilities_Email.eqEmail)(x.email)(y.email) && Data_Eq.eq(eqUsername)(x.username)(y.username);
        };
    }
};
var encodeJsonUsername = {
    encodeJson: Control_Lazy.defer(Control_Lazy.lazyFn)(function (v) {
        return Data_Argonaut_Encode_Aeson.encode(Data_Functor_Contravariant.cmap(Data_Op.contravariantOp)(Data_Newtype.unwrap())(Data_Argonaut_Encode_Aeson.value(Data_Argonaut_Encode_Class.encodeJsonJString)));
    })
};
var encodeJsonUserId = {
    encodeJson: Control_Lazy.defer(Control_Lazy.lazyFn)(function (v) {
        return Data_Argonaut_Encode_Aeson.encode(Data_Functor_Contravariant.cmap(Data_Op.contravariantOp)(Data_Newtype.unwrap())(Data_Argonaut_Encode_Aeson.value(Data_UUID_Argonaut.encodeJsonUUID)));
    })
};
var encodeJsonUpdateUserReque = {
    encodeJson: Control_Lazy.defer(Control_Lazy.lazyFn)(function (v) {
        return Data_Argonaut_Encode_Aeson.encode(Data_Functor_Contravariant.cmap(Data_Op.contravariantOp)(Data_Newtype.unwrap())(Data_Argonaut_Encode_Aeson.record()(Data_Argonaut_Encode_Aeson.rowListEncoderEncoderCons({
            reflectSymbol: function () {
                return "newEmail";
            }
        })()()()()(Data_Argonaut_Encode_Aeson.rowListEncoderEncoderCons({
            reflectSymbol: function () {
                return "newUsername";
            }
        })()()()()(Data_Argonaut_Encode_Aeson.rowListEncoderNil)))({
            newEmail: Data_Argonaut_Encode_Aeson.maybe(Data_Argonaut_Encode_Aeson.value(Purescriptify_Utilities_Email.encodeJsonEmail)),
            newUsername: Data_Argonaut_Encode_Aeson.maybe(Data_Argonaut_Encode_Aeson.value(encodeJsonUsername))
        })));
    })
};
var encodeJsonError = {
    encodeJson: Control_Lazy.defer(Control_Lazy.lazyFn)(function (v) {
        return Data_Argonaut_Encode_Aeson.encode(Data_Functor_Contravariant.cmap(Data_Op.contravariantOp)(Data_Newtype.unwrap())(Data_Argonaut_Encode_Aeson.record()(Data_Argonaut_Encode_Aeson.rowListEncoderEncoderCons({
            reflectSymbol: function () {
                return "error";
            }
        })()()()()(Data_Argonaut_Encode_Aeson.rowListEncoderEncoderCons({
            reflectSymbol: function () {
                return "status";
            }
        })()()()()(Data_Argonaut_Encode_Aeson.rowListEncoderNil)))({
            error: Data_Argonaut_Encode_Aeson.value(Data_Argonaut_Encode_Class.encodeJsonJString),
            status: Data_Argonaut_Encode_Aeson.value(Data_Argonaut_Encode_Class.encodeJsonInt)
        })));
    })
};
var encodeJsonCreatedAt = {
    encodeJson: Control_Lazy.defer(Control_Lazy.lazyFn)(function (v) {
        return Data_Argonaut_Encode_Aeson.encode(Data_Functor_Contravariant.cmap(Data_Op.contravariantOp)(Data_Newtype.unwrap())(Data_Argonaut_Encode_Aeson.value(Data_Argonaut_JSONDateTime.encodeJsonJSONDateTime)));
    })
};
var encodeJsonUserData = {
    encodeJson: Control_Lazy.defer(Control_Lazy.lazyFn)(function (v) {
        return Data_Argonaut_Encode_Aeson.encode(Data_Functor_Contravariant.cmap(Data_Op.contravariantOp)(Data_Newtype.unwrap())(Data_Argonaut_Encode_Aeson.record()(Data_Argonaut_Encode_Aeson.rowListEncoderEncoderCons({
            reflectSymbol: function () {
                return "created";
            }
        })()()()()(Data_Argonaut_Encode_Aeson.rowListEncoderEncoderCons({
            reflectSymbol: function () {
                return "email";
            }
        })()()()()(Data_Argonaut_Encode_Aeson.rowListEncoderEncoderCons({
            reflectSymbol: function () {
                return "username";
            }
        })()()()()(Data_Argonaut_Encode_Aeson.rowListEncoderNil))))({
            email: Data_Argonaut_Encode_Aeson.value(Purescriptify_Utilities_Email.encodeJsonEmail),
            username: Data_Argonaut_Encode_Aeson.value(encodeJsonUsername),
            created: Data_Argonaut_Encode_Aeson.value(encodeJsonCreatedAt)
        })));
    })
};
var encodeJsonUser = {
    encodeJson: Control_Lazy.defer(Control_Lazy.lazyFn)(function (v) {
        return Data_Argonaut_Encode_Aeson.encode(Data_Functor_Contravariant.cmap(Data_Op.contravariantOp)(Data_Newtype.unwrap())(Data_Argonaut_Encode_Aeson.record()(Data_Argonaut_Encode_Aeson.rowListEncoderEncoderCons({
            reflectSymbol: function () {
                return "id";
            }
        })()()()()(Data_Argonaut_Encode_Aeson.rowListEncoderEncoderCons({
            reflectSymbol: function () {
                return "info";
            }
        })()()()()(Data_Argonaut_Encode_Aeson.rowListEncoderNil)))({
            id: Data_Argonaut_Encode_Aeson.value(encodeJsonUserId),
            info: Data_Argonaut_Encode_Aeson.value(encodeJsonUserData)
        })));
    })
};
var encodeJsonCreateUserReque = {
    encodeJson: Control_Lazy.defer(Control_Lazy.lazyFn)(function (v) {
        return Data_Argonaut_Encode_Aeson.encode(Data_Functor_Contravariant.cmap(Data_Op.contravariantOp)(Data_Newtype.unwrap())(Data_Argonaut_Encode_Aeson.record()(Data_Argonaut_Encode_Aeson.rowListEncoderEncoderCons({
            reflectSymbol: function () {
                return "email";
            }
        })()()()()(Data_Argonaut_Encode_Aeson.rowListEncoderEncoderCons({
            reflectSymbol: function () {
                return "username";
            }
        })()()()()(Data_Argonaut_Encode_Aeson.rowListEncoderNil)))({
            email: Data_Argonaut_Encode_Aeson.value(Purescriptify_Utilities_Email.encodeJsonEmail),
            username: Data_Argonaut_Encode_Aeson.value(encodeJsonUsername)
        })));
    })
};
var decodeJsonUsername = {
    decodeJson: Control_Lazy.defer(Control_Lazy.lazyFn)(function (v) {
        return Data_Argonaut_Decode_Aeson.decode(Data_Functor.map(Control_Monad_Reader_Trans.functorReaderT(Data_Either.functorEither))(Username)(Data_Argonaut_Decode_Aeson.value(Data_Argonaut_Decode_Class.decodeJsonString)));
    })
};
var decodeJsonUserId = {
    decodeJson: Control_Lazy.defer(Control_Lazy.lazyFn)(function (v) {
        return Data_Argonaut_Decode_Aeson.decode(Data_Functor.map(Control_Monad_Reader_Trans.functorReaderT(Data_Either.functorEither))(UserId)(Data_Argonaut_Decode_Aeson.value(Data_UUID_Argonaut.decodeJsonUUID)));
    })
};
var decodeJsonUpdateUserReque = {
    decodeJson: Control_Lazy.defer(Control_Lazy.lazyFn)(function (v) {
        return Data_Argonaut_Decode_Aeson.decode(Data_Functor.map(Control_Monad_Reader_Trans.functorReaderT(Data_Either.functorEither))(UpdateUserRequest)(Data_Argonaut_Decode_Aeson.record()(Data_Argonaut_Decode_Aeson.rowListDecoderDecoderCons({
            reflectSymbol: function () {
                return "newEmail";
            }
        })()()()()(Data_Argonaut_Decode_Aeson.rowListDecoderDecoderCons({
            reflectSymbol: function () {
                return "newUsername";
            }
        })()()()()(Data_Argonaut_Decode_Aeson.rowListDecoderNil)))("UpdateUserRequest")({
            newEmail: Data_Argonaut_Decode_Aeson.maybe(Data_Argonaut_Decode_Aeson.value(Purescriptify_Utilities_Email.decodeJsonEmail)),
            newUsername: Data_Argonaut_Decode_Aeson.maybe(Data_Argonaut_Decode_Aeson.value(decodeJsonUsername))
        })));
    })
};
var decodeJsonError = {
    decodeJson: Control_Lazy.defer(Control_Lazy.lazyFn)(function (v) {
        return Data_Argonaut_Decode_Aeson.decode(Data_Functor.map(Control_Monad_Reader_Trans.functorReaderT(Data_Either.functorEither))($$Error)(Data_Argonaut_Decode_Aeson.record()(Data_Argonaut_Decode_Aeson.rowListDecoderDecoderCons({
            reflectSymbol: function () {
                return "error";
            }
        })()()()()(Data_Argonaut_Decode_Aeson.rowListDecoderDecoderCons({
            reflectSymbol: function () {
                return "status";
            }
        })()()()()(Data_Argonaut_Decode_Aeson.rowListDecoderNil)))("Error")({
            error: Data_Argonaut_Decode_Aeson.value(Data_Argonaut_Decode_Class.decodeJsonString),
            status: Data_Argonaut_Decode_Aeson.value(Data_Argonaut_Decode_Class.decodeJsonInt)
        })));
    })
};
var decodeJsonCreatedAt = {
    decodeJson: Control_Lazy.defer(Control_Lazy.lazyFn)(function (v) {
        return Data_Argonaut_Decode_Aeson.decode(Data_Functor.map(Control_Monad_Reader_Trans.functorReaderT(Data_Either.functorEither))(CreatedAt)(Data_Argonaut_Decode_Aeson.value(Data_Argonaut_JSONDateTime.decodeJsonJSONDateTime)));
    })
};
var decodeJsonUserData = {
    decodeJson: Control_Lazy.defer(Control_Lazy.lazyFn)(function (v) {
        return Data_Argonaut_Decode_Aeson.decode(Data_Functor.map(Control_Monad_Reader_Trans.functorReaderT(Data_Either.functorEither))(UserData)(Data_Argonaut_Decode_Aeson.record()(Data_Argonaut_Decode_Aeson.rowListDecoderDecoderCons({
            reflectSymbol: function () {
                return "created";
            }
        })()()()()(Data_Argonaut_Decode_Aeson.rowListDecoderDecoderCons({
            reflectSymbol: function () {
                return "email";
            }
        })()()()()(Data_Argonaut_Decode_Aeson.rowListDecoderDecoderCons({
            reflectSymbol: function () {
                return "username";
            }
        })()()()()(Data_Argonaut_Decode_Aeson.rowListDecoderNil))))("UserData")({
            email: Data_Argonaut_Decode_Aeson.value(Purescriptify_Utilities_Email.decodeJsonEmail),
            username: Data_Argonaut_Decode_Aeson.value(decodeJsonUsername),
            created: Data_Argonaut_Decode_Aeson.value(decodeJsonCreatedAt)
        })));
    })
};
var decodeJsonUser = {
    decodeJson: Control_Lazy.defer(Control_Lazy.lazyFn)(function (v) {
        return Data_Argonaut_Decode_Aeson.decode(Data_Functor.map(Control_Monad_Reader_Trans.functorReaderT(Data_Either.functorEither))(User)(Data_Argonaut_Decode_Aeson.record()(Data_Argonaut_Decode_Aeson.rowListDecoderDecoderCons({
            reflectSymbol: function () {
                return "id";
            }
        })()()()()(Data_Argonaut_Decode_Aeson.rowListDecoderDecoderCons({
            reflectSymbol: function () {
                return "info";
            }
        })()()()()(Data_Argonaut_Decode_Aeson.rowListDecoderNil)))("User")({
            id: Data_Argonaut_Decode_Aeson.value(decodeJsonUserId),
            info: Data_Argonaut_Decode_Aeson.value(decodeJsonUserData)
        })));
    })
};
var decodeJsonCreateUserReque = {
    decodeJson: Control_Lazy.defer(Control_Lazy.lazyFn)(function (v) {
        return Data_Argonaut_Decode_Aeson.decode(Data_Functor.map(Control_Monad_Reader_Trans.functorReaderT(Data_Either.functorEither))(CreateUserRequest)(Data_Argonaut_Decode_Aeson.record()(Data_Argonaut_Decode_Aeson.rowListDecoderDecoderCons({
            reflectSymbol: function () {
                return "email";
            }
        })()()()()(Data_Argonaut_Decode_Aeson.rowListDecoderDecoderCons({
            reflectSymbol: function () {
                return "username";
            }
        })()()()()(Data_Argonaut_Decode_Aeson.rowListDecoderNil)))("CreateUserRequest")({
            email: Data_Argonaut_Decode_Aeson.value(Purescriptify_Utilities_Email.decodeJsonEmail),
            username: Data_Argonaut_Decode_Aeson.value(decodeJsonUsername)
        })));
    })
};

//------------------------------------------------------------------------------
var _Username = function (dictProfunctor) {
    return Data_Lens_Iso_Newtype["_Newtype"]()()(dictProfunctor);
};

//------------------------------------------------------------------------------
var _UserId = function (dictProfunctor) {
    return Data_Lens_Iso_Newtype["_Newtype"]()()(dictProfunctor);
};

//------------------------------------------------------------------------------
var _UserData = function (dictProfunctor) {
    return Data_Lens_Iso_Newtype["_Newtype"]()()(dictProfunctor);
};

//------------------------------------------------------------------------------
var _User = function (dictProfunctor) {
    return Data_Lens_Iso_Newtype["_Newtype"]()()(dictProfunctor);
};

//------------------------------------------------------------------------------
var _UpdateUserRequest = function (dictProfunctor) {
    return Data_Lens_Iso_Newtype["_Newtype"]()()(dictProfunctor);
};

//------------------------------------------------------------------------------
var _Error = function (dictProfunctor) {
    return Data_Lens_Iso_Newtype["_Newtype"]()()(dictProfunctor);
};

//------------------------------------------------------------------------------
var _CreatedAt = function (dictProfunctor) {
    return Data_Lens_Iso_Newtype["_Newtype"]()()(dictProfunctor);
};

//------------------------------------------------------------------------------
var _CreateUserRequest = function (dictProfunctor) {
    return Data_Lens_Iso_Newtype["_Newtype"]()()(dictProfunctor);
};
module.exports = {
    CreateUserRequest: CreateUserRequest,
    "_CreateUserRequest": _CreateUserRequest,
    CreatedAt: CreatedAt,
    "_CreatedAt": _CreatedAt,
    "Error": $$Error,
    "_Error": _Error,
    UpdateUserRequest: UpdateUserRequest,
    "_UpdateUserRequest": _UpdateUserRequest,
    User: User,
    "_User": _User,
    UserData: UserData,
    "_UserData": _UserData,
    UserId: UserId,
    "_UserId": _UserId,
    Username: Username,
    "_Username": _Username,
    showCreateUserRequest: showCreateUserRequest,
    eqCreateUserRequest: eqCreateUserRequest,
    encodeJsonCreateUserReque: encodeJsonCreateUserReque,
    decodeJsonCreateUserReque: decodeJsonCreateUserReque,
    genericCreateUserRequest_: genericCreateUserRequest_,
    newtypeCreateUserRequest_: newtypeCreateUserRequest_,
    eqCreatedAt: eqCreatedAt,
    ordCreatedAt: ordCreatedAt,
    showCreatedAt: showCreatedAt,
    encodeJsonCreatedAt: encodeJsonCreatedAt,
    decodeJsonCreatedAt: decodeJsonCreatedAt,
    genericCreatedAt_: genericCreatedAt_,
    newtypeCreatedAt_: newtypeCreatedAt_,
    eqError: eqError,
    ordError: ordError,
    showError: showError,
    encodeJsonError: encodeJsonError,
    decodeJsonError: decodeJsonError,
    genericError_: genericError_,
    newtypeError_: newtypeError_,
    showUpdateUserRequest: showUpdateUserRequest,
    eqUpdateUserRequest: eqUpdateUserRequest,
    encodeJsonUpdateUserReque: encodeJsonUpdateUserReque,
    decodeJsonUpdateUserReque: decodeJsonUpdateUserReque,
    genericUpdateUserRequest_: genericUpdateUserRequest_,
    newtypeUpdateUserRequest_: newtypeUpdateUserRequest_,
    showUser: showUser,
    eqUser: eqUser,
    encodeJsonUser: encodeJsonUser,
    decodeJsonUser: decodeJsonUser,
    genericUser_: genericUser_,
    newtypeUser_: newtypeUser_,
    showUserData: showUserData,
    eqUserData: eqUserData,
    encodeJsonUserData: encodeJsonUserData,
    decodeJsonUserData: decodeJsonUserData,
    genericUserData_: genericUserData_,
    newtypeUserData_: newtypeUserData_,
    toPathSegmentUserId: toPathSegmentUserId,
    eqUserId: eqUserId,
    ordUserId: ordUserId,
    showUserId: showUserId,
    encodeJsonUserId: encodeJsonUserId,
    decodeJsonUserId: decodeJsonUserId,
    genericUserId_: genericUserId_,
    newtypeUserId_: newtypeUserId_,
    eqUsername: eqUsername,
    ordUsername: ordUsername,
    showUsername: showUsername,
    encodeJsonUsername: encodeJsonUsername,
    decodeJsonUsername: decodeJsonUsername,
    genericUsername_: genericUsername_,
    newtypeUsername_: newtypeUsername_
};
