// Routing.Duplex.Generic.Syntax/index.js
"use strict";
var Control_Applicative = require("../Control.Applicative/index.js");
var Data_Unit = require("../Data.Unit/index.js");
var Routing_Duplex = require("../Routing.Duplex/index.js");
var Routing_Duplex_Generic = require("../Routing.Duplex.Generic/index.js");
var gsepStringString = {
    gsep: function (a) {
        return function (b) {
            return Routing_Duplex.prefix(a)(Routing_Duplex.prefix(b)(Control_Applicative.pure(Routing_Duplex.applicativeRouteDuplex)(Data_Unit.unit)));
        };
    }
};
var gsepStringRoute = function (dictGRouteDuplexCtr) {
    return {
        gsep: function (a) {
            var $7 = Routing_Duplex.prefix(a);
            var $8 = Routing_Duplex_Generic.gRouteDuplexCtr(dictGRouteDuplexCtr);
            return function ($9) {
                return $7($8($9));
            };
        }
    };
};
var gsepRouteString = function (dictGRouteDuplexCtr) {
    return {
        gsep: (function () {
            var $10 = Routing_Duplex_Generic.gRouteDuplexCtr(dictGRouteDuplexCtr);
            return function ($11) {
                return Routing_Duplex.suffix($10($11));
            };
        })()
    };
};
var gsepProduct = function (dictGRouteDuplexCtr) {
    return {
        gsep: Routing_Duplex_Generic.product(dictGRouteDuplexCtr)
    };
};
var gsep = function (dict) {
    return dict.gsep;
};
var gparamsString = function (dictRouteDuplexParams) {
    return {
        gparams: function (a) {
            var $12 = Routing_Duplex.prefix(a);
            var $13 = Routing_Duplex.params(dictRouteDuplexParams);
            return function ($14) {
                return $12($13($14));
            };
        }
    };
};
var gparamsRoute = function (dictRouteDuplexParams) {
    return {
        gparams: function (a) {
            var $15 = Routing_Duplex_Generic.product(Routing_Duplex_Generic.gRouteAll)(a);
            var $16 = Routing_Duplex.params(dictRouteDuplexParams);
            return function ($17) {
                return $15($16($17));
            };
        }
    };
};
var gparams = function (dict) {
    return dict.gparams;
};
module.exports = {
    gparams: gparams,
    gsep: gsep,
    gsepStringString: gsepStringString,
    gsepStringRoute: gsepStringRoute,
    gsepRouteString: gsepRouteString,
    gsepProduct: gsepProduct,
    gparamsString: gparamsString,
    gparamsRoute: gparamsRoute
};
