// Test.Spec.Style/index.js
"use strict";
var Ansi_Codes = require("../Ansi.Codes/index.js");
var Data_Array = require("../Data.Array/index.js");
var Data_Foldable = require("../Data.Foldable/index.js");
var Data_List_NonEmpty = require("../Data.List.NonEmpty/index.js");
var Data_Maybe = require("../Data.Maybe/index.js");
var Data_String_CodeUnits = require("../Data.String.CodeUnits/index.js");
var yellow = [ new Ansi_Codes.PForeground(Ansi_Codes.Yellow.value) ];
var styled = function (as) {
    return function (str) {
        var v = Data_List_NonEmpty.fromFoldable(Data_Foldable.foldableArray)(as);
        if (v instanceof Data_Maybe.Nothing) {
            return str;
        };
        if (v instanceof Data_Maybe.Just) {
            return Ansi_Codes.escapeCodeToString(new Ansi_Codes.Graphics(v.value0)) + (str + Ansi_Codes.escapeCodeToString(new Ansi_Codes.Graphics(Data_List_NonEmpty.singleton(Ansi_Codes.Reset.value))));
        };
        throw new Error("Failed pattern match at Test.Spec.Style (line 16, column 3 - line 22, column 80): " + [ v.constructor.name ]);
    };
};
var red = [ new Ansi_Codes.PForeground(Ansi_Codes.Red.value) ];
var magenta = [ new Ansi_Codes.PForeground(Ansi_Codes.Magenta.value) ];
var indent = function (i) {
    return Data_String_CodeUnits.fromCharArray(Data_Array.replicate(i)(" "));
};
var green = [ new Ansi_Codes.PForeground(Ansi_Codes.Green.value) ];
var dim = [ new Ansi_Codes.PMode(Ansi_Codes.Dim.value) ];
var cyan = [ new Ansi_Codes.PForeground(Ansi_Codes.Cyan.value) ];
var bold = [ new Ansi_Codes.PMode(Ansi_Codes.Bold.value) ];
module.exports = {
    styled: styled,
    red: red,
    green: green,
    yellow: yellow,
    cyan: cyan,
    dim: dim,
    bold: bold,
    magenta: magenta,
    indent: indent
};
