// Purescriptify.Aggregate.Class/index.js

// File auto generated by purescript-bridge! --
"use strict";
var Control_Lazy = require("../Control.Lazy/index.js");
var Control_Monad_Reader_Trans = require("../Control.Monad.Reader.Trans/index.js");
var Data_Argonaut_Decode_Aeson = require("../Data.Argonaut.Decode.Aeson/index.js");
var Data_Argonaut_Encode_Aeson = require("../Data.Argonaut.Encode.Aeson/index.js");
var Data_Argonaut_JSONDateTime = require("../Data.Argonaut.JSONDateTime/index.js");
var Data_Either = require("../Data.Either/index.js");
var Data_Eq = require("../Data.Eq/index.js");
var Data_Functor = require("../Data.Functor/index.js");
var Data_Functor_Contravariant = require("../Data.Functor.Contravariant/index.js");
var Data_Lens_Iso_Newtype = require("../Data.Lens.Iso.Newtype/index.js");
var Data_Newtype = require("../Data.Newtype/index.js");
var Data_Op = require("../Data.Op/index.js");
var Data_Ord = require("../Data.Ord/index.js");
var Data_Show_Generic = require("../Data.Show.Generic/index.js");
var LastEdit = function (x) {
    return x;
};
var newtypeLastEdit_ = {
    Coercible0: function () {
        return undefined;
    }
};
var genericLastEdit_ = {
    to: function (x) {
        return x;
    },
    from: function (x) {
        return x;
    }
};
var showLastEdit = {
    show: function (a) {
        return Data_Show_Generic.genericShow(genericLastEdit_)(Data_Show_Generic.genericShowConstructor(Data_Show_Generic.genericShowArgsArgument(Data_Argonaut_JSONDateTime.showJSONDateTime))({
            reflectSymbol: function () {
                return "LastEdit";
            }
        }))(a);
    }
};
var eqLastEdit = {
    eq: function (x) {
        return function (y) {
            return Data_Eq.eq(Data_Argonaut_JSONDateTime.eqJSONDateTime)(x)(y);
        };
    }
};
var ordLastEdit = {
    compare: function (x) {
        return function (y) {
            return Data_Ord.compare(Data_Argonaut_JSONDateTime.ordJSONDateTime)(x)(y);
        };
    },
    Eq0: function () {
        return eqLastEdit;
    }
};
var encodeJsonLastEdit = {
    encodeJson: Control_Lazy.defer(Control_Lazy.lazyFn)(function (v) {
        return Data_Argonaut_Encode_Aeson.encode(Data_Functor_Contravariant.cmap(Data_Op.contravariantOp)(Data_Newtype.unwrap())(Data_Argonaut_Encode_Aeson.value(Data_Argonaut_JSONDateTime.encodeJsonJSONDateTime)));
    })
};
var decodeJsonLastEdit = {
    decodeJson: Control_Lazy.defer(Control_Lazy.lazyFn)(function (v) {
        return Data_Argonaut_Decode_Aeson.decode(Data_Functor.map(Control_Monad_Reader_Trans.functorReaderT(Data_Either.functorEither))(LastEdit)(Data_Argonaut_Decode_Aeson.value(Data_Argonaut_JSONDateTime.decodeJsonJSONDateTime)));
    })
};

//------------------------------------------------------------------------------
var _LastEdit = function (dictProfunctor) {
    return Data_Lens_Iso_Newtype["_Newtype"]()()(dictProfunctor);
};
module.exports = {
    LastEdit: LastEdit,
    "_LastEdit": _LastEdit,
    eqLastEdit: eqLastEdit,
    ordLastEdit: ordLastEdit,
    showLastEdit: showLastEdit,
    encodeJsonLastEdit: encodeJsonLastEdit,
    decodeJsonLastEdit: decodeJsonLastEdit,
    genericLastEdit_: genericLastEdit_,
    newtypeLastEdit_: newtypeLastEdit_
};
