// Purescriptify.AppM/index.js
"use strict";
var Control_Applicative = require("../Control.Applicative/index.js");
var Control_Bind = require("../Control.Bind/index.js");
var Control_Monad_Reader_Class = require("../Control.Monad.Reader.Class/index.js");
var Control_Monad_Reader_Trans = require("../Control.Monad.Reader.Trans/index.js");
var Data_Either = require("../Data.Either/index.js");
var Data_Functor = require("../Data.Functor/index.js");
var Data_Lens_Getter = require("../Data.Lens.Getter/index.js");
var Data_Lens_Internal_Forget = require("../Data.Lens.Internal.Forget/index.js");
var Data_Lens_Record = require("../Data.Lens.Record/index.js");
var Data_Lens_Setter = require("../Data.Lens.Setter/index.js");
var Data_Maybe = require("../Data.Maybe/index.js");
var Data_Ordering = require("../Data.Ordering/index.js");
var Data_Profunctor_Strong = require("../Data.Profunctor.Strong/index.js");
var Data_Semigroup = require("../Data.Semigroup/index.js");
var Data_Unit = require("../Data.Unit/index.js");
var Effect = require("../Effect/index.js");
var Effect_Aff = require("../Effect.Aff/index.js");
var Effect_Aff_Class = require("../Effect.Aff.Class/index.js");
var Effect_Class = require("../Effect.Class/index.js");
var Effect_Class_Console = require("../Effect.Class.Console/index.js");
var Effect_Now = require("../Effect.Now/index.js");
var Halogen_Subscription = require("../Halogen.Subscription/index.js");
var Purescriptify_API_Error = require("../Purescriptify.API.Error/index.js");
var Purescriptify_Capability_Global = require("../Purescriptify.Capability.Global/index.js");
var Purescriptify_Config = require("../Purescriptify.Config/index.js");
var Purescriptify_Data_Log = require("../Purescriptify.Data.Log/index.js");
var React_Halo_Component = require("../React.Halo.Component/index.js");
var React_Halo_Internal_Control = require("../React.Halo.Internal.Control/index.js");
var Servant_PureScript = require("../Servant.PureScript/index.js");
var ServerAPI = require("../ServerAPI/index.js");
var Type_Proxy = require("../Type.Proxy/index.js");
var URI_Authority = require("../URI.Authority/index.js");
var URI_Host = require("../URI.Host/index.js");
var URI_Host_RegName = require("../URI.Host.RegName/index.js");
var URI_Path_Segment = require("../URI.Path.Segment/index.js");
var URI_RelativePart = require("../URI.RelativePart/index.js");
var URI_RelativeRef = require("../URI.RelativeRef/index.js");
var Dev = (function () {
    function Dev() {

    };
    Dev.value = new Dev();
    return Dev;
})();
var Prod = (function () {
    function Prod() {

    };
    Prod.value = new Prod();
    return Prod;
})();
var AppM = function (x) {
    return x;
};
var showLogLevel = {
    show: function (v) {
        if (v instanceof Dev) {
            return "Dev";
        };
        if (v instanceof Prod) {
            return "Prod";
        };
        throw new Error("Failed pattern match at Purescriptify.AppM (line 57, column 1 - line 59, column 21): " + [ v.constructor.name ]);
    }
};
var setAuthority = function (apiUrl) {
    return function (req) {
        var rel = Data_Lens_Getter.viewOn(req)(Data_Lens_Record.prop({
            reflectSymbol: function () {
                return "uri";
            }
        })()()(Type_Proxy["Proxy"].value)(Data_Lens_Internal_Forget.strongForget));
        var auth = new URI_Authority.Authority(Data_Maybe.Nothing.value, new URI_Host.NameAddress(URI_Host_RegName.unsafeFromString(apiUrl)));
        var v = Data_Maybe.fromMaybe([  ])(Data_Lens_Getter.viewOn(rel)((function () {
            var $36 = URI_RelativeRef["_relPart"](Data_Lens_Internal_Forget.strongForget);
            var $37 = URI_RelativePart["_relPath"](Data_Lens_Internal_Forget.wanderForget(Data_Maybe.monoidMaybe(Data_Semigroup.semigroupArray)));
            return function ($38) {
                return $36($37($38));
            };
        })()));
        var newPart = new URI_RelativePart.RelativePartAuth(auth, Data_Functor.map(Data_Functor.functorArray)(URI_Path_Segment.segmentFromString)(v));
        var updated = Data_Lens_Setter.set(URI_RelativeRef["_relPart"](Data_Profunctor_Strong.strongFn))(newPart)(rel);
        return Data_Lens_Setter.set(Data_Lens_Record.prop({
            reflectSymbol: function () {
                return "uri";
            }
        })()()(Type_Proxy["Proxy"].value)(Data_Profunctor_Strong.strongFn))(updated)(req);
    };
};
var runAppM = function (env) {
    return function (v) {
        return Control_Monad_Reader_Trans.runReaderT(v)(env);
    };
};
var monadEffectAppM = Control_Monad_Reader_Trans.monadEffectReader(Effect_Aff.monadEffectAff);
var monadAppM = Control_Monad_Reader_Trans.monadReaderT(Effect_Aff.monadAff);
var monadGlobalAppM = {
    emitGlobalAction: function (act) {
        return Control_Bind.bind(Control_Monad_Reader_Trans.bindReaderT(Effect_Aff.bindAff))(Control_Monad_Reader_Class.asks(Control_Monad_Reader_Trans.monadAskReaderT(Effect_Aff.monadAff))(function (s) {
            return s.globalIO.actionsListener;
        }))(function (listener) {
            return Effect_Class.liftEffect(Control_Monad_Reader_Trans.monadEffectReader(Effect_Aff.monadEffectAff))(Halogen_Subscription.notify(listener)(act));
        });
    },
    emitGlobalEvent: function (ev) {
        return Control_Bind.bind(Control_Monad_Reader_Trans.bindReaderT(Effect_Aff.bindAff))(Control_Monad_Reader_Class.asks(Control_Monad_Reader_Trans.monadAskReaderT(Effect_Aff.monadAff))(function (s) {
            return s.globalIO.eventsListener;
        }))(function (listener) {
            return Effect_Class.liftEffect(Control_Monad_Reader_Trans.monadEffectReader(Effect_Aff.monadEffectAff))(Halogen_Subscription.notify(listener)(ev));
        });
    },
    getGlobalEventsEmitter: Control_Monad_Reader_Class.asks(Control_Monad_Reader_Trans.monadAskReaderT(Effect_Aff.monadAff))(function (s) {
        return s.globalIO.eventsEmitter;
    }),
    getGlobalActionsEmitter: Control_Monad_Reader_Class.asks(Control_Monad_Reader_Trans.monadAskReaderT(Effect_Aff.monadAff))(function (s) {
        return s.globalIO.actionsEmitter;
    }),
    Monad0: function () {
        return monadAppM;
    }
};
var nowAppM = {
    now: Effect_Class.liftEffect(monadEffectAppM)(Effect_Now.now),
    nowDate: Effect_Class.liftEffect(monadEffectAppM)(Effect_Now.nowDate),
    nowTime: Effect_Class.liftEffect(monadEffectAppM)(Effect_Now.nowTime),
    nowDateTime: Effect_Class.liftEffect(monadEffectAppM)(Effect_Now.nowDateTime),
    Monad0: function () {
        return monadAppM;
    }
};
var monadAjaxAppM = {
    request: function (dictContentType) {
        return function (dictContentType1) {
            return function (api) {
                var $39 = Servant_PureScript.request(Servant_PureScript.monadAjaxReaderT(Servant_PureScript.monadAjaxAff))(dictContentType)(dictContentType1)(api);
                var $40 = setAuthority(Purescriptify_Config.apiUrl);
                return function ($41) {
                    return AppM($39($40($41)));
                };
            };
        };
    },
    Monad0: function () {
        return monadAppM;
    }
};
var monadAffAppM = Effect_Aff_Class.monadAffReader(Effect_Aff_Class.monadAffAff);
var functorAppM = Control_Monad_Reader_Trans.functorReaderT(Effect_Aff.functorAff);

// | Halo
var monadHaloAppM = {
    component: function (name) {
        return function (spec) {
            return Control_Bind.bind(Control_Monad_Reader_Trans.bindReaderT(Effect_Aff.bindAff))(Control_Monad_Reader_Class.ask(Control_Monad_Reader_Trans.monadAskReaderT(Effect_Aff.monadAff)))(function (env) {
                return Effect_Class.liftEffect(Control_Monad_Reader_Trans.monadEffectReader(Effect_Aff.monadEffectAff))(React_Halo_Component.component(name)({
                    context: spec.context,
                    initialState: spec.initialState,
                    "eval": (function () {
                        var $42 = React_Halo_Internal_Control.hoist(functorAppM)(runAppM(env));
                        return function ($43) {
                            return $42(spec["eval"]($43));
                        };
                    })(),
                    render: spec.render
                }));
            });
        };
    },
    Monad0: function () {
        return monadAppM;
    }
};
var eqLogLevel = {
    eq: function (x) {
        return function (y) {
            if (x instanceof Dev && y instanceof Dev) {
                return true;
            };
            if (x instanceof Prod && y instanceof Prod) {
                return true;
            };
            return false;
        };
    }
};
var ordLogLevel = {
    compare: function (x) {
        return function (y) {
            if (x instanceof Dev && y instanceof Dev) {
                return Data_Ordering.EQ.value;
            };
            if (x instanceof Dev) {
                return Data_Ordering.LT.value;
            };
            if (y instanceof Dev) {
                return Data_Ordering.GT.value;
            };
            if (x instanceof Prod && y instanceof Prod) {
                return Data_Ordering.EQ.value;
            };
            throw new Error("Failed pattern match at Purescriptify.AppM (line 63, column 1 - line 63, column 44): " + [ x.constructor.name, y.constructor.name ]);
        };
    },
    Eq0: function () {
        return eqLogLevel;
    }
};
var callApi = function (dictMonad) {
    return function (dictMonadGlobal) {
        return function (apiCall) {
            return Control_Bind.bind(dictMonad.Bind1())(apiCall)(function (result) {
                if (result instanceof Data_Either.Left) {
                    return Control_Bind.discard(Control_Bind.discardUnit)(dictMonad.Bind1())(Purescriptify_Capability_Global.emitGlobalEvent(dictMonadGlobal)(new Purescriptify_Capability_Global.APIErrorOccurred(new Purescriptify_API_Error.APIError(result.value0))))(function () {
                        return Control_Applicative.pure(dictMonad.Applicative0())(new Data_Either.Left(new Purescriptify_API_Error.APIError(result.value0)));
                    });
                };
                if (result instanceof Data_Either.Right) {
                    return Control_Applicative.pure(dictMonad.Applicative0())(new Data_Either.Right(result.value0));
                };
                throw new Error("Failed pattern match at Purescriptify.AppM (line 156, column 5 - line 160, column 56): " + [ result.constructor.name ]);
            });
        };
    };
};
var bindAppM = Control_Monad_Reader_Trans.bindReaderT(Effect_Aff.bindAff);
var logMessagesAppM = {
    logMessage: function (log) {
        return Control_Bind.bind(bindAppM)(Control_Monad_Reader_Class.asks(Control_Monad_Reader_Trans.monadAskReaderT(Effect_Aff.monadAff))(function (v) {
            return v.logLevel;
        }))(function (logLevel) {
            return Effect_Class.liftEffect(monadEffectAppM)((function () {
                var v = Purescriptify_Data_Log.reason(log);
                if (logLevel instanceof Prod && v instanceof Purescriptify_Data_Log.Debug) {
                    return Control_Applicative.pure(Effect.applicativeEffect)(Data_Unit.unit);
                };
                return Effect_Class_Console.log(Effect_Class.monadEffectEffect)(Purescriptify_Data_Log.message(log));
            })());
        });
    },
    Monad0: function () {
        return monadAppM;
    }
};
var monadRoutingAppM = {
    read: Control_Bind.bindFlipped(bindAppM)(Effect_Class.liftEffect(monadEffectAppM))(Control_Monad_Reader_Class.asks(Control_Monad_Reader_Trans.monadAskReaderT(Effect_Aff.monadAff))(function (v) {
        return v.routing.read;
    })),
    reload: Control_Bind.bindFlipped(bindAppM)(Effect_Class.liftEffect(monadEffectAppM))(Control_Monad_Reader_Class.asks(Control_Monad_Reader_Trans.monadAskReaderT(Effect_Aff.monadAff))(function (v) {
        return v.routing.reload;
    })),
    getEmitter: Control_Monad_Reader_Class.asks(Control_Monad_Reader_Trans.monadAskReaderT(Effect_Aff.monadAff))(function (v) {
        return v.routing.emitter;
    }),
    navigate: function (route) {
        return Control_Bind.bind(bindAppM)(Control_Monad_Reader_Class.asks(Control_Monad_Reader_Trans.monadAskReaderT(Effect_Aff.monadAff))(function (v) {
            return v.routing.navigate;
        }))(function (f) {
            return Effect_Class.liftEffect(monadEffectAppM)(f(route));
        });
    },
    redirect: function (route) {
        return Control_Bind.bind(bindAppM)(Control_Monad_Reader_Class.asks(Control_Monad_Reader_Trans.monadAskReaderT(Effect_Aff.monadAff))(function (v) {
            return v.routing.redirect;
        }))(function (f) {
            return Effect_Class.liftEffect(monadEffectAppM)(f(route));
        });
    },
    Monad0: function () {
        return monadAppM;
    }
};
var applyAppM = Control_Monad_Reader_Trans.applyReaderT(Effect_Aff.applyAff);
var applicativeAppM = Control_Monad_Reader_Trans.applicativeReaderT(Effect_Aff.applicativeAff);
var getAuthHeader = Control_Applicative.pure(applicativeAppM)(new Data_Maybe.Just("Bearer eyJhbGciOiJIUzI1NiJ9.eyJ1c2VybmFtZSI6InRlc3QiLCJleHAiOjE5NzM1NDg4MDB9.Li3xg_6ikx7PQlPR6ca_WB6KV3xJnBnjDnB3AbQUFWI"));
var monadUsersAppM = {
    listUsers: callApi(monadAppM)(monadGlobalAppM)(ServerAPI.getUsers(monadAjaxAppM)),
    newUser: (function () {
        var $44 = callApi(monadAppM)(monadGlobalAppM);
        var $45 = ServerAPI.postUser(monadAjaxAppM);
        return function ($46) {
            return $44($45($46));
        };
    })(),
    deleteUser: function (uId) {
        return Control_Bind.bind(bindAppM)(getAuthHeader)(function (authHeader) {
            return callApi(monadAppM)(monadGlobalAppM)(ServerAPI.deleteUserByUserId(monadAjaxAppM)(authHeader)(uId));
        });
    },
    Monad0: function () {
        return monadAppM;
    }
};
module.exports = {
    Dev: Dev,
    Prod: Prod,
    AppM: AppM,
    runAppM: runAppM,
    getAuthHeader: getAuthHeader,
    callApi: callApi,
    setAuthority: setAuthority,
    showLogLevel: showLogLevel,
    eqLogLevel: eqLogLevel,
    ordLogLevel: ordLogLevel,
    monadHaloAppM: monadHaloAppM,
    functorAppM: functorAppM,
    applyAppM: applyAppM,
    applicativeAppM: applicativeAppM,
    bindAppM: bindAppM,
    monadAppM: monadAppM,
    monadEffectAppM: monadEffectAppM,
    monadAffAppM: monadAffAppM,
    nowAppM: nowAppM,
    logMessagesAppM: logMessagesAppM,
    monadAjaxAppM: monadAjaxAppM,
    monadGlobalAppM: monadGlobalAppM,
    monadRoutingAppM: monadRoutingAppM,
    monadUsersAppM: monadUsersAppM
};
