// Purescriptify.Page.Home/index.js
"use strict";
var Control_Applicative = require("../Control.Applicative/index.js");
var Control_Bind = require("../Control.Bind/index.js");
var Control_Monad_State_Class = require("../Control.Monad.State.Class/index.js");
var Data_Argonaut_JSONDateTime = require("../Data.Argonaut.JSONDateTime/index.js");
var Data_Array = require("../Data.Array/index.js");
var Data_Either = require("../Data.Either/index.js");
var Data_Eq = require("../Data.Eq/index.js");
var Data_Functor = require("../Data.Functor/index.js");
var Data_Maybe = require("../Data.Maybe/index.js");
var Data_Newtype = require("../Data.Newtype/index.js");
var Data_Semigroup = require("../Data.Semigroup/index.js");
var Data_Show = require("../Data.Show/index.js");
var Data_UUID_Argonaut = require("../Data.UUID.Argonaut/index.js");
var Data_Unit = require("../Data.Unit/index.js");
var Debug = require("../Debug/index.js");
var Network_RemoteData = require("../Network.RemoteData/index.js");
var Purescriptify_API_Types = require("../Purescriptify.API.Types/index.js");
var Purescriptify_Capability_Halo = require("../Purescriptify.Capability.Halo/index.js");
var Purescriptify_Capability_Log = require("../Purescriptify.Capability.Log/index.js");
var Purescriptify_Capability_Now = require("../Purescriptify.Capability.Now/index.js");
var Purescriptify_Capability_Users = require("../Purescriptify.Capability.Users/index.js");
var Purescriptify_Component_NewUserRow = require("../Purescriptify.Component.NewUserRow/index.js");
var Purescriptify_Config = require("../Purescriptify.Config/index.js");
var Purescriptify_Utilities_Email = require("../Purescriptify.Utilities.Email/index.js");
var React_Basic = require("../React.Basic/index.js");
var React_Basic_DOM = require("../React.Basic.DOM/index.js");
var React_Basic_DOM_Generated = require("../React.Basic.DOM.Generated/index.js");
var React_Basic_Events = require("../React.Basic.Events/index.js");
var React_Basic_Hooks_Internal = require("../React.Basic.Hooks.Internal/index.js");
var React_Halo_Internal_Control = require("../React.Halo.Internal.Control/index.js");
var React_Halo_Internal_Eval = require("../React.Halo.Internal.Eval/index.js");
var Type_Equality = require("../Type.Equality/index.js");
var Initialize = (function () {
    function Initialize() {

    };
    Initialize.value = new Initialize();
    return Initialize;
})();
var AddUser = (function () {
    function AddUser(value0) {
        this.value0 = value0;
    };
    AddUser.create = function (value0) {
        return new AddUser(value0);
    };
    return AddUser;
})();
var DeleteUser = (function () {
    function DeleteUser(value0) {
        this.value0 = value0;
    };
    DeleteUser.create = function (value0) {
        return new DeleteUser(value0);
    };
    return DeleteUser;
})();
var mkHomePage = function (dictMonadEffect) {
    return function (dictMonadUsers) {
        return function (dictNow) {
            return function (dictLogMessages) {
                return function (dictMonadHalo) {
                    var render = function (newUserRow) {
                        return function (v) {
                            var message = function (m) {
                                return React_Basic_DOM_Generated.section()({
                                    className: "rounded-lg bg-white p-4",
                                    children: [ React_Basic_DOM.text(m) ]
                                });
                            };
                            var headerEl = function (t) {
                                return React_Basic_DOM_Generated.th()({
                                    className: "p-6 text-gray-9 font-bold",
                                    children: [ React_Basic_DOM.text(t) ]
                                });
                            };
                            var cell = function (t) {
                                return React_Basic_DOM_Generated.td()({
                                    className: "p-6 text-gray-9",
                                    children: [ React_Basic_DOM.text(t) ]
                                });
                            };
                            var actions = function (id) {
                                var $$delete = React_Basic_DOM_Generated.button()({
                                    className: "p-2 text-white bg-red-4 rounded-md",
                                    onClick: React_Basic_Events.handler_(v.send(new DeleteUser(id))),
                                    children: [ React_Basic_DOM.text("Delete") ]
                                });
                                return React_Basic_DOM_Generated.td()({
                                    className: "",
                                    children: [ $$delete ]
                                });
                            };
                            var toUserRow = function (v1) {
                                return React_Basic_DOM_Generated.tr()({
                                    className: "",
                                    children: [ cell(Data_UUID_Argonaut.toString(Data_Newtype.unwrap()(v1.id))), cell(Data_Newtype.unwrap()(v1.info.username)), cell(Data_Show.show(Purescriptify_Utilities_Email.showEmail)(v1.info.email)), cell(Data_Show.show(Data_Argonaut_JSONDateTime.showJSONDateTime)(Data_Newtype.unwrap()(v1.info.created))), actions(v1.id) ]
                                });
                            };
                            return React_Basic_DOM_Generated.section()({
                                className: "w-full h-screen bg-blue-50 flex flex-col items-center justify-center",
                                children: [ (function () {
                                    if (v.state.usersR instanceof Network_RemoteData.Success) {
                                        return React_Basic.fragment([ React_Basic_DOM_Generated.table()({
                                            className: "table-auto border bg-white border-blue-100 rounded-lg border-separate",
                                            children: [ React_Basic_DOM_Generated.thead()({
                                                children: [ React_Basic_DOM_Generated.tr()({
                                                    children: [ headerEl("User Id"), headerEl("Username"), headerEl("Email"), headerEl("Created At"), headerEl("Actions") ]
                                                }) ]
                                            }), React_Basic_DOM_Generated.tbody()({
                                                children: Data_Functor.map(Data_Functor.functorArray)(toUserRow)(v.state.usersR.value0)
                                            }) ]
                                        }), newUserRow({
                                            onSubmit: function ($55) {
                                                return v.send(AddUser.create($55));
                                            }
                                        }) ]);
                                    };
                                    if (v.state.usersR instanceof Network_RemoteData.Loading) {
                                        return message("Loading stuff...");
                                    };
                                    if (v.state.usersR instanceof Network_RemoteData.Failure) {
                                        return message("Error! Make sure the API is running on " + (Purescriptify_Config.apiUrl + "."));
                                    };
                                    if (v.state.usersR instanceof Network_RemoteData.NotAsked) {
                                        return message("Preparing the world...");
                                    };
                                    throw new Error("Failed pattern match at Purescriptify.Page.Home (line 115, column 13 - line 143, column 70): " + [ v.state.usersR.constructor.name ]);
                                })() ]
                            });
                        };
                    };
                    var initialState = function (v) {
                        return function (v1) {
                            return {
                                usersR: Network_RemoteData.NotAsked.value
                            };
                        };
                    };
                    var handleAction = function (v) {
                        if (v instanceof Initialize) {
                            return Control_Bind.discard(Control_Bind.discardUnit)(React_Halo_Internal_Control.bindHaloM)(Control_Monad_State_Class.modify_(React_Halo_Internal_Control.monadStateHaloM)(function (v1) {
                                var $33 = {};
                                for (var $34 in v1) {
                                    if ({}.hasOwnProperty.call(v1, $34)) {
                                        $33[$34] = v1[$34];
                                    };
                                };
                                $33.usersR = Network_RemoteData.Loading.value;
                                return $33;
                            }))(function () {
                                return Control_Bind.bind(React_Halo_Internal_Control.bindHaloM)(Purescriptify_Capability_Users.listUsers(Purescriptify_Capability_Users.monadUsersHalogenM(dictMonadUsers)))(function (usersR) {
                                    return Control_Monad_State_Class.modify_(React_Halo_Internal_Control.monadStateHaloM)(function (v1) {
                                        var $36 = {};
                                        for (var $37 in v1) {
                                            if ({}.hasOwnProperty.call(v1, $37)) {
                                                $36[$37] = v1[$37];
                                            };
                                        };
                                        $36.usersR = Network_RemoteData.fromEither(usersR);
                                        return $36;
                                    });
                                });
                            });
                        };
                        if (v instanceof AddUser) {
                            return Control_Bind.bind(React_Halo_Internal_Control.bindHaloM)(Purescriptify_Capability_Users.newUser(Purescriptify_Capability_Users.monadUsersHalogenM(dictMonadUsers))(v.value0))(function (result) {
                                if (result instanceof Data_Either.Right) {
                                    return Control_Monad_State_Class.modify_(React_Halo_Internal_Control.monadStateHaloM)(function (currentState) {
                                        var updatedUsers = Data_Functor.map(Network_RemoteData.functorRemoteData)(function (users) {
                                            return Data_Semigroup.append(Data_Semigroup.semigroupArray)(users)([ result.value0 ]);
                                        })(currentState.usersR);
                                        var $40 = {};
                                        for (var $41 in currentState) {
                                            if ({}.hasOwnProperty.call(currentState, $41)) {
                                                $40[$41] = currentState[$41];
                                            };
                                        };
                                        $40.usersR = updatedUsers;
                                        return $40;
                                    });
                                };
                                if (result instanceof Data_Either.Left) {
                                    return Control_Bind.discard(Control_Bind.discardUnit)(React_Halo_Internal_Control.bindHaloM)(Debug.traceM()(React_Halo_Internal_Control.monadHaloM)(result.value0))(function () {
                                        return Purescriptify_Capability_Log.logError(Purescriptify_Capability_Log.logMessagesHalogenM(dictLogMessages))(Purescriptify_Capability_Now.nowHalogenM(dictNow))("Could not add a new user.");
                                    });
                                };
                                throw new Error("Failed pattern match at Purescriptify.Page.Home (line 86, column 7 - line 96, column 47): " + [ result.constructor.name ]);
                            });
                        };
                        if (v instanceof DeleteUser) {
                            return Control_Bind.bind(React_Halo_Internal_Control.bindHaloM)(Purescriptify_Capability_Users.deleteUser(Purescriptify_Capability_Users.monadUsersHalogenM(dictMonadUsers))(v.value0))(function (result) {
                                return Control_Bind.discard(Control_Bind.discardUnit)(React_Halo_Internal_Control.bindHaloM)((function () {
                                    if (result instanceof Data_Either.Right) {
                                        return Control_Monad_State_Class.modify_(React_Halo_Internal_Control.monadStateHaloM)(function (currentState) {
                                            var updatedUsers = Data_Functor.map(Network_RemoteData.functorRemoteData)(Data_Array.filter(function (v1) {
                                                return Data_Eq.notEq(Purescriptify_API_Types.eqUserId)(v1.id)(v.value0);
                                            }))(currentState.usersR);
                                            var $49 = {};
                                            for (var $50 in currentState) {
                                                if ({}.hasOwnProperty.call(currentState, $50)) {
                                                    $49[$50] = currentState[$50];
                                                };
                                            };
                                            $49.usersR = updatedUsers;
                                            return $49;
                                        });
                                    };
                                    if (result instanceof Data_Either.Left) {
                                        return Purescriptify_Capability_Log.logError(Purescriptify_Capability_Log.logMessagesHalogenM(dictLogMessages))(Purescriptify_Capability_Now.nowHalogenM(dictNow))("Could not delete user.");
                                    };
                                    throw new Error("Failed pattern match at Purescriptify.Page.Home (line 99, column 7 - line 108, column 44): " + [ result.constructor.name ]);
                                })())(function () {
                                    return Control_Applicative.pure(React_Halo_Internal_Control.applicativeHaloM)(Data_Unit.unit);
                                });
                            });
                        };
                        throw new Error("Failed pattern match at Purescriptify.Page.Home (line 79, column 18 - line 109, column 16): " + [ v.constructor.name ]);
                    };
                    var $$eval = React_Halo_Internal_Eval.mkEval(function (v) {
                        return {
                            onInitialize: function (v1) {
                                return new Data_Maybe.Just(Initialize.value);
                            },
                            onUpdate: v.onUpdate,
                            onAction: handleAction,
                            onFinalize: v.onFinalize
                        };
                    });
                    var context = function (v) {
                        return Control_Applicative.pure(React_Basic_Hooks_Internal.applicativeRender(Type_Equality.refl))(Data_Unit.unit);
                    };
                    return Control_Bind.bind((dictLogMessages.Monad0()).Bind1())(Purescriptify_Component_NewUserRow.mkNewUserRow(dictMonadEffect))(function (newUserRow) {
                        return Purescriptify_Capability_Halo.component(dictMonadHalo)("HomePage")({
                            context: context,
                            initialState: initialState,
                            "eval": $$eval,
                            render: render(newUserRow)
                        });
                    });
                };
            };
        };
    };
};
module.exports = {
    Initialize: Initialize,
    AddUser: AddUser,
    DeleteUser: DeleteUser,
    mkHomePage: mkHomePage
};
