// React.Basic.DOM.SVG/index.js

// | ----------------------------------------
// | THIS FILE IS GENERATED -- DO NOT EDIT IT
// | ----------------------------------------
"use strict";
var Effect_Unsafe = require("../Effect.Unsafe/index.js");
var React_Basic = require("../React.Basic/index.js");
var React_Basic_DOM_Internal = require("../React.Basic.DOM.Internal/index.js");
var _view$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("view"));
var view$prime = function () {
    return _view$prime;
};
var view = function () {
    return React_Basic.element(view$prime());
};
var view_ = function (children) {
    return view()({
        children: children
    });
};
var _video$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("video"));
var video$prime = function () {
    return _video$prime;
};
var video = function () {
    return React_Basic.element(video$prime());
};
var video_ = function (children) {
    return video()({
        children: children
    });
};
var _use$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("use"));
var use$prime = function () {
    return _use$prime;
};
var use = function () {
    return React_Basic.element(use$prime());
};
var use_ = function (children) {
    return use()({
        children: children
    });
};
var _unknown$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("unknown"));
var unknown$prime = function () {
    return _unknown$prime;
};
var unknown = function () {
    return React_Basic.element(unknown$prime());
};
var unknown_ = function (children) {
    return unknown()({
        children: children
    });
};
var _tspan$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("tspan"));
var tspan$prime = function () {
    return _tspan$prime;
};
var tspan = function () {
    return React_Basic.element(tspan$prime());
};
var tspan_ = function (children) {
    return tspan()({
        children: children
    });
};
var _title$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("title"));
var title$prime = function () {
    return _title$prime;
};
var title = function () {
    return React_Basic.element(title$prime());
};
var title_ = function (children) {
    return title()({
        children: children
    });
};
var _textPath$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("textPath"));
var textPath$prime = function () {
    return _textPath$prime;
};
var textPath = function () {
    return React_Basic.element(textPath$prime());
};
var textPath_ = function (children) {
    return textPath()({
        children: children
    });
};
var _textArea$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("textArea"));
var textArea$prime = function () {
    return _textArea$prime;
};
var textArea = function () {
    return React_Basic.element(textArea$prime());
};
var textArea_ = function (children) {
    return textArea()({
        children: children
    });
};
var _text$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("text"));
var text$prime = function () {
    return _text$prime;
};
var text = function () {
    return React_Basic.element(text$prime());
};
var text_ = function (children) {
    return text()({
        children: children
    });
};
var _tbreak$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("tbreak"));
var tbreak$prime = function () {
    return _tbreak$prime;
};
var tbreak = function () {
    return React_Basic.element(tbreak$prime());
};
var tbreak_ = function (children) {
    return tbreak()({
        children: children
    });
};
var _symbol$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("symbol"));
var symbol$prime = function () {
    return _symbol$prime;
};
var symbol = function () {
    return React_Basic.element(symbol$prime());
};
var symbol_ = function (children) {
    return symbol()({
        children: children
    });
};
var _switch$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("switch"));
var switch$prime = function () {
    return _switch$prime;
};
var $$switch = function () {
    return React_Basic.element(switch$prime());
};
var switch_ = function (children) {
    return $$switch()({
        children: children
    });
};
var _svg$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("svg"));
var svg$prime = function () {
    return _svg$prime;
};
var svg = function () {
    return React_Basic.element(svg$prime());
};
var svg_ = function (children) {
    return svg()({
        children: children
    });
};
var _style$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("style"));
var style$prime = function () {
    return _style$prime;
};
var style = function () {
    return React_Basic.element(style$prime());
};
var style_ = function (children) {
    return style()({
        children: children
    });
};
var _stop$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("stop"));
var stop$prime = function () {
    return _stop$prime;
};
var stop = function () {
    return React_Basic.element(stop$prime());
};
var stop_ = function (children) {
    return stop()({
        children: children
    });
};
var _set$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("set"));
var set$prime = function () {
    return _set$prime;
};
var set = function () {
    return React_Basic.element(set$prime());
};
var set_ = function (children) {
    return set()({
        children: children
    });
};
var _script$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("script"));
var script$prime = function () {
    return _script$prime;
};
var script = function () {
    return React_Basic.element(script$prime());
};
var script_ = function (children) {
    return script()({
        children: children
    });
};
var _rect$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("rect"));
var rect$prime = function () {
    return _rect$prime;
};
var rect = function () {
    return React_Basic.element(rect$prime());
};
var rect_ = function (children) {
    return rect()({
        children: children
    });
};
var _radialGradient$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("radialGradient"));
var radialGradient$prime = function () {
    return _radialGradient$prime;
};
var radialGradient = function () {
    return React_Basic.element(radialGradient$prime());
};
var radialGradient_ = function (children) {
    return radialGradient()({
        children: children
    });
};
var _prefetch$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("prefetch"));
var prefetch$prime = function () {
    return _prefetch$prime;
};
var prefetch = function () {
    return React_Basic.element(prefetch$prime());
};
var prefetch_ = function (children) {
    return prefetch()({
        children: children
    });
};
var _polyline$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("polyline"));
var polyline$prime = function () {
    return _polyline$prime;
};
var polyline = function () {
    return React_Basic.element(polyline$prime());
};
var polyline_ = function (children) {
    return polyline()({
        children: children
    });
};
var _polygon$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("polygon"));
var polygon$prime = function () {
    return _polygon$prime;
};
var polygon = function () {
    return React_Basic.element(polygon$prime());
};
var polygon_ = function (children) {
    return polygon()({
        children: children
    });
};
var _pattern$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("pattern"));
var pattern$prime = function () {
    return _pattern$prime;
};
var pattern = function () {
    return React_Basic.element(pattern$prime());
};
var pattern_ = function (children) {
    return pattern()({
        children: children
    });
};
var _path$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("path"));
var path$prime = function () {
    return _path$prime;
};
var path = function () {
    return React_Basic.element(path$prime());
};
var path_ = function (children) {
    return path()({
        children: children
    });
};
var _mpath$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("mpath"));
var mpath$prime = function () {
    return _mpath$prime;
};
var mpath = function () {
    return React_Basic.element(mpath$prime());
};
var mpath_ = function (children) {
    return mpath()({
        children: children
    });
};
var _metadata$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("metadata"));
var metadata$prime = function () {
    return _metadata$prime;
};
var metadata = function () {
    return React_Basic.element(metadata$prime());
};
var metadata_ = function (children) {
    return metadata()({
        children: children
    });
};
var _mask$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("mask"));
var mask$prime = function () {
    return _mask$prime;
};
var mask = function () {
    return React_Basic.element(mask$prime());
};
var mask_ = function (children) {
    return mask()({
        children: children
    });
};
var _marker$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("marker"));
var marker$prime = function () {
    return _marker$prime;
};
var marker = function () {
    return React_Basic.element(marker$prime());
};
var marker_ = function (children) {
    return marker()({
        children: children
    });
};
var _listener$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("listener"));
var listener$prime = function () {
    return _listener$prime;
};
var listener = function () {
    return React_Basic.element(listener$prime());
};
var listener_ = function (children) {
    return listener()({
        children: children
    });
};
var _linearGradient$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("linearGradient"));
var linearGradient$prime = function () {
    return _linearGradient$prime;
};
var linearGradient = function () {
    return React_Basic.element(linearGradient$prime());
};
var linearGradient_ = function (children) {
    return linearGradient()({
        children: children
    });
};
var _line$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("line"));
var line$prime = function () {
    return _line$prime;
};
var line = function () {
    return React_Basic.element(line$prime());
};
var line_ = function (children) {
    return line()({
        children: children
    });
};
var _image$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("image"));
var image$prime = function () {
    return _image$prime;
};
var image = function () {
    return React_Basic.element(image$prime());
};
var image_ = function (children) {
    return image()({
        children: children
    });
};
var _iframe$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("iframe"));
var iframe$prime = function () {
    return _iframe$prime;
};
var iframe = function () {
    return React_Basic.element(iframe$prime());
};
var iframe_ = function (children) {
    return iframe()({
        children: children
    });
};
var _handler$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("handler"));
var handler$prime = function () {
    return _handler$prime;
};
var handler = function () {
    return React_Basic.element(handler$prime());
};
var handler_ = function (children) {
    return handler()({
        children: children
    });
};
var _g$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("g"));
var g$prime = function () {
    return _g$prime;
};
var g = function () {
    return React_Basic.element(g$prime());
};
var g_ = function (children) {
    return g()({
        children: children
    });
};
var _foreignObject$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("foreignObject"));
var foreignObject$prime = function () {
    return _foreignObject$prime;
};
var foreignObject = function () {
    return React_Basic.element(foreignObject$prime());
};
var foreignObject_ = function (children) {
    return foreignObject()({
        children: children
    });
};
var _filter$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("filter"));
var filter$prime = function () {
    return _filter$prime;
};
var filter = function () {
    return React_Basic.element(filter$prime());
};
var filter_ = function (children) {
    return filter()({
        children: children
    });
};
var _feTurbulence$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("feTurbulence"));
var feTurbulence$prime = function () {
    return _feTurbulence$prime;
};
var feTurbulence = function () {
    return React_Basic.element(feTurbulence$prime());
};
var feTurbulence_ = function (children) {
    return feTurbulence()({
        children: children
    });
};
var _feTile$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("feTile"));
var feTile$prime = function () {
    return _feTile$prime;
};
var feTile = function () {
    return React_Basic.element(feTile$prime());
};
var feTile_ = function (children) {
    return feTile()({
        children: children
    });
};
var _feSpotLight$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("feSpotLight"));
var feSpotLight$prime = function () {
    return _feSpotLight$prime;
};
var feSpotLight = function () {
    return React_Basic.element(feSpotLight$prime());
};
var feSpotLight_ = function (children) {
    return feSpotLight()({
        children: children
    });
};
var _feSpecularLighting$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("feSpecularLighting"));
var feSpecularLighting$prime = function () {
    return _feSpecularLighting$prime;
};
var feSpecularLighting = function () {
    return React_Basic.element(feSpecularLighting$prime());
};
var feSpecularLighting_ = function (children) {
    return feSpecularLighting()({
        children: children
    });
};
var _fePointLight$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("fePointLight"));
var fePointLight$prime = function () {
    return _fePointLight$prime;
};
var fePointLight = function () {
    return React_Basic.element(fePointLight$prime());
};
var fePointLight_ = function (children) {
    return fePointLight()({
        children: children
    });
};
var _feOffset$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("feOffset"));
var feOffset$prime = function () {
    return _feOffset$prime;
};
var feOffset = function () {
    return React_Basic.element(feOffset$prime());
};
var feOffset_ = function (children) {
    return feOffset()({
        children: children
    });
};
var _feMorphology$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("feMorphology"));
var feMorphology$prime = function () {
    return _feMorphology$prime;
};
var feMorphology = function () {
    return React_Basic.element(feMorphology$prime());
};
var feMorphology_ = function (children) {
    return feMorphology()({
        children: children
    });
};
var _feMergeNode$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("feMergeNode"));
var feMergeNode$prime = function () {
    return _feMergeNode$prime;
};
var feMergeNode = function () {
    return React_Basic.element(feMergeNode$prime());
};
var feMergeNode_ = function (children) {
    return feMergeNode()({
        children: children
    });
};
var _feMerge$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("feMerge"));
var feMerge$prime = function () {
    return _feMerge$prime;
};
var feMerge = function () {
    return React_Basic.element(feMerge$prime());
};
var feMerge_ = function (children) {
    return feMerge()({
        children: children
    });
};
var _feImage$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("feImage"));
var feImage$prime = function () {
    return _feImage$prime;
};
var feImage = function () {
    return React_Basic.element(feImage$prime());
};
var feImage_ = function (children) {
    return feImage()({
        children: children
    });
};
var _feGaussianBlur$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("feGaussianBlur"));
var feGaussianBlur$prime = function () {
    return _feGaussianBlur$prime;
};
var feGaussianBlur = function () {
    return React_Basic.element(feGaussianBlur$prime());
};
var feGaussianBlur_ = function (children) {
    return feGaussianBlur()({
        children: children
    });
};
var _feFuncR$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("feFuncR"));
var feFuncR$prime = function () {
    return _feFuncR$prime;
};
var feFuncR = function () {
    return React_Basic.element(feFuncR$prime());
};
var feFuncR_ = function (children) {
    return feFuncR()({
        children: children
    });
};
var _feFuncG$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("feFuncG"));
var feFuncG$prime = function () {
    return _feFuncG$prime;
};
var feFuncG = function () {
    return React_Basic.element(feFuncG$prime());
};
var feFuncG_ = function (children) {
    return feFuncG()({
        children: children
    });
};
var _feFuncB$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("feFuncB"));
var feFuncB$prime = function () {
    return _feFuncB$prime;
};
var feFuncB = function () {
    return React_Basic.element(feFuncB$prime());
};
var feFuncB_ = function (children) {
    return feFuncB()({
        children: children
    });
};
var _feFuncA$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("feFuncA"));
var feFuncA$prime = function () {
    return _feFuncA$prime;
};
var feFuncA = function () {
    return React_Basic.element(feFuncA$prime());
};
var feFuncA_ = function (children) {
    return feFuncA()({
        children: children
    });
};
var _feFlood$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("feFlood"));
var feFlood$prime = function () {
    return _feFlood$prime;
};
var feFlood = function () {
    return React_Basic.element(feFlood$prime());
};
var feFlood_ = function (children) {
    return feFlood()({
        children: children
    });
};
var _feDropShadow$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("feDropShadow"));
var feDropShadow$prime = function () {
    return _feDropShadow$prime;
};
var feDropShadow = function () {
    return React_Basic.element(feDropShadow$prime());
};
var feDropShadow_ = function (children) {
    return feDropShadow()({
        children: children
    });
};
var _feDistantLight$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("feDistantLight"));
var feDistantLight$prime = function () {
    return _feDistantLight$prime;
};
var feDistantLight = function () {
    return React_Basic.element(feDistantLight$prime());
};
var feDistantLight_ = function (children) {
    return feDistantLight()({
        children: children
    });
};
var _feDisplacementMap$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("feDisplacementMap"));
var feDisplacementMap$prime = function () {
    return _feDisplacementMap$prime;
};
var feDisplacementMap = function () {
    return React_Basic.element(feDisplacementMap$prime());
};
var feDisplacementMap_ = function (children) {
    return feDisplacementMap()({
        children: children
    });
};
var _feDiffuseLighting$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("feDiffuseLighting"));
var feDiffuseLighting$prime = function () {
    return _feDiffuseLighting$prime;
};
var feDiffuseLighting = function () {
    return React_Basic.element(feDiffuseLighting$prime());
};
var feDiffuseLighting_ = function (children) {
    return feDiffuseLighting()({
        children: children
    });
};
var _feConvolveMatrix$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("feConvolveMatrix"));
var feConvolveMatrix$prime = function () {
    return _feConvolveMatrix$prime;
};
var feConvolveMatrix = function () {
    return React_Basic.element(feConvolveMatrix$prime());
};
var feConvolveMatrix_ = function (children) {
    return feConvolveMatrix()({
        children: children
    });
};
var _feComposite$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("feComposite"));
var feComposite$prime = function () {
    return _feComposite$prime;
};
var feComposite = function () {
    return React_Basic.element(feComposite$prime());
};
var feComposite_ = function (children) {
    return feComposite()({
        children: children
    });
};
var _feComponentTransfer$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("feComponentTransfer"));
var feComponentTransfer$prime = function () {
    return _feComponentTransfer$prime;
};
var feComponentTransfer = function () {
    return React_Basic.element(feComponentTransfer$prime());
};
var feComponentTransfer_ = function (children) {
    return feComponentTransfer()({
        children: children
    });
};
var _feColorMatrix$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("feColorMatrix"));
var feColorMatrix$prime = function () {
    return _feColorMatrix$prime;
};
var feColorMatrix = function () {
    return React_Basic.element(feColorMatrix$prime());
};
var feColorMatrix_ = function (children) {
    return feColorMatrix()({
        children: children
    });
};
var _feBlend$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("feBlend"));
var feBlend$prime = function () {
    return _feBlend$prime;
};
var feBlend = function () {
    return React_Basic.element(feBlend$prime());
};
var feBlend_ = function (children) {
    return feBlend()({
        children: children
    });
};
var _ellipse$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("ellipse"));
var ellipse$prime = function () {
    return _ellipse$prime;
};
var ellipse = function () {
    return React_Basic.element(ellipse$prime());
};
var ellipse_ = function (children) {
    return ellipse()({
        children: children
    });
};
var _discard$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("discard"));
var discard$prime = function () {
    return _discard$prime;
};
var discard = function () {
    return React_Basic.element(discard$prime());
};
var discard_ = function (children) {
    return discard()({
        children: children
    });
};
var _desc$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("desc"));
var desc$prime = function () {
    return _desc$prime;
};
var desc = function () {
    return React_Basic.element(desc$prime());
};
var desc_ = function (children) {
    return desc()({
        children: children
    });
};
var _defs$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("defs"));
var defs$prime = function () {
    return _defs$prime;
};
var defs = function () {
    return React_Basic.element(defs$prime());
};
var defs_ = function (children) {
    return defs()({
        children: children
    });
};
var _clipPath$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("clipPath"));
var clipPath$prime = function () {
    return _clipPath$prime;
};
var clipPath = function () {
    return React_Basic.element(clipPath$prime());
};
var clipPath_ = function (children) {
    return clipPath()({
        children: children
    });
};
var _circle$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("circle"));
var circle$prime = function () {
    return _circle$prime;
};
var circle = function () {
    return React_Basic.element(circle$prime());
};
var circle_ = function (children) {
    return circle()({
        children: children
    });
};
var _canvas$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("canvas"));
var canvas$prime = function () {
    return _canvas$prime;
};
var canvas = function () {
    return React_Basic.element(canvas$prime());
};
var canvas_ = function (children) {
    return canvas()({
        children: children
    });
};
var _audio$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("audio"));
var audio$prime = function () {
    return _audio$prime;
};
var audio = function () {
    return React_Basic.element(audio$prime());
};
var audio_ = function (children) {
    return audio()({
        children: children
    });
};
var _animation$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("animation"));
var animation$prime = function () {
    return _animation$prime;
};
var animation = function () {
    return React_Basic.element(animation$prime());
};
var animation_ = function (children) {
    return animation()({
        children: children
    });
};
var _animateTransform$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("animateTransform"));
var animateTransform$prime = function () {
    return _animateTransform$prime;
};
var animateTransform = function () {
    return React_Basic.element(animateTransform$prime());
};
var animateTransform_ = function (children) {
    return animateTransform()({
        children: children
    });
};
var _animateMotion$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("animateMotion"));
var animateMotion$prime = function () {
    return _animateMotion$prime;
};
var animateMotion = function () {
    return React_Basic.element(animateMotion$prime());
};
var animateMotion_ = function (children) {
    return animateMotion()({
        children: children
    });
};
var _animateColor$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("animateColor"));
var animateColor$prime = function () {
    return _animateColor$prime;
};
var animateColor = function () {
    return React_Basic.element(animateColor$prime());
};
var animateColor_ = function (children) {
    return animateColor()({
        children: children
    });
};
var _animate$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("animate"));
var animate$prime = function () {
    return _animate$prime;
};
var animate = function () {
    return React_Basic.element(animate$prime());
};
var animate_ = function (children) {
    return animate()({
        children: children
    });
};
var _a$prime = Effect_Unsafe.unsafePerformEffect(React_Basic_DOM_Internal.unsafeCreateDOMComponent("a"));
var a$prime = function () {
    return _a$prime;
};
var a = function () {
    return React_Basic.element(a$prime());
};
var a_ = function (children) {
    return a()({
        children: children
    });
};
module.exports = {
    a: a,
    a_: a_,
    "a'": a$prime,
    "_a'": _a$prime,
    animate: animate,
    animate_: animate_,
    "animate'": animate$prime,
    "_animate'": _animate$prime,
    animateColor: animateColor,
    animateColor_: animateColor_,
    "animateColor'": animateColor$prime,
    "_animateColor'": _animateColor$prime,
    animateMotion: animateMotion,
    animateMotion_: animateMotion_,
    "animateMotion'": animateMotion$prime,
    "_animateMotion'": _animateMotion$prime,
    animateTransform: animateTransform,
    animateTransform_: animateTransform_,
    "animateTransform'": animateTransform$prime,
    "_animateTransform'": _animateTransform$prime,
    animation: animation,
    animation_: animation_,
    "animation'": animation$prime,
    "_animation'": _animation$prime,
    audio: audio,
    audio_: audio_,
    "audio'": audio$prime,
    "_audio'": _audio$prime,
    canvas: canvas,
    canvas_: canvas_,
    "canvas'": canvas$prime,
    "_canvas'": _canvas$prime,
    circle: circle,
    circle_: circle_,
    "circle'": circle$prime,
    "_circle'": _circle$prime,
    clipPath: clipPath,
    clipPath_: clipPath_,
    "clipPath'": clipPath$prime,
    "_clipPath'": _clipPath$prime,
    defs: defs,
    defs_: defs_,
    "defs'": defs$prime,
    "_defs'": _defs$prime,
    desc: desc,
    desc_: desc_,
    "desc'": desc$prime,
    "_desc'": _desc$prime,
    discard: discard,
    discard_: discard_,
    "discard'": discard$prime,
    "_discard'": _discard$prime,
    ellipse: ellipse,
    ellipse_: ellipse_,
    "ellipse'": ellipse$prime,
    "_ellipse'": _ellipse$prime,
    feBlend: feBlend,
    feBlend_: feBlend_,
    "feBlend'": feBlend$prime,
    "_feBlend'": _feBlend$prime,
    feColorMatrix: feColorMatrix,
    feColorMatrix_: feColorMatrix_,
    "feColorMatrix'": feColorMatrix$prime,
    "_feColorMatrix'": _feColorMatrix$prime,
    feComponentTransfer: feComponentTransfer,
    feComponentTransfer_: feComponentTransfer_,
    "feComponentTransfer'": feComponentTransfer$prime,
    "_feComponentTransfer'": _feComponentTransfer$prime,
    feComposite: feComposite,
    feComposite_: feComposite_,
    "feComposite'": feComposite$prime,
    "_feComposite'": _feComposite$prime,
    feConvolveMatrix: feConvolveMatrix,
    feConvolveMatrix_: feConvolveMatrix_,
    "feConvolveMatrix'": feConvolveMatrix$prime,
    "_feConvolveMatrix'": _feConvolveMatrix$prime,
    feDiffuseLighting: feDiffuseLighting,
    feDiffuseLighting_: feDiffuseLighting_,
    "feDiffuseLighting'": feDiffuseLighting$prime,
    "_feDiffuseLighting'": _feDiffuseLighting$prime,
    feDisplacementMap: feDisplacementMap,
    feDisplacementMap_: feDisplacementMap_,
    "feDisplacementMap'": feDisplacementMap$prime,
    "_feDisplacementMap'": _feDisplacementMap$prime,
    feDistantLight: feDistantLight,
    feDistantLight_: feDistantLight_,
    "feDistantLight'": feDistantLight$prime,
    "_feDistantLight'": _feDistantLight$prime,
    feDropShadow: feDropShadow,
    feDropShadow_: feDropShadow_,
    "feDropShadow'": feDropShadow$prime,
    "_feDropShadow'": _feDropShadow$prime,
    feFlood: feFlood,
    feFlood_: feFlood_,
    "feFlood'": feFlood$prime,
    "_feFlood'": _feFlood$prime,
    feFuncA: feFuncA,
    feFuncA_: feFuncA_,
    "feFuncA'": feFuncA$prime,
    "_feFuncA'": _feFuncA$prime,
    feFuncB: feFuncB,
    feFuncB_: feFuncB_,
    "feFuncB'": feFuncB$prime,
    "_feFuncB'": _feFuncB$prime,
    feFuncG: feFuncG,
    feFuncG_: feFuncG_,
    "feFuncG'": feFuncG$prime,
    "_feFuncG'": _feFuncG$prime,
    feFuncR: feFuncR,
    feFuncR_: feFuncR_,
    "feFuncR'": feFuncR$prime,
    "_feFuncR'": _feFuncR$prime,
    feGaussianBlur: feGaussianBlur,
    feGaussianBlur_: feGaussianBlur_,
    "feGaussianBlur'": feGaussianBlur$prime,
    "_feGaussianBlur'": _feGaussianBlur$prime,
    feImage: feImage,
    feImage_: feImage_,
    "feImage'": feImage$prime,
    "_feImage'": _feImage$prime,
    feMerge: feMerge,
    feMerge_: feMerge_,
    "feMerge'": feMerge$prime,
    "_feMerge'": _feMerge$prime,
    feMergeNode: feMergeNode,
    feMergeNode_: feMergeNode_,
    "feMergeNode'": feMergeNode$prime,
    "_feMergeNode'": _feMergeNode$prime,
    feMorphology: feMorphology,
    feMorphology_: feMorphology_,
    "feMorphology'": feMorphology$prime,
    "_feMorphology'": _feMorphology$prime,
    feOffset: feOffset,
    feOffset_: feOffset_,
    "feOffset'": feOffset$prime,
    "_feOffset'": _feOffset$prime,
    fePointLight: fePointLight,
    fePointLight_: fePointLight_,
    "fePointLight'": fePointLight$prime,
    "_fePointLight'": _fePointLight$prime,
    feSpecularLighting: feSpecularLighting,
    feSpecularLighting_: feSpecularLighting_,
    "feSpecularLighting'": feSpecularLighting$prime,
    "_feSpecularLighting'": _feSpecularLighting$prime,
    feSpotLight: feSpotLight,
    feSpotLight_: feSpotLight_,
    "feSpotLight'": feSpotLight$prime,
    "_feSpotLight'": _feSpotLight$prime,
    feTile: feTile,
    feTile_: feTile_,
    "feTile'": feTile$prime,
    "_feTile'": _feTile$prime,
    feTurbulence: feTurbulence,
    feTurbulence_: feTurbulence_,
    "feTurbulence'": feTurbulence$prime,
    "_feTurbulence'": _feTurbulence$prime,
    filter: filter,
    filter_: filter_,
    "filter'": filter$prime,
    "_filter'": _filter$prime,
    foreignObject: foreignObject,
    foreignObject_: foreignObject_,
    "foreignObject'": foreignObject$prime,
    "_foreignObject'": _foreignObject$prime,
    g: g,
    g_: g_,
    "g'": g$prime,
    "_g'": _g$prime,
    handler: handler,
    handler_: handler_,
    "handler'": handler$prime,
    "_handler'": _handler$prime,
    iframe: iframe,
    iframe_: iframe_,
    "iframe'": iframe$prime,
    "_iframe'": _iframe$prime,
    image: image,
    image_: image_,
    "image'": image$prime,
    "_image'": _image$prime,
    line: line,
    line_: line_,
    "line'": line$prime,
    "_line'": _line$prime,
    linearGradient: linearGradient,
    linearGradient_: linearGradient_,
    "linearGradient'": linearGradient$prime,
    "_linearGradient'": _linearGradient$prime,
    listener: listener,
    listener_: listener_,
    "listener'": listener$prime,
    "_listener'": _listener$prime,
    marker: marker,
    marker_: marker_,
    "marker'": marker$prime,
    "_marker'": _marker$prime,
    mask: mask,
    mask_: mask_,
    "mask'": mask$prime,
    "_mask'": _mask$prime,
    metadata: metadata,
    metadata_: metadata_,
    "metadata'": metadata$prime,
    "_metadata'": _metadata$prime,
    mpath: mpath,
    mpath_: mpath_,
    "mpath'": mpath$prime,
    "_mpath'": _mpath$prime,
    path: path,
    path_: path_,
    "path'": path$prime,
    "_path'": _path$prime,
    pattern: pattern,
    pattern_: pattern_,
    "pattern'": pattern$prime,
    "_pattern'": _pattern$prime,
    polygon: polygon,
    polygon_: polygon_,
    "polygon'": polygon$prime,
    "_polygon'": _polygon$prime,
    polyline: polyline,
    polyline_: polyline_,
    "polyline'": polyline$prime,
    "_polyline'": _polyline$prime,
    prefetch: prefetch,
    prefetch_: prefetch_,
    "prefetch'": prefetch$prime,
    "_prefetch'": _prefetch$prime,
    radialGradient: radialGradient,
    radialGradient_: radialGradient_,
    "radialGradient'": radialGradient$prime,
    "_radialGradient'": _radialGradient$prime,
    rect: rect,
    rect_: rect_,
    "rect'": rect$prime,
    "_rect'": _rect$prime,
    script: script,
    script_: script_,
    "script'": script$prime,
    "_script'": _script$prime,
    set: set,
    set_: set_,
    "set'": set$prime,
    "_set'": _set$prime,
    stop: stop,
    stop_: stop_,
    "stop'": stop$prime,
    "_stop'": _stop$prime,
    style: style,
    style_: style_,
    "style'": style$prime,
    "_style'": _style$prime,
    svg: svg,
    svg_: svg_,
    "svg'": svg$prime,
    "_svg'": _svg$prime,
    "switch": $$switch,
    switch_: switch_,
    "switch'": switch$prime,
    "_switch'": _switch$prime,
    symbol: symbol,
    symbol_: symbol_,
    "symbol'": symbol$prime,
    "_symbol'": _symbol$prime,
    tbreak: tbreak,
    tbreak_: tbreak_,
    "tbreak'": tbreak$prime,
    "_tbreak'": _tbreak$prime,
    text: text,
    text_: text_,
    "text'": text$prime,
    "_text'": _text$prime,
    textArea: textArea,
    textArea_: textArea_,
    "textArea'": textArea$prime,
    "_textArea'": _textArea$prime,
    textPath: textPath,
    textPath_: textPath_,
    "textPath'": textPath$prime,
    "_textPath'": _textPath$prime,
    title: title,
    title_: title_,
    "title'": title$prime,
    "_title'": _title$prime,
    tspan: tspan,
    tspan_: tspan_,
    "tspan'": tspan$prime,
    "_tspan'": _tspan$prime,
    unknown: unknown,
    unknown_: unknown_,
    "unknown'": unknown$prime,
    "_unknown'": _unknown$prime,
    use: use,
    use_: use_,
    "use'": use$prime,
    "_use'": _use$prime,
    video: video,
    video_: video_,
    "video'": video$prime,
    "_video'": _video$prime,
    view: view,
    view_: view_,
    "view'": view$prime,
    "_view'": _view$prime
};
