// React.Halo.Internal.State/index.js
"use strict";
var Data_Map_Internal = require("../Data.Map.Internal/index.js");
var Effect_Ref = require("../Effect.Ref/index.js");

// | HThe alo component state used during evaluation.
var HaloState = function (x) {
    return x;
};

// | Issue a new identifier, unique to this component.
var fresh = function (f) {
    return function (v) {
        return Effect_Ref["modify'"](function (a) {
            return {
                state: a + 1 | 0,
                value: f(a)
            };
        })(v.fresh);
    };
};

// | Creates a starting `HaloState`, ready for initialization.
var createInitialState = function (v) {
    return function __do() {
        var finalized = Effect_Ref["new"](false)();
        var fresh$prime = Effect_Ref["new"](0)();
        var props = Effect_Ref["new"](v.props)();
        var context = Effect_Ref["new"](v.context)();
        var state = Effect_Ref["new"](v.state)();
        var subscriptions = Effect_Ref["new"](Data_Map_Internal.empty)();
        var forks = Effect_Ref["new"](Data_Map_Internal.empty)();
        return {
            "eval": v["eval"],
            update: v.update,
            finalized: finalized,
            props: props,
            context: context,
            state: state,
            fresh: fresh$prime,
            subscriptions: subscriptions,
            forks: forks
        };
    };
};
module.exports = {
    HaloState: HaloState,
    createInitialState: createInitialState,
    fresh: fresh
};
