// Web.File.FileReader/index.js
"use strict";
var $foreign = require("./foreign.js");
var Data_Maybe = require("../Data.Maybe/index.js");
var Unsafe_Coerce = require("../Unsafe.Coerce/index.js");
var Web_File_FileReader_ReadyState = require("../Web.File.FileReader.ReadyState/index.js");
var Web_Internal_FFI = require("../Web.Internal.FFI/index.js");
var toEventTarget = Unsafe_Coerce.unsafeCoerce;
var readyState = function (fr) {
    return function __do() {
        var rs = $foreign.readyStateImpl(fr)();
        return Data_Maybe.fromJust()(Web_File_FileReader_ReadyState.toEnumReadyState(rs));
    };
};
var fromEventTarget = Web_Internal_FFI.unsafeReadProtoTagged("FileReader");
module.exports = {
    fromEventTarget: fromEventTarget,
    toEventTarget: toEventTarget,
    readyState: readyState,
    fileReader: $foreign.fileReader,
    error: $foreign.error,
    result: $foreign.result,
    abort: $foreign.abort,
    readAsText: $foreign.readAsText,
    readAsArrayBuffer: $foreign.readAsArrayBuffer,
    readAsDataURL: $foreign.readAsDataURL
};
