// Purescriptify.API.Error/index.js
"use strict";
var Control_Applicative = require("../Control.Applicative/index.js");
var Data_Argonaut_Core = require("../Data.Argonaut.Core/index.js");
var Data_Argonaut_Decode_Class = require("../Data.Argonaut.Decode.Class/index.js");
var Data_Argonaut_Decode_Error = require("../Data.Argonaut.Decode.Error/index.js");
var Data_Either = require("../Data.Either/index.js");
var Data_Functor = require("../Data.Functor/index.js");
var Data_Maybe = require("../Data.Maybe/index.js");
var Data_Unit = require("../Data.Unit/index.js");
var Purescriptify_API_Types = require("../Purescriptify.API.Types/index.js");
var Servant_PureScript = require("../Servant.PureScript/index.js");
var ValidationError = (function () {
    function ValidationError(value0) {
        this.value0 = value0;
    };
    ValidationError.create = function (value0) {
        return new ValidationError(value0);
    };
    return ValidationError;
})();
var UnknownError = (function () {
    function UnknownError(value0) {
        this.value0 = value0;
    };
    UnknownError.create = function (value0) {
        return new UnknownError(value0);
    };
    return UnknownError;
})();
var APIError = (function () {
    function APIError(value0) {
        this.value0 = value0;
    };
    APIError.create = function (value0) {
        return new APIError(value0);
    };
    return APIError;
})();
var UnauthorizedAccess = (function () {
    function UnauthorizedAccess() {

    };
    UnauthorizedAccess.value = new UnauthorizedAccess();
    return UnauthorizedAccess;
})();
var withErrorContents = function (dictMonad) {
    return function (v) {
        return function (v1) {
            if (v instanceof APIError) {
                var v2 = Data_Functor.map(Data_Maybe.functorMaybe)(function (v3) {
                    return v3.body;
                })(v.value0.response);
                if (v2 instanceof Data_Maybe.Just) {
                    var v3 = Data_Argonaut_Decode_Class.decodeJson(Purescriptify_API_Types.decodeJsonError)(v2.value0);
                    if (v3 instanceof Data_Either.Right) {
                        return v1(new ValidationError(v3.value0));
                    };
                    if (v3 instanceof Data_Either.Left) {
                        return v1(new UnknownError(new Data_Maybe.Just(v2.value0)));
                    };
                    throw new Error("Failed pattern match at Purescriptify.API.Error (line 21, column 15 - line 23, column 57): " + [ v3.constructor.name ]);
                };
                if (v2 instanceof Data_Maybe.Nothing) {
                    return v1(new UnknownError(Data_Maybe.Nothing.value));
                };
                throw new Error("Failed pattern match at Purescriptify.API.Error (line 20, column 64 - line 24, column 43): " + [ v2.constructor.name ]);
            };
            return Control_Applicative.pure(dictMonad.Applicative0())(Data_Unit.unit);
        };
    };
};
var printAPIError = function (v) {
    if (v instanceof APIError) {
        return Servant_PureScript.printAjaxError(Data_Argonaut_Core.stringify)(Data_Argonaut_Decode_Error.printJsonDecodeError)(v.value0);
    };
    if (v instanceof UnauthorizedAccess) {
        return "Unauthorized access.";
    };
    throw new Error("Failed pattern match at Purescriptify.API.Error (line 28, column 1 - line 28, column 36): " + [ v.constructor.name ]);
};
module.exports = {
    APIError: APIError,
    UnauthorizedAccess: UnauthorizedAccess,
    ValidationError: ValidationError,
    UnknownError: UnknownError,
    withErrorContents: withErrorContents,
    printAPIError: printAPIError
};
