// Purescriptify.Component.NewUserRow/index.js
"use strict";
var Control_Applicative = require("../Control.Applicative/index.js");
var Data_Function = require("../Data.Function/index.js");
var Data_Maybe = require("../Data.Maybe/index.js");
var Effect_Class = require("../Effect.Class/index.js");
var Purescriptify_Utilities_Email = require("../Purescriptify.Utilities.Email/index.js");
var React_Basic = require("../React.Basic/index.js");
var React_Basic_DOM = require("../React.Basic.DOM/index.js");
var React_Basic_DOM_Events = require("../React.Basic.DOM.Events/index.js");
var React_Basic_DOM_Generated = require("../React.Basic.DOM.Generated/index.js");
var React_Basic_Events = require("../React.Basic.Events/index.js");
var React_Basic_Hooks = require("../React.Basic.Hooks/index.js");
var React_Basic_Hooks_Internal = require("../React.Basic.Hooks.Internal/index.js");
var Type_Equality = require("../Type.Equality/index.js");
var mkNewUserRow = function (dictMonadEffect) {
    return Effect_Class.liftEffect(dictMonadEffect)(React_Basic_Hooks.component("NewUserRow")(function (v) {
        return React_Basic_Hooks_Internal.bind(React_Basic_Hooks_Internal.ixBindRender)(React_Basic_Hooks.useState(""))(function (v1) {
            return React_Basic_Hooks_Internal.bind(React_Basic_Hooks_Internal.ixBindRender)(React_Basic_Hooks.useState(""))(function (v2) {
                return React_Basic_Hooks_Internal.bind(React_Basic_Hooks_Internal.ixBindRender)(React_Basic_Hooks.useState(Data_Maybe.Nothing.value))(function (v3) {
                    var clearFields = function __do() {
                        v1.value1(Data_Function["const"](""))();
                        v2.value1(Data_Function["const"](""))();
                        return v3.value1(Data_Function["const"](Data_Maybe.Nothing.value))();
                    };
                    var handleSubmit = (function () {
                        var v4 = Purescriptify_Utilities_Email.mkEmail(v2.value0);
                        if (v4 instanceof Data_Maybe.Just) {
                            return function __do() {
                                v3.value1(Data_Function["const"](Data_Maybe.Nothing.value))();
                                clearFields();
                                return v.onSubmit({
                                    username: v1.value0,
                                    email: v4.value0
                                })();
                            };
                        };
                        if (v4 instanceof Data_Maybe.Nothing) {
                            return v3.value1(Data_Function["const"](new Data_Maybe.Just("Invalid email.")));
                        };
                        throw new Error("Failed pattern match at Purescriptify.Component.NewUserRow (line 37, column 13 - line 42, column 76): " + [ v4.constructor.name ]);
                    })();
                    return Control_Applicative.pure(React_Basic_Hooks_Internal.applicativeRender(Type_Equality.refl))(React_Basic_DOM_Generated.div()({
                        className: "mt-8 space-y-4 flex flex-col bg-white rounded-lg border-blue-100 p-2",
                        children: [ (function () {
                            if (v3.value0 instanceof Data_Maybe.Just) {
                                return React_Basic_DOM_Generated.div()({
                                    className: "bg-red-4 text-white p-2 rounded-lg",
                                    children: [ React_Basic_DOM.text(v3.value0.value0) ]
                                });
                            };
                            if (v3.value0 instanceof Data_Maybe.Nothing) {
                                return React_Basic.empty;
                            };
                            throw new Error("Failed pattern match at Purescriptify.Component.NewUserRow (line 47, column 21 - line 53, column 45): " + [ v3.value0.constructor.name ]);
                        })(), React_Basic_DOM_Generated.div()({
                            className: "space-x-4 flex flex-row items-center",
                            children: [ React_Basic_DOM_Generated.input()({
                                type: "text",
                                className: "p-2 border border-slate-100 rounded-md",
                                placeholder: "Username",
                                value: v1.value0,
                                onChange: React_Basic_Events.handler(React_Basic_DOM_Events.targetValue)(function (val) {
                                    return v1.value1(Data_Function["const"](Data_Maybe.fromMaybe("")(val)));
                                })
                            }), React_Basic_DOM_Generated.input()({
                                type: "text",
                                className: "p-2 border border-slate-100 rounded-md",
                                placeholder: "Email",
                                value: v2.value0,
                                onChange: React_Basic_Events.handler(React_Basic_DOM_Events.targetValue)(function (val) {
                                    return v2.value1(Data_Function["const"](Data_Maybe.fromMaybe("")(val)));
                                })
                            }), React_Basic_DOM_Generated.button()({
                                className: "p-2 bg-gold-4 rounded-md text-gray-9",
                                onClick: React_Basic_Events.handler_(handleSubmit),
                                children: [ React_Basic_DOM.text("Add User") ]
                            }) ]
                        }) ]
                    }));
                });
            });
        });
    }));
};
module.exports = {
    mkNewUserRow: mkNewUserRow
};
