// Purescriptify.Capability.Now/index.js
"use strict";
var Control_Monad_Trans_Class = require("../Control.Monad.Trans.Class/index.js");
var React_Halo_Internal_Control = require("../React.Halo.Internal.Control/index.js");
var nowTime = function (dict) {
    return dict.nowTime;
};
var nowDateTime = function (dict) {
    return dict.nowDateTime;
};
var nowDate = function (dict) {
    return dict.nowDate;
};
var now = function (dict) {
    return dict.now;
};
var nowHalogenM = function (dictNow) {
    return {
        now: Control_Monad_Trans_Class.lift(React_Halo_Internal_Control.monadTransHaloM)(dictNow.Monad0())(now(dictNow)),
        nowDate: Control_Monad_Trans_Class.lift(React_Halo_Internal_Control.monadTransHaloM)(dictNow.Monad0())(nowDate(dictNow)),
        nowTime: Control_Monad_Trans_Class.lift(React_Halo_Internal_Control.monadTransHaloM)(dictNow.Monad0())(nowTime(dictNow)),
        nowDateTime: Control_Monad_Trans_Class.lift(React_Halo_Internal_Control.monadTransHaloM)(dictNow.Monad0())(nowDateTime(dictNow)),
        Monad0: function () {
            return React_Halo_Internal_Control.monadHaloM;
        }
    };
};
module.exports = {
    now: now,
    nowDate: nowDate,
    nowDateTime: nowDateTime,
    nowTime: nowTime,
    nowHalogenM: nowHalogenM
};
