// Data.MediaType.Common/index.js
"use strict";

// | The `text/xml` media type.
var textXML = "text/xml";

// | The `text/plain` media type.
var textPlain = "text/plain";

// | The `text/html` media type.
var textHTML = "text/html";

// | The `text/csv` media type.
var textCSV = "text/csv";

// | The `text/css` media type.
var textCSS = "text/css";

// | The `multipart/form-data` media type.
var multipartFormData = "multipart/form-data";

// | The `image/png` media type.
var imagePNG = "image/png";

// | The `image/jpeg` media type.
var imageJPEG = "image/jpeg";

// | The `image/gif` media type.
var imageGIF = "image/gif";

// | The `application/xml` media type.
var applicationXML = "application/xml";

// | The `application/octet-stream` media type.
var applicationOctetStream = "application/octet-stream";

// | The `application/javascript` media type.
var applicationJavascript = "application/javascript";

// | The `application/json` media type.
var applicationJSON = "application/json";

// | The `application/x-www-form-urlencoded` media type.
var applicationFormURLEncoded = "application/x-www-form-urlencoded";
module.exports = {
    applicationFormURLEncoded: applicationFormURLEncoded,
    applicationJSON: applicationJSON,
    applicationJavascript: applicationJavascript,
    applicationOctetStream: applicationOctetStream,
    applicationXML: applicationXML,
    imageGIF: imageGIF,
    imageJPEG: imageJPEG,
    imagePNG: imagePNG,
    multipartFormData: multipartFormData,
    textCSV: textCSV,
    textHTML: textHTML,
    textPlain: textPlain,
    textXML: textXML,
    textCSS: textCSS
};
