// Type.RowList/index.js
"use strict";
var Type_Data_RowList = require("../Type.Data.RowList/index.js");
var rowListSetImpl = function (dictTypeEquals) {
    return function (dictTypeEquals1) {
        return function () {
            return {};
        };
    };
};
var rowListRemoveNil = {};
var rowListRemoveCons = function () {
    return function () {
        return function () {
            return {};
        };
    };
};
var rowListNubNil = {};
var rowListNubCons = function (dictTypeEquals) {
    return function (dictTypeEquals1) {
        return function (dictTypeEquals2) {
            return function () {
                return function () {
                    return {};
                };
            };
        };
    };
};
var rowListAppendNil = function (dictTypeEquals) {
    return {};
};
var rowListAppendCons = function () {
    return function (dictTypeEquals) {
        return {};
    };
};
var listToRowNil = {};
var listToRowCons = function () {
    return function () {
        return {};
    };
};
module.exports = {
    listToRowNil: listToRowNil,
    listToRowCons: listToRowCons,
    rowListRemoveNil: rowListRemoveNil,
    rowListRemoveCons: rowListRemoveCons,
    rowListSetImpl: rowListSetImpl,
    rowListNubNil: rowListNubNil,
    rowListNubCons: rowListNubCons,
    rowListAppendNil: rowListAppendNil,
    rowListAppendCons: rowListAppendCons,
    RLProxy: Type_Data_RowList.RLProxy
};
