// Data.Lens.Index/index.js
"use strict";
var Data_Array = require("../Data.Array/index.js");
var Data_Array_NonEmpty = require("../Data.Array.NonEmpty/index.js");
var Data_Either = require("../Data.Either/index.js");
var Data_Eq = require("../Data.Eq/index.js");
var Data_Lens_AffineTraversal = require("../Data.Lens.AffineTraversal/index.js");
var Data_Lens_Iso_Newtype = require("../Data.Lens.Iso.Newtype/index.js");
var Data_Lens_Lens = require("../Data.Lens.Lens/index.js");
var Data_Lens_Prism_Maybe = require("../Data.Lens.Prism.Maybe/index.js");
var Data_List = require("../Data.List/index.js");
var Data_Map_Internal = require("../Data.Map.Internal/index.js");
var Data_Maybe = require("../Data.Maybe/index.js");
var Data_Set = require("../Data.Set/index.js");
var Data_Unit = require("../Data.Unit/index.js");
var Foreign_Object = require("../Foreign.Object/index.js");
var ix = function (dict) {
    return dict.ix;
};
var indexSet = function (dictOrd) {
    return {
        ix: function (x) {
            return function (dictStrong) {
                return function (dictChoice) {
                    var set = function (xs) {
                        return function (v) {
                            return xs;
                        };
                    };
                    var pre = function (xs) {
                        var $27 = Data_Set.member(dictOrd)(x)(xs);
                        if ($27) {
                            return new Data_Either.Right(Data_Unit.unit);
                        };
                        return new Data_Either.Left(xs);
                    };
                    return Data_Lens_AffineTraversal.affineTraversal(set)(pre)(dictStrong)(dictChoice);
                };
            };
        }
    };
};
var indexNonEmptyArray = {
    ix: function (n) {
        return function (dictStrong) {
            return function (dictChoice) {
                var set = function (s) {
                    return function (b) {
                        return Data_Maybe.fromMaybe(s)(Data_Array_NonEmpty.updateAt(n)(b)(s));
                    };
                };
                var pre = function (s) {
                    return Data_Maybe.maybe(new Data_Either.Left(s))(Data_Either.Right.create)(Data_Array_NonEmpty.index(s)(n));
                };
                return Data_Lens_AffineTraversal.affineTraversal(set)(pre)(dictStrong)(dictChoice);
            };
        };
    }
};
var indexMaybe = {
    ix: function (v) {
        return function (dictStrong) {
            return function (dictChoice) {
                return Data_Lens_Prism_Maybe["_Just"](dictChoice);
            };
        };
    }
};
var indexMap = function (dictOrd) {
    return {
        ix: function (k) {
            return function (dictStrong) {
                return function (dictChoice) {
                    var set = function (s) {
                        return function (b) {
                            return Data_Map_Internal.update(dictOrd)(function (v) {
                                return new Data_Maybe.Just(b);
                            })(k)(s);
                        };
                    };
                    var pre = function (s) {
                        return Data_Maybe.maybe(new Data_Either.Left(s))(Data_Either.Right.create)(Data_Map_Internal.lookup(dictOrd)(k)(s));
                    };
                    return Data_Lens_AffineTraversal.affineTraversal(set)(pre)(dictStrong)(dictChoice);
                };
            };
        }
    };
};
var indexList = {
    ix: function (n) {
        return function (dictStrong) {
            return function (dictChoice) {
                var set = function (s) {
                    return function (b) {
                        return Data_Maybe.fromMaybe(s)(Data_List.updateAt(n)(b)(s));
                    };
                };
                var pre = function (s) {
                    return Data_Maybe.maybe(new Data_Either.Left(s))(Data_Either.Right.create)(Data_List.index(s)(n));
                };
                return Data_Lens_AffineTraversal.affineTraversal(set)(pre)(dictStrong)(dictChoice);
            };
        };
    }
};
var indexIdentity = {
    ix: function (v) {
        return function (dictStrong) {
            return function (dictChoice) {
                return Data_Lens_Iso_Newtype["_Newtype"]()()(dictChoice.Profunctor0());
            };
        };
    }
};
var indexForeignObject = {
    ix: function (k) {
        return function (dictStrong) {
            return function (dictChoice) {
                var set = function (s) {
                    return function (b) {
                        return Foreign_Object.update(function (v) {
                            return new Data_Maybe.Just(b);
                        })(k)(s);
                    };
                };
                var pre = function (s) {
                    return Data_Maybe.maybe(new Data_Either.Left(s))(Data_Either.Right.create)(Foreign_Object.lookup(k)(s));
                };
                return Data_Lens_AffineTraversal.affineTraversal(set)(pre)(dictStrong)(dictChoice);
            };
        };
    }
};
var indexFn = function (dictEq) {
    return {
        ix: function (i) {
            return function (dictStrong) {
                return function (dictChoice) {
                    return Data_Lens_Lens.lens(function (f) {
                        return f(i);
                    })(function (f) {
                        return function (a) {
                            return function (j) {
                                var $28 = Data_Eq.eq(dictEq)(i)(j);
                                if ($28) {
                                    return a;
                                };
                                return f(j);
                            };
                        };
                    })(dictStrong);
                };
            };
        }
    };
};
var indexArray = {
    ix: function (n) {
        return function (dictStrong) {
            return function (dictChoice) {
                var set = function (s) {
                    return function (b) {
                        return Data_Maybe.fromMaybe(s)(Data_Array.updateAt(n)(b)(s));
                    };
                };
                var pre = function (s) {
                    return Data_Maybe.maybe(new Data_Either.Left(s))(Data_Either.Right.create)(Data_Array.index(s)(n));
                };
                return Data_Lens_AffineTraversal.affineTraversal(set)(pre)(dictStrong)(dictChoice);
            };
        };
    }
};
module.exports = {
    ix: ix,
    indexFn: indexFn,
    indexMaybe: indexMaybe,
    indexIdentity: indexIdentity,
    indexArray: indexArray,
    indexNonEmptyArray: indexNonEmptyArray,
    indexList: indexList,
    indexSet: indexSet,
    indexMap: indexMap,
    indexForeignObject: indexForeignObject
};
