// Node.FS.Stats/index.js
"use strict";
var $foreign = require("./foreign.js");
var Data_JSDate = require("../Data.JSDate/index.js");
var Data_Maybe = require("../Data.Maybe/index.js");

// | Stats wrapper to provide a usable interface to the underlying properties and methods.
var Stats = (function () {
    function Stats(value0) {
        this.value0 = value0;
    };
    Stats.create = function (value0) {
        return new Stats(value0);
    };
    return Stats;
})();
var statusChangedTime = function (v) {
    return Data_Maybe.fromJust()(Data_JSDate.toDateTime(v.value0.ctime));
};
var showStats = {
    show: function (v) {
        return "Stats " + $foreign.showStatsObj(v.value0);
    }
};
var modifiedTime = function (v) {
    return Data_Maybe.fromJust()(Data_JSDate.toDateTime(v.value0.mtime));
};
var isSymbolicLink = function (v) {
    return $foreign.statsMethod("isSymbolicLink", v.value0);
};
var isSocket = function (v) {
    return $foreign.statsMethod("isSocket", v.value0);
};
var isFile = function (v) {
    return $foreign.statsMethod("isFile", v.value0);
};
var isFIFO = function (v) {
    return $foreign.statsMethod("isFIFO", v.value0);
};
var isDirectory = function (v) {
    return $foreign.statsMethod("isDirectory", v.value0);
};
var isCharacterDevice = function (v) {
    return $foreign.statsMethod("isCharacterDevice", v.value0);
};
var isBlockDevice = function (v) {
    return $foreign.statsMethod("isBlockDevice", v.value0);
};
var accessedTime = function (v) {
    return Data_Maybe.fromJust()(Data_JSDate.toDateTime(v.value0.atime));
};
module.exports = {
    Stats: Stats,
    isFile: isFile,
    isDirectory: isDirectory,
    isBlockDevice: isBlockDevice,
    isCharacterDevice: isCharacterDevice,
    isFIFO: isFIFO,
    isSocket: isSocket,
    isSymbolicLink: isSymbolicLink,
    accessedTime: accessedTime,
    modifiedTime: modifiedTime,
    statusChangedTime: statusChangedTime,
    showStats: showStats
};
