// Purescriptify.Data.Route/index.js
"use strict";
var Data_Generic_Rep = require("../Data.Generic.Rep/index.js");
var Routing_Duplex = require("../Routing.Duplex/index.js");
var Routing_Duplex_Generic = require("../Routing.Duplex.Generic/index.js");
var Home = (function () {
    function Home() {

    };
    Home.value = new Home();
    return Home;
})();
var $$Error = (function () {
    function $$Error() {

    };
    $$Error.value = new $$Error();
    return $$Error;
})();
var genericRoute_ = {
    to: function (x) {
        if (x instanceof Data_Generic_Rep.Inl) {
            return Home.value;
        };
        if (x instanceof Data_Generic_Rep.Inr) {
            return $$Error.value;
        };
        throw new Error("Failed pattern match at Purescriptify.Data.Route (line 13, column 1 - line 13, column 32): " + [ x.constructor.name ]);
    },
    from: function (x) {
        if (x instanceof Home) {
            return new Data_Generic_Rep.Inl(Data_Generic_Rep.NoArguments.value);
        };
        if (x instanceof $$Error) {
            return new Data_Generic_Rep.Inr(Data_Generic_Rep.NoArguments.value);
        };
        throw new Error("Failed pattern match at Purescriptify.Data.Route (line 13, column 1 - line 13, column 32): " + [ x.constructor.name ]);
    }
};
var eqRoute = {
    eq: function (x) {
        return function (y) {
            if (x instanceof Home && y instanceof Home) {
                return true;
            };
            if (x instanceof $$Error && y instanceof $$Error) {
                return true;
            };
            return false;
        };
    }
};
var routeCodec = Routing_Duplex["default"]($$Error.value)(Routing_Duplex.root(Routing_Duplex_Generic.sum(genericRoute_)(Routing_Duplex_Generic.gRouteSum(Routing_Duplex_Generic.gRouteConstructor({
    reflectSymbol: function () {
        return "Home";
    }
})()(Routing_Duplex_Generic.gRouteNoArguments))(Routing_Duplex_Generic.gRouteConstructor({
    reflectSymbol: function () {
        return "Error";
    }
})()(Routing_Duplex_Generic.gRouteNoArguments)))({
    Home: Routing_Duplex_Generic.noArgs,
    "Error": Routing_Duplex_Generic.noArgs
})));
module.exports = {
    Home: Home,
    "Error": $$Error,
    routeCodec: routeCodec,
    genericRoute_: genericRoute_,
    eqRoute: eqRoute
};
