// URI.Host.RegName/index.js
"use strict";
var Control_Alt = require("../Control.Alt/index.js");
var Data_Array_NonEmpty = require("../Data.Array.NonEmpty/index.js");
var Data_Array_NonEmpty_Internal = require("../Data.Array.NonEmpty.Internal/index.js");
var Data_Functor = require("../Data.Functor/index.js");
var Data_Identity = require("../Data.Identity/index.js");
var Data_Show = require("../Data.Show/index.js");
var Data_String_NonEmpty_CodeUnits = require("../Data.String.NonEmpty.CodeUnits/index.js");
var Data_String_NonEmpty_Internal = require("../Data.String.NonEmpty.Internal/index.js");
var Text_Parsing_Parser = require("../Text.Parsing.Parser/index.js");
var URI_Common = require("../URI.Common/index.js");

// | The reg-name variation of the host part of a URI. A reg-name is probably
// | more commonly referred to as just a host name or domain name (but it is
// | actually a name, rather than an IP address).
var RegName = function (x) {
    return x;
};

// | Returns the string value for the reg-name without percent-decoding. Only
// | "unsafe" in the sense that values this produces may need further decoding,
// | the name is more for symmetry with the `fromString`/`unsafeFromString`
// | pairing.
var unsafeToString = function (v) {
    return v;
};

// | Constructs a query value from a string directly - no percent-encoding
// | will be applied. This is useful when using a custom encoding scheme for
// | the query, to prevent double-encoding.
var unsafeFromString = RegName;

// | Returns the string value for a reg-name, percent-decoding any characters
// | that require it.
// |
// | ``` purescript
// | toString (unsafeFromString "foo.com") = "foo.com"
// | toString (unsafeFromString "foo%3Abar") = "foo:bar"
// | ```
var toString = function (v) {
    return URI_Common["decodeURIComponent'"](v);
};
var showRegName = {
    show: function (v) {
        return "(RegName.unsafeFromString " + (Data_Show.show(Data_String_NonEmpty_Internal.showNonEmptyString)(v) + ")");
    }
};
var semigroupRegName = Data_String_NonEmpty_Internal.semigroupNonEmptyString;

// | The supported reg-name characters, excluding percent-encodings.
var regNameChar = Control_Alt.alt(Text_Parsing_Parser.altParserT(Data_Identity.monadIdentity))(URI_Common.unreserved)(URI_Common.subDelims);

// | A printer for reg-names.
var print = function ($6) {
    return Data_String_NonEmpty_Internal.toString(unsafeToString($6));
};

// | A parser for reg-names.
var parser = (function () {
    var p = Control_Alt.alt(Text_Parsing_Parser.altParserT(Data_Identity.monadIdentity))(URI_Common.pctEncoded)(Data_Functor.map(Text_Parsing_Parser.functorParserT(Data_Identity.functorIdentity))(Data_String_NonEmpty_CodeUnits.singleton)(regNameChar));
    return Data_Functor.map(Text_Parsing_Parser.functorParserT(Data_Identity.functorIdentity))((function () {
        var $7 = Data_String_NonEmpty_Internal.join1With(Data_Array_NonEmpty_Internal.foldable1NonEmptyArray)("");
        return function ($8) {
            return RegName($7($8));
        };
    })())(Data_Array_NonEmpty.some(Text_Parsing_Parser.alternativeParserT(Data_Identity.monadIdentity))(Text_Parsing_Parser.lazyParserT)(p));
})();
var ordRegName = Data_String_NonEmpty_Internal.ordNonEmptyString;

// | Constructs a reg-name value from a string, percent-encoding any characters
// | that require it. Note that running this on a string that has already had
// | percent-encoding applied will double-encode it, for those situations use
// | `unsafeFromString` instead.
// |
// | ``` purescript
// | fromString "foo.com" = unsafeFromString "foo.com"
// | fromString "foo:bar" = unsafeFromString "foo%3Abar"
// | fromString "foo%3Abar" = unsafeFromString "foo%253Abar"
// | ```
var fromString = (function () {
    var $9 = URI_Common["printEncoded'"](regNameChar);
    return function ($10) {
        return RegName($9($10));
    };
})();
var eqRegName = Data_String_NonEmpty_Internal.eqNonEmptyString;
module.exports = {
    fromString: fromString,
    toString: toString,
    unsafeFromString: unsafeFromString,
    unsafeToString: unsafeToString,
    parser: parser,
    print: print,
    regNameChar: regNameChar,
    eqRegName: eqRegName,
    ordRegName: ordRegName,
    semigroupRegName: semigroupRegName,
    showRegName: showRegName
};
