// Tidy.Precedence/index.js
"use strict";
var Control_Applicative = require("../Control.Applicative/index.js");
var Control_Bind = require("../Control.Bind/index.js");
var Data_Array_NonEmpty = require("../Data.Array.NonEmpty/index.js");
var Data_Array_NonEmpty_Internal = require("../Data.Array.NonEmpty.Internal/index.js");
var Data_Eq = require("../Data.Eq/index.js");
var Data_Foldable = require("../Data.Foldable/index.js");
var Data_Function = require("../Data.Function/index.js");
var Data_Functor = require("../Data.Functor/index.js");
var Data_HeytingAlgebra = require("../Data.HeytingAlgebra/index.js");
var Data_List_Types = require("../Data.List.Types/index.js");
var Data_Map_Internal = require("../Data.Map.Internal/index.js");
var Data_Maybe = require("../Data.Maybe/index.js");
var Data_Monoid = require("../Data.Monoid/index.js");
var Data_Ord = require("../Data.Ord/index.js");
var Data_Ordering = require("../Data.Ordering/index.js");
var Data_Semigroup = require("../Data.Semigroup/index.js");
var Data_Tuple = require("../Data.Tuple/index.js");
var PureScript_CST_Types = require("../PureScript.CST.Types/index.js");
var OperatorType = (function () {
    function OperatorType() {

    };
    OperatorType.value = new OperatorType();
    return OperatorType;
})();
var OperatorValue = (function () {
    function OperatorValue() {

    };
    OperatorValue.value = new OperatorValue();
    return OperatorValue;
})();
var QualifiedOperator = (function () {
    function QualifiedOperator(value0, value1, value2) {
        this.value0 = value0;
        this.value1 = value1;
        this.value2 = value2;
    };
    QualifiedOperator.create = function (value0) {
        return function (value1) {
            return function (value2) {
                return new QualifiedOperator(value0, value1, value2);
            };
        };
    };
    return QualifiedOperator;
})();
var OpList = (function () {
    function OpList(value0, value1, value2) {
        this.value0 = value0;
        this.value1 = value1;
        this.value2 = value2;
    };
    OpList.create = function (value0) {
        return function (value1) {
            return function (value2) {
                return new OpList(value0, value1, value2);
            };
        };
    };
    return OpList;
})();
var OpPure = (function () {
    function OpPure(value0) {
        this.value0 = value0;
    };
    OpPure.create = function (value0) {
        return new OpPure(value0);
    };
    return OpPure;
})();
var OpHead = (function () {
    function OpHead(value0) {
        this.value0 = value0;
    };
    OpHead.create = function (value0) {
        return new OpHead(value0);
    };
    return OpHead;
})();
var OpPrec = (function () {
    function OpPrec(value0, value1, value2) {
        this.value0 = value0;
        this.value1 = value1;
        this.value2 = value2;
    };
    OpPrec.create = function (value0) {
        return function (value1) {
            return function (value2) {
                return new OpPrec(value0, value1, value2);
            };
        };
    };
    return OpPrec;
})();
var snoc = function (prevOps) {
    return function (nextPrec) {
        return function (nextOps) {
            var v = Data_Array_NonEmpty.unsnoc(prevOps);
            return Data_Array_NonEmpty["snoc'"](v.init)(new Data_Tuple.Tuple(v.last.value0, new OpList(v.last.value1, nextPrec, nextOps)));
        };
    };
};
var unwind = (function () {
    var go = function ($copy_prec) {
        return function ($copy_ops) {
            return function ($copy_v) {
                var $tco_var_prec = $copy_prec;
                var $tco_var_ops = $copy_ops;
                var $tco_done = false;
                var $tco_result;
                function $tco_loop(prec, ops, v) {
                    if (v instanceof OpHead) {
                        $tco_done = true;
                        return new OpList(v.value0, prec, ops);
                    };
                    if (v instanceof OpPrec) {
                        $tco_var_prec = v.value1;
                        $tco_var_ops = snoc(v.value2)(prec)(ops);
                        $copy_v = v.value0;
                        return;
                    };
                    throw new Error("Failed pattern match at Tidy.Precedence (line 101, column 17 - line 104, column 46): " + [ v.constructor.name ]);
                };
                while (!$tco_done) {
                    $tco_result = $tco_loop($tco_var_prec, $tco_var_ops, $copy_v);
                };
                return $tco_result;
            };
        };
    };
    return function (v) {
        if (v instanceof OpHead) {
            return v.value0;
        };
        if (v instanceof OpPrec) {
            return go(v.value1)(v.value2)(v.value0);
        };
        throw new Error("Failed pattern match at Tidy.Precedence (line 97, column 10 - line 99, column 41): " + [ v.constructor.name ]);
    };
})();
var push = function ($copy_stk) {
    return function ($copy_chs) {
        var $tco_var_stk = $copy_stk;
        var $tco_done = false;
        var $tco_result;
        function $tco_loop(stk, chs) {
            if (chs instanceof Data_List_Types.Nil) {
                $tco_done = true;
                return stk;
            };
            if (chs instanceof Data_List_Types.Cons && chs.value1 instanceof Data_List_Types.Nil) {
                if (stk instanceof OpHead) {
                    $tco_done = true;
                    return new OpPrec(stk, chs.value0.value0, chs.value0.value1);
                };
                if (stk instanceof OpPrec) {
                    var v = Data_Ord.compare(Data_Ord.ordInt)(chs.value0.value0)(stk.value1);
                    if (v instanceof Data_Ordering.EQ) {
                        $tco_done = true;
                        return new OpPrec(stk.value0, stk.value1, Data_Semigroup.append(Data_Array_NonEmpty_Internal.semigroupNonEmptyArray)(stk.value2)(chs.value0.value1));
                    };
                    if (v instanceof Data_Ordering.GT) {
                        $tco_done = true;
                        return new OpPrec(stk, chs.value0.value0, chs.value0.value1);
                    };
                    if (v instanceof Data_Ordering.LT) {
                        $tco_var_stk = stk.value0;
                        $copy_chs = new Data_List_Types.Cons(new Data_Tuple.Tuple(stk.value1, stk.value2), chs);
                        return;
                    };
                    throw new Error("Failed pattern match at Tidy.Precedence (line 82, column 9 - line 85, column 60): " + [ v.constructor.name ]);
                };
                throw new Error("Failed pattern match at Tidy.Precedence (line 78, column 5 - line 85, column 60): " + [ stk.constructor.name ]);
            };
            if (chs instanceof Data_List_Types.Cons) {
                if (stk instanceof OpHead) {
                    $tco_var_stk = new OpHead(new OpList(stk.value0, chs.value0.value0, chs.value0.value1));
                    $copy_chs = chs.value1;
                    return;
                };
                if (stk instanceof OpPrec) {
                    var v = Data_Ord.compare(Data_Ord.ordInt)(chs.value0.value0)(stk.value1);
                    if (v instanceof Data_Ordering.EQ) {
                        $tco_var_stk = new OpPrec(stk.value0, stk.value1, Data_Semigroup.append(Data_Array_NonEmpty_Internal.semigroupNonEmptyArray)(stk.value2)(chs.value0.value1));
                        $copy_chs = chs.value1;
                        return;
                    };
                    if (v instanceof Data_Ordering.GT) {
                        $tco_var_stk = new OpPrec(stk.value0, stk.value1, snoc(stk.value2)(chs.value0.value0)(chs.value0.value1));
                        $copy_chs = chs.value1;
                        return;
                    };
                    if (v instanceof Data_Ordering.LT) {
                        $tco_var_stk = stk.value0;
                        $copy_chs = new Data_List_Types.Cons(new Data_Tuple.Tuple(stk.value1, snoc(stk.value2)(chs.value0.value0)(chs.value0.value1)), chs.value1);
                        return;
                    };
                    throw new Error("Failed pattern match at Tidy.Precedence (line 91, column 9 - line 94, column 85): " + [ v.constructor.name ]);
                };
                throw new Error("Failed pattern match at Tidy.Precedence (line 87, column 5 - line 94, column 85): " + [ stk.constructor.name ]);
            };
            throw new Error("Failed pattern match at Tidy.Precedence (line 75, column 16 - line 94, column 85): " + [ chs.constructor.name ]);
        };
        while (!$tco_done) {
            $tco_result = $tco_loop($tco_var_stk, $copy_chs);
        };
        return $tco_result;
    };
};
var eqOperatorNamespace = {
    eq: function (x) {
        return function (y) {
            if (x instanceof OperatorType && y instanceof OperatorType) {
                return true;
            };
            if (x instanceof OperatorValue && y instanceof OperatorValue) {
                return true;
            };
            return false;
        };
    }
};
var eqQualifiedOperator = {
    eq: function (x) {
        return function (y) {
            return Data_Eq.eq(Data_Maybe.eqMaybe(PureScript_CST_Types.eqModuleName))(x.value0)(y.value0) && Data_Eq.eq(eqOperatorNamespace)(x.value1)(y.value1) && Data_Eq.eq(PureScript_CST_Types.eqOperator)(x.value2)(y.value2);
        };
    }
};
var ordOperatorNamespace = {
    compare: function (x) {
        return function (y) {
            if (x instanceof OperatorType && y instanceof OperatorType) {
                return Data_Ordering.EQ.value;
            };
            if (x instanceof OperatorType) {
                return Data_Ordering.LT.value;
            };
            if (y instanceof OperatorType) {
                return Data_Ordering.GT.value;
            };
            if (x instanceof OperatorValue && y instanceof OperatorValue) {
                return Data_Ordering.EQ.value;
            };
            throw new Error("Failed pattern match at Tidy.Precedence (line 33, column 1 - line 33, column 62): " + [ x.constructor.name, y.constructor.name ]);
        };
    },
    Eq0: function () {
        return eqOperatorNamespace;
    }
};
var insertOperator = function (v) {
    return function (prec) {
        var opKey = new Data_Tuple.Tuple(v.value1, v.value2);
        return Data_Map_Internal.alter(Data_Maybe.ordMaybe(PureScript_CST_Types.ordModuleName))(function (v1) {
            if (v1 instanceof Data_Maybe.Nothing) {
                return new Data_Maybe.Just(Data_Map_Internal.singleton(opKey)(prec));
            };
            if (v1 instanceof Data_Maybe.Just) {
                return new Data_Maybe.Just(Data_Map_Internal.insert(Data_Tuple.ordTuple(ordOperatorNamespace)(PureScript_CST_Types.ordOperator))(opKey)(prec)(v1.value0));
            };
            throw new Error("Failed pattern match at Tidy.Precedence (line 159, column 5 - line 163, column 41): " + [ v1.constructor.name ]);
        })(v.value0);
    };
};
var lookupOperator = function (v) {
    return function (precMap) {
        return Control_Bind.bind(Data_Maybe.bindMaybe)(Data_Map_Internal.lookup(Data_Maybe.ordMaybe(PureScript_CST_Types.ordModuleName))(v.value0)(precMap))(Data_Map_Internal.lookup(Data_Tuple.ordTuple(ordOperatorNamespace)(PureScript_CST_Types.ordOperator))(new Data_Tuple.Tuple(v.value1, v.value2)));
    };
};
var remapOperatorTo = function (newModName) {
    return function (v) {
        return function (precMap) {
            return Data_Maybe.fromMaybe(precMap)(Control_Bind.bind(Data_Maybe.bindMaybe)(lookupOperator(v)(precMap))(function (prec) {
                return Control_Applicative.pure(Data_Maybe.applicativeMaybe)(insertOperator(new QualifiedOperator(newModName, v.value1, v.value2))(prec)(precMap));
            }));
        };
    };
};
var ordQualifiedOperator = {
    compare: function (x) {
        return function (y) {
            var v = Data_Ord.compare(Data_Maybe.ordMaybe(PureScript_CST_Types.ordModuleName))(x.value0)(y.value0);
            if (v instanceof Data_Ordering.LT) {
                return Data_Ordering.LT.value;
            };
            if (v instanceof Data_Ordering.GT) {
                return Data_Ordering.GT.value;
            };
            var v1 = Data_Ord.compare(ordOperatorNamespace)(x.value1)(y.value1);
            if (v1 instanceof Data_Ordering.LT) {
                return Data_Ordering.LT.value;
            };
            if (v1 instanceof Data_Ordering.GT) {
                return Data_Ordering.GT.value;
            };
            return Data_Ord.compare(PureScript_CST_Types.ordOperator)(x.value2)(y.value2);
        };
    },
    Eq0: function () {
        return eqQualifiedOperator;
    }
};
var remapModuleTo = function (newModName) {
    return function (modName) {
        return function (precMap) {
            return Data_Maybe.fromMaybe(precMap)(Control_Bind.bind(Data_Maybe.bindMaybe)(Data_Map_Internal.lookup(Data_Maybe.ordMaybe(PureScript_CST_Types.ordModuleName))(new Data_Maybe.Just(modName))(precMap))(function (ops) {
                return Control_Applicative.pure(Data_Maybe.applicativeMaybe)(Data_Map_Internal.alter(Data_Maybe.ordMaybe(PureScript_CST_Types.ordModuleName))(function (v) {
                    if (v instanceof Data_Maybe.Nothing) {
                        return new Data_Maybe.Just(ops);
                    };
                    if (v instanceof Data_Maybe.Just) {
                        return new Data_Maybe.Just(Data_Map_Internal.union(Data_Tuple.ordTuple(ordOperatorNamespace)(PureScript_CST_Types.ordOperator))(ops)(v.value0));
                    };
                    throw new Error("Failed pattern match at Tidy.Precedence (line 184, column 7 - line 188, column 38): " + [ v.constructor.name ]);
                })(newModName)(precMap));
            }));
        };
    };
};
var remapModuleToHiding = function (dictFoldable) {
    return function (hiding) {
        return function (newModName) {
            return function (modName) {
                return function (precMap) {
                    return Data_Maybe.fromMaybe(precMap)(Control_Bind.bind(Data_Maybe.bindMaybe)(Data_Map_Internal.lookup(Data_Maybe.ordMaybe(PureScript_CST_Types.ordModuleName))(new Data_Maybe.Just(modName))(precMap))(function (ops) {
                        var filteredOps = Data_Map_Internal.filterKeys(Data_Tuple.ordTuple(ordOperatorNamespace)(PureScript_CST_Types.ordOperator))((function () {
                            var $179 = Data_HeytingAlgebra.not(Data_HeytingAlgebra.heytingAlgebraBoolean);
                            var $180 = Data_Function.flip(Data_Foldable.elem(dictFoldable)(eqQualifiedOperator))(hiding);
                            var $181 = Data_Tuple.uncurry(QualifiedOperator.create(new Data_Maybe.Just(modName)));
                            return function ($182) {
                                return $179($180($181($182)));
                            };
                        })())(ops);
                        return Control_Applicative.pure(Data_Maybe.applicativeMaybe)(Data_Map_Internal.alter(Data_Maybe.ordMaybe(PureScript_CST_Types.ordModuleName))(function (v) {
                            if (v instanceof Data_Maybe.Nothing) {
                                return new Data_Maybe.Just(filteredOps);
                            };
                            if (v instanceof Data_Maybe.Just) {
                                return new Data_Maybe.Just(Data_Map_Internal.union(Data_Tuple.ordTuple(ordOperatorNamespace)(PureScript_CST_Types.ordOperator))(filteredOps)(v.value0));
                            };
                            throw new Error("Failed pattern match at Tidy.Precedence (line 205, column 7 - line 209, column 46): " + [ v.constructor.name ]);
                        })(newModName)(precMap));
                    }));
                };
            };
        };
    };
};
var remapOperators = (function () {
    var goImport = function (modName) {
        return function (v) {
            if (v instanceof PureScript_CST_Types.ImportOp) {
                return [ new QualifiedOperator(new Data_Maybe.Just(modName), OperatorValue.value, v.value0.name) ];
            };
            if (v instanceof PureScript_CST_Types.ImportTypeOp) {
                return [ new QualifiedOperator(new Data_Maybe.Just(modName), OperatorType.value, v.value1.name) ];
            };
            return [  ];
        };
    };
    var goImportDecl = function (precMap) {
        return function (v) {
            var newModName = Data_Functor.map(Data_Maybe.functorMaybe)(function (v1) {
                return v1.value1.name;
            })(v.qualified);
            if (v.names instanceof Data_Maybe.Nothing) {
                return remapModuleTo(newModName)(v.module.name)(precMap);
            };
            if (v.names instanceof Data_Maybe.Just) {
                var impOps = Data_Semigroup.append(Data_Semigroup.semigroupArray)(goImport(v.module.name)(v.names.value0.value1.value.head))(Data_Foldable.foldMap(Data_Foldable.foldableArray)(Data_Monoid.monoidArray)((function () {
                    var $183 = goImport(v.module.name);
                    return function ($184) {
                        return $183(Data_Tuple.snd($184));
                    };
                })())(v.names.value0.value1.value.tail));
                var $136 = Data_Maybe.isJust(v.names.value0.value0);
                if ($136) {
                    return remapModuleToHiding(Data_Foldable.foldableArray)(impOps)(newModName)(v.module.name)(precMap);
                };
                return Data_Foldable.foldl(Data_Foldable.foldableArray)(Data_Function.flip(remapOperatorTo(newModName)))(precMap)(impOps);
            };
            throw new Error("Failed pattern match at Tidy.Precedence (line 121, column 5 - line 132, column 67): " + [ v.names.constructor.name ]);
        };
    };
    var goDecl = function (modName) {
        return function (precMap) {
            return function (v) {
                if (v instanceof PureScript_CST_Types.DeclFixity) {
                    if (v.value0.operator instanceof PureScript_CST_Types.FixityValue) {
                        return insertOperator(new QualifiedOperator(new Data_Maybe.Just(modName), OperatorValue.value, v.value0.operator.value2.name))(v.value0.prec.value1)(insertOperator(new QualifiedOperator(Data_Maybe.Nothing.value, OperatorValue.value, v.value0.operator.value2.name))(v.value0.prec.value1)(precMap));
                    };
                    if (v.value0.operator instanceof PureScript_CST_Types.FixityType) {
                        return insertOperator(new QualifiedOperator(new Data_Maybe.Just(modName), OperatorType.value, v.value0.operator.value3.name))(v.value0.prec.value1)(insertOperator(new QualifiedOperator(Data_Maybe.Nothing.value, OperatorType.value, v.value0.operator.value3.name))(v.value0.prec.value1)(precMap));
                    };
                    throw new Error("Failed pattern match at Tidy.Precedence (line 144, column 7 - line 152, column 85): " + [ v.value0.operator.constructor.name ]);
                };
                return precMap;
            };
        };
    };
    var goModule = function (precMap) {
        return function (v) {
            return Data_Foldable.foldl(Data_Foldable.foldableArray)(goDecl(v.header.name.name))(Data_Foldable.foldl(Data_Foldable.foldableArray)(goImportDecl)(precMap)(v.header.imports))(v.body.decls);
        };
    };
    return goModule;
})();
var defaultPrecedence = 10;
var toOperatorTree = function (precMap) {
    return function (getOperator) {
        return function (init) {
            var go = function (stk) {
                return function (v) {
                    var v1 = getOperator(v.value0);
                    var prec = Data_Maybe.fromMaybe(defaultPrecedence)(Control_Bind.bindFlipped(Data_Maybe.bindMaybe)(Data_Map_Internal.lookup(Data_Tuple.ordTuple(ordOperatorNamespace)(PureScript_CST_Types.ordOperator))(new Data_Tuple.Tuple(v1.value1, v1.value2)))(Data_Map_Internal.lookup(Data_Maybe.ordMaybe(PureScript_CST_Types.ordModuleName))(v1.value0)(precMap)));
                    var opCh = Control_Applicative.pure(Data_Array_NonEmpty_Internal.applicativeNonEmptyArray)(new Data_Tuple.Tuple(v.value0, new OpPure(v.value1)));
                    return push(stk)(Control_Applicative.pure(Data_List_Types.applicativeList)(new Data_Tuple.Tuple(prec, opCh)));
                };
            };
            var $185 = Data_Foldable.foldl(Data_Array_NonEmpty_Internal.foldableNonEmptyArray)(go)(new OpHead(new OpPure(init)));
            return function ($186) {
                return unwind($185($186));
            };
        };
    };
};
module.exports = {
    QualifiedOperator: QualifiedOperator,
    OperatorType: OperatorType,
    OperatorValue: OperatorValue,
    OpList: OpList,
    OpPure: OpPure,
    defaultPrecedence: defaultPrecedence,
    toOperatorTree: toOperatorTree,
    remapOperators: remapOperators,
    insertOperator: insertOperator,
    lookupOperator: lookupOperator,
    remapOperatorTo: remapOperatorTo,
    remapModuleTo: remapModuleTo,
    remapModuleToHiding: remapModuleToHiding,
    eqOperatorNamespace: eqOperatorNamespace,
    ordOperatorNamespace: ordOperatorNamespace,
    eqQualifiedOperator: eqQualifiedOperator,
    ordQualifiedOperator: ordQualifiedOperator
};
