// Purescriptify.Component.Routing/index.js
"use strict";
var Control_Applicative = require("../Control.Applicative/index.js");
var Control_Bind = require("../Control.Bind/index.js");
var Control_Category = require("../Control.Category/index.js");
var Data_Either = require("../Data.Either/index.js");
var Data_Foldable = require("../Data.Foldable/index.js");
var Data_Function = require("../Data.Function/index.js");
var Data_Tuple = require("../Data.Tuple/index.js");
var Data_Unit = require("../Data.Unit/index.js");
var Effect = require("../Effect/index.js");
var Effect_Ref = require("../Effect.Ref/index.js");
var Halogen_Subscription = require("../Halogen.Subscription/index.js");
var Purescriptify_Data_Route = require("../Purescriptify.Data.Route/index.js");
var React_Basic = require("../React.Basic/index.js");
var React_Basic_Hooks = require("../React.Basic.Hooks/index.js");
var React_Basic_Hooks_Internal = require("../React.Basic.Hooks.Internal/index.js");
var Routing_Duplex = require("../Routing.Duplex/index.js");
var Routing_PushState = require("../Routing.PushState/index.js");
var Type_Equality = require("../Type.Equality/index.js");
var Web_HTML = require("../Web.HTML/index.js");
var Web_HTML_Location = require("../Web.HTML.Location/index.js");
var Web_HTML_Window = require("../Web.HTML.Window/index.js");
var Web_Router = require("../Web.Router/index.js");
var Web_Router_Driver_PushState = require("../Web.Router.Driver.PushState/index.js");
var Web_Router_Types = require("../Web.Router.Types/index.js");
var mkRoutingManager = function __do() {
    var $$interface = Routing_PushState.makeInterface();
    var v = $$interface.locationState();
    var value = Effect_Ref["new"](Data_Either.either(Data_Function["const"](Purescriptify_Data_Route["Error"].value))(Control_Category.identity(Control_Category.categoryFn))(Routing_Duplex.parse(Purescriptify_Data_Route.routeCodec)(v.path)))();
    var v1 = Halogen_Subscription.create();
    var driver = Web_Router_Driver_PushState.makeDriver_(Data_Foldable.foldableEither)(Routing_Duplex.parse(Purescriptify_Data_Route.routeCodec))(Routing_Duplex.print(Purescriptify_Data_Route.routeCodec))($$interface);
    var router = Web_Router.makeRouter(function (v2) {
        return function (v3) {
            return Web_Router["continue"];
        };
    })(function (v2) {
        if (v2 instanceof Web_Router_Types.Resolved) {
            return function __do() {
                var newRoute = Effect_Ref.modify(Data_Function["const"](v2.value1))(value)();
                return Halogen_Subscription.notify(v1.listener)(newRoute)();
            };
        };
        return Control_Applicative.pure(Effect.applicativeEffect)(Data_Unit.unit);
    })(driver)();
    var component = React_Basic_Hooks.component("Router")(function (v2) {
        return React_Basic_Hooks_Internal.discard(React_Basic_Hooks_Internal.ixBindRender)(React_Basic_Hooks.useEffectOnce(router.initialize))(function () {
            return Control_Applicative.pure(React_Basic_Hooks_Internal.applicativeRender(Type_Equality.refl))(React_Basic.empty);
        });
    })();
    return new Data_Tuple.Tuple({
        read: Effect_Ref.read(value),
        emitter: v1.emitter,
        navigate: router.navigate,
        redirect: router.redirect,
        reload: Control_Bind.bind(Effect.bindEffect)(Control_Bind.bind(Effect.bindEffect)(Web_HTML.window)(Web_HTML_Window.location))(Web_HTML_Location.reload)
    }, component(Data_Unit.unit));
};
module.exports = {
    mkRoutingManager: mkRoutingManager
};
