// Data.Functor.Nu/index.js
"use strict";
var Control_Comonad_Store = require("../Control.Comonad.Store/index.js");
var Data_Exists = require("../Data.Exists/index.js");
var Data_Function = require("../Data.Function/index.js");
var Data_Functor = require("../Data.Functor/index.js");
var NuF = function (x) {
    return x;
};

// | `Nu f` is the greatest fixed point of the functor `f`, when it exists.
var Nu = function (x) {
    return x;
};
var unfold = function (pos) {
    return function (peek) {
        return Data_Exists.mkExists(Control_Comonad_Store.store(peek)(pos));
    };
};
var observeF = function (dictFunctor) {
    return function (v) {
        var v1 = Control_Comonad_Store.runStore(v);
        return Data_Functor.map(dictFunctor)(Data_Function.flip(unfold)(v1.value0))(v1.value0(v1.value1));
    };
};
var observe = function (dictFunctor) {
    return function (v) {
        return Data_Exists.runExists(observeF(dictFunctor))(v);
    };
};
var newtypeNu = {
    Coercible0: function () {
        return undefined;
    }
};
module.exports = {
    Nu: Nu,
    NuF: NuF,
    unfold: unfold,
    observe: observe,
    newtypeNu: newtypeNu
};
