// Control.Comonad.Env/index.js

// | This module defines the `Env` comonad.
"use strict";
var Control_Comonad_Env_Class = require("../Control.Comonad.Env.Class/index.js");
var Control_Comonad_Env_Trans = require("../Control.Comonad.Env.Trans/index.js");
var Data_Functor = require("../Data.Functor/index.js");
var Data_Identity = require("../Data.Identity/index.js");
var Data_Newtype = require("../Data.Newtype/index.js");
var Data_Tuple = require("../Data.Tuple/index.js");

// | Change the environment type in an `Env` computation.
var withEnv = Control_Comonad_Env_Trans.withEnvT;

// | Unwrap a value in the `Env` comonad.
var runEnv = function (v) {
    return Data_Functor.map(Data_Tuple.functorTuple)(Data_Newtype.unwrap())(v);
};

// | Change the data type in an `Env` computation.
var mapEnv = Data_Functor.map(Control_Comonad_Env_Trans.functorEnvT(Data_Identity.functorIdentity));

// | Create a value in context in the `Env` comonad.
var env = function (e) {
    return function (a) {
        return new Data_Tuple.Tuple(e, a);
    };
};
module.exports = {
    runEnv: runEnv,
    withEnv: withEnv,
    mapEnv: mapEnv,
    env: env,
    ask: Control_Comonad_Env_Class.ask,
    asks: Control_Comonad_Env_Class.asks,
    local: Control_Comonad_Env_Class.local,
    EnvT: Control_Comonad_Env_Trans.EnvT,
    mapEnvT: Control_Comonad_Env_Trans.mapEnvT,
    runEnvT: Control_Comonad_Env_Trans.runEnvT,
    withEnvT: Control_Comonad_Env_Trans.withEnvT
};
