// Data.Lens.Zoom/index.js

// | This module defines functions for zooming in a state monad.
"use strict";
var Control_Monad_State_Trans = require("../Control.Monad.State.Trans/index.js");
var Data_Lens_Internal_Focusing = require("../Data.Lens.Internal.Focusing/index.js");
var Data_Lens_Types = require("../Data.Lens.Types/index.js");
var Data_Newtype = require("../Data.Newtype/index.js");
var Data_Profunctor_Star = require("../Data.Profunctor.Star/index.js");

// | Zooms into a substate in a `StateT` transformer.
var zoom = function (p) {
    var $0 = Data_Newtype.underF()()()()(Data_Lens_Internal_Focusing.Focusing)(Data_Newtype.under()()(Data_Profunctor_Star.Star)(p));
    return function ($1) {
        return Control_Monad_State_Trans.StateT($0(Control_Monad_State_Trans.runStateT($1)));
    };
};
module.exports = {
    zoom: zoom,
    Exchange: Data_Lens_Types.Exchange,
    Forget: Data_Lens_Types.Forget,
    Indexed: Data_Lens_Types.Indexed,
    Market: Data_Lens_Types.Market,
    Re: Data_Lens_Types.Re,
    Shop: Data_Lens_Types.Shop,
    Stall: Data_Lens_Types.Stall,
    Tagged: Data_Lens_Types.Tagged,
    wander: Data_Lens_Types.wander
};
